---
title: "Tangle Bearing Neurons Viewer"
output:
  flexdashboard::flex_dashboard:
    source_code: embed
runtime: shiny
---

```{r global, include = FALSE}
#    This file is part of tangle-bearing-neurons.
#    Copyright (C) 2024  Emir Turkes, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

library(conflicted)
library(scales)
library(ComplexHeatmap)
library(circlize)
library(DT)
library(GSEABase)

`%notin%` <- Negate(`%in%`)

data_sets <- c(
  "Laser-capture Microdissection Proteomics - Human AD BA9 - New Data",
  "Laser-capture Microdissection Proteomics - Human AD BA9 - Old Data",
  "Single-cell Transcriptomics - Human AD BA9"
)
data_types <- c(
  "log2 LFQ Intensity", "log2 LFQ Intensity", "log2 CPM",
  "log2 Enrichment Score", "log2 Enrichment Score", "log2 Enrichment Score"
)

data_dir <- file.path("..", "data")

gene_sets <- readRDS(file.path(data_dir, "gene_sets.rds"))
gene_anno <- readRDS(file.path(data_dir, "gene_anno.rds"))

meta <- vector("list", length = length(data_sets))
deg <- vector("list", length = length(data_sets))
gse <- vector("list", length = length(data_sets))
res_deg <- vector("list", length = length(data_sets))
res_gse <- vector("list", length = length(data_sets))

data_dir <- list.dirs(data_dir)
data_dir <- data_dir[c(2, 3, 4)]
for (i in seq_along(data_sets)) {
  meta[[i]] <- readRDS(file.path(data_dir[i], "meta.rds"))
  deg[[i]] <- readRDS(file.path(data_dir[i], "deg.rds"))
  gse[[i]] <- readRDS(file.path(data_dir[i], "gse.rds"))
  res_deg[[i]] <- readRDS(file.path(data_dir[i], "res_deg.rds"))
  res_gse[[i]] <- readRDS(file.path(data_dir[i], "res_gse.rds"))
}

top_anno <- vector("list", length = length(data_sets))
for (i in seq_along(data_sets)) {
  n <- length(unique(meta[[i]]$donor))
  anno_colour <- hue_pal()(n)
  top_anno[[i]] <- HeatmapAnnotation(
    Donor = meta[[i]]$donor,
    group = anno_block(
      gp = gpar(fill = "lightgray", col = "lightgray"),
      labels = unique(meta[[i]]$tau),
      labels_gp = gpar(col = "black", fontsize = 14),
      height = unit(6, "mm")
    ),
    col = list(Donor = setNames(anno_colour, unique(meta[[i]]$donor))),
    simple_anno_size = unit(4, "mm"),
    show_legend = c("Donor" = FALSE),
    annotation_name_gp = gpar(fontsize = 12)
  )
}

p_adj <- p.adjust(
  unlist(c(lapply(res_deg, `[[`, "P.value"), lapply(res_gse, `[[`, "P.value"))),
  method = "BH"
)
p_adj <- split(
  p_adj,
  f = rep(
    seq_along(c(res_deg, res_gse)),
    times = sapply(c(res_deg, res_gse), FUN = nrow)
  )
)
for (i in seq_along(res_deg)) {
  res_deg[[i]]$P.value.adj <- p_adj[[i]]
}
for (i in seq_along(res_gse)) {
  res_gse[[i]]$P.value.adj <- p_adj[[i + length(data_sets)]]
}

priority_gse <- c(
  `Tau & Neurofibrillary Tangle` =
    "tau-protein | tau protein | neurofi | inclusion | aggrep",
  `Microtubule & Cytoskeletal` =
    "microtubule | dynein | cytoskele | laminin",
  `Death, Stress, & Ageing` =
    "death | necro | senescen | stress",
  `Protein Folding` =
    "folding | folded",
  `Synapse, Axon, & Dendrite` =
    "synap | axon | dendrit",
  `Exocytosis & Endocytosis` =
    "exocyt | endocyt",
  `Apoptosis & Autophagy`=
    "apopto | autophag | phagocyt",
  `Proteasomal & Lysosomal`=
    "proteasom | lysosom | endoplasmic | Golgi | ESCRT | ubiqui",
  `Heat Shock`=
    "Hsp | heat shock",
  `Glycosaminoglycan`=
    "heparan | chondroitin | sulfotransferase | glycosaminoglycan",
  `Transcription & Translation` =
    "transcription | translation",
  `Amyloid & Lipoprotein` =
    "amyloid | lipoprotein",
  `Lipid Metabolism`=
    "fatty | cholesterol | insulin",
  `Synaptic Pruning` =
    "pruning | phosphatidylserine | complement",
  `T Cell, B Cell, & Other Immune Cell`=
    "T cell | B cell | leukocyte | myeloid | lymphocyte",
  `Inflammatory Response` =
    "immun | inflammat | cortisol | interleukin | NF-kappaB | toll-like",
  `Macrophage & Astrocyte` =
    "macrophage | astrocyte",
  `Oligodendrocyte & Myelin` =
    "oligodendrocyte | Schwann | myelin"
)

range_gse <- vector("list", length = length(data_sets))
sig_gse <- vector("list", length = length(data_sets))
top_gse <- vector("list", length = length(data_sets))
n_gse <- vector("list", length = length(data_sets))

for (i in seq_along(data_sets)) {

  range_gse[[i]] <- c(
    min(gse[[i]]), (min(gse[[i]]) + max(gse[[i]])) / 2, max(gse[[i]])
  )

  sig <- ifelse(res_gse[[i]]$P.value.adj < 0.05, yes = 1, no = 0)
  direction <- ifelse(res_gse[[i]]$Coef > 0, yes = 1, no = -1)
  sig <- sig * direction

  names(sig) <- rownames(res_gse[[i]])
  sig <- sig[c(which(sig == 1), which(sig == -1), which(sig == 0))]
  sig <- ifelse(sig == -1, yes = "↓", no = ifelse(sig == 1, yes = "↑", no = ""))

  sig <- data.frame(CTRL = rep("", times = length(sig)), NFT = sig)
  sig_gse[[i]] <- vector("list", length = 2)
  for (j in seq_along(sig)) {
    sig_gse[[i]][[j]] <- replicate(nrow(meta[[i]]) / 2, expr = sig[ , j])
  }
  sig_gse[[i]] <- do.call(cbind, args = sig_gse[[i]])

  res_gse[[i]] <- res_gse[[i]][match(rownames(sig), rownames(res_gse[[i]])), ]
  rownames(sig_gse[[i]]) <- res_gse[[i]]$GeneSet

  top_gse[[i]] <- sapply(
    priority_gse,
    function (pattern) {
      individual_patterns <- unlist(strsplit(pattern, " \\| "))
      matches <- sapply(
        individual_patterns, function(individual_pattern) {
          grep(
            paste0(".*", individual_pattern, ".*"),
            rownames(res_gse[[i]][res_gse[[i]]$P.value.adj < 0.05, ]),
            value = TRUE
          )
        }
      )
      unique(unlist(matches))
    }
  )

}

range_deg <- vector("list", length = length(data_sets))
sig_deg <- vector("list", length = length(data_sets))
top_deg <- vector("list", length = length(data_sets))
n_deg <- vector("list", length = length(data_sets))

for (i in seq_along(data_sets)) {

  range_deg[[i]] <- c(
    min(deg[[i]]), (min(deg[[i]]) + max(deg[[i]])) / 2, max(deg[[i]])
  )

  sig <- ifelse(res_deg[[i]]$P.value.adj < 0.05, yes = 1, no = 0)
  direction <- ifelse(res_deg[[i]]$Coef > 0, yes = 1, no = -1)
  sig <- sig * direction

  names(sig) <- rownames(res_deg[[i]])
  sig <- sig[c(which(sig == 1), which(sig == -1), which(sig == 0))]
  sig <- ifelse(sig == -1, yes = "↓", no = ifelse(sig == 1, yes = "↑", no = ""))

  sig <- data.frame(CTRL = rep("", times = length(sig)), NFT = sig)
  sig_deg[[i]] <- vector("list", length = 2)
  for (j in seq_along(sig)) {
    sig_deg[[i]][[j]] <- replicate(nrow(meta[[i]]) / 2, expr = sig[ , j])
  }
  sig_deg[[i]] <- do.call(cbind, args = sig_deg[[i]])

  res_deg[[i]] <- res_deg[[i]][match(rownames(sig), rownames(res_deg[[i]])), ]
  rownames(sig_deg[[i]]) <- res_deg[[i]]$Gene

}

datatable_download_exp_15 <- function(dt) {
  datatable(
    dt,
    options = list(
      pageLength = 15,
      scrollX = TRUE,
      dom = "Blfrtip",
      buttons = list(
        "copy", "print",
        list(
          extend = "collection", buttons = c("csv", "excel", "pdf"),
          text = "Download"
        )
      ),
      rowCallback = JS(
        "function(row, data) {",
        "for (i = 1; i < data.length; i++) {",
        "if (data[i]>=1000 | data[i]<1000) {",
        "$('td:eq('+i+')', row).html(data[i].toExponential(2));}}}"
      )
    ),
    extensions = "Buttons"
  )
}
```

Single-cell LCM Proteomics - Pathways
===

```{r}
idx <- 1
sub_idx <- 0
```

Row {.tabset}
---

### Tau & Neurofibrillary Tangle

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 1
    comb_idx <- 11

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Microtubule & Cytoskeletal

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 2
    comb_idx <- 12

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Death, Stress, & Ageing

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 3
    comb_idx <- 13

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Protein Folding

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 4
    comb_idx <- 14

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Synapse, Axon, & Dendrite

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 5
    comb_idx <- 15

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Exocytosis & Endocytosis

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 6
    comb_idx <- 16

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Apoptosis & Autophagy

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 7
    comb_idx <- 17

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Proteasomal & Lysosomal

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 8
    comb_idx <- 18

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Heat Shock

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 9
    comb_idx <- 19

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Glycosaminoglycan

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 10
    comb_idx <- 110

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Transcription & Translation

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 11
    comb_idx <- 111

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Amyloid & Lipoprotein

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 12
    comb_idx <- 112

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Lipid Metabolism

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 13
    comb_idx <- 113

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Synaptic Pruning

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 14
    comb_idx <- 114

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### T Cell, B Cell, & Other Immune Cell

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 15
    comb_idx <- 115

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Inflammatory Response

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 16
    comb_idx <- 116

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Macrophage & Astrocyte

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 17
    comb_idx <- 117

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Oligodendrocyte & Myelin

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1
    sub_idx <- 18
    comb_idx <- 118

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

Pooled LCM Proteomics - Pathways
===

```{r}
idx <- 2
sub_idx <- 0
```

Row {.tabset}
---

### Tau & Neurofibrillary Tangle

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 1
    comb_idx <- 21

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Microtubule & Cytoskeletal

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 2
    comb_idx <- 22

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Death, Stress, & Ageing

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 3
    comb_idx <- 23

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Protein Folding

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 4
    comb_idx <- 24

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Synapse, Axon, & Dendrite

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 5
    comb_idx <- 25

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Exocytosis & Endocytosis

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 6
    comb_idx <- 26

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Apoptosis & Autophagy

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 7
    comb_idx <- 27

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Proteasomal & Lysosomal

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 8
    comb_idx <- 28

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Heat Shock

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 9
    comb_idx <- 29

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Glycosaminoglycan

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 10
    comb_idx <- 210

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Transcription & Translation

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 11
    comb_idx <- 211

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Amyloid & Lipoprotein

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 12
    comb_idx <- 212

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Lipid Metabolism

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 13
    comb_idx <- 213

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Synaptic Pruning

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 14
    comb_idx <- 214

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### T Cell, B Cell, & Other Immune Cell

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 15
    comb_idx <- 215

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Inflammatory Response

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 16
    comb_idx <- 216

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Macrophage & Astrocyte

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 17
    comb_idx <- 217

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Oligodendrocyte & Myelin

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2
    sub_idx <- 18
    comb_idx <- 218

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

Single-cell FACS Transcriptomics - Pathways
===

```{r}
idx <- 3
sub_idx <- 0
```

Row {.tabset}
---

### Tau & Neurofibrillary Tangle

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 1
    comb_idx <- 31

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Microtubule & Cytoskeletal

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 2
    comb_idx <- 32

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Death, Stress, & Ageing

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 3
    comb_idx <- 33

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Protein Folding

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 4
    comb_idx <- 34

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Synapse, Axon, & Dendrite

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 5
    comb_idx <- 35

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Exocytosis & Endocytosis

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 6
    comb_idx <- 36

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Apoptosis & Autophagy

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 7
    comb_idx <- 37

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Proteasomal & Lysosomal

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 8
    comb_idx <- 38

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Heat Shock

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 9
    comb_idx <- 39

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Glycosaminoglycan

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 10
    comb_idx <- 310

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Transcription & Translation

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 11
    comb_idx <- 311

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Amyloid & Lipoprotein

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 12
    comb_idx <- 312

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Lipid Metabolism

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 13
    comb_idx <- 313

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Synaptic Pruning

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 14
    comb_idx <- 314

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### T Cell, B Cell, & Other Immune Cell

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 15
    comb_idx <- 315

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Inflammatory Response

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 16
    comb_idx <- 316

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Macrophage & Astrocyte

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 17
    comb_idx <- 317

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

### Oligodendrocyte & Myelin

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

selectizeInput(
  paste0("gse", comb_idx),
  label = "Type to Search Pathways:",
  choices = top_gse[[idx]][[sub_idx]],
  selected = top_gse[[idx]][[sub_idx]][1:20],
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", comb_idx), label = NULL,
  c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 3
    sub_idx <- 18
    comb_idx <- 318

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", comb_idx)]], ,
      drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", comb_idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", comb_idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          "Significant ", names(priority_gse)[sub_idx], " Pathways ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1450,
  height = 615,
  res = 96
)
```

</div>
</div>

```{r, include = FALSE}
rm(data_dir, n, anno_colour, p_adj, direction, priority_gse, priority_deg)
gc()
```
