---
title: "Tangle Bearing Neurons Viewer"
output:
  flexdashboard::flex_dashboard:
    orientation: rows
runtime: shiny
---

```{r global, include = FALSE}
#    This file is part of tangle-bearing-neurons.
#    Copyright (C) 2024  Emir Turkes, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

library(conflicted)
conflicts_prefer(igraph::degree, base::intersect, DT::JS, graphics::layout)
library(scales)
library(InteractiveComplexHeatmap)
library(circlize)
library(GSEABase)
library(reshape2)
library(dplyr)
library(igraph)
library(qgraph)
library(visNetwork)
library(purrr)
library(networkD3)
library(DT)
library(tidyr)
library(ggplot2)
library(plotly)
library(scuttle)
library(flextable)
library(eulerr)
library(RColorBrewer)

`%notin%` <- Negate(`%in%`)

datatable_download_exp_15 <- function(dt) {
  datatable(
    dt,
    options = list(
      pageLength = 15,
      scrollX = TRUE,
      dom = "Blfrtip",
      buttons = list(
        "copy", "print",
        list(
          extend = "collection", buttons = c("csv", "excel", "pdf"),
          text = "Download"
        )
      ),
      rowCallback = JS(
        "function(row, data) {",
        "for (i = 1; i < data.length; i++) {",
        "if (data[i]>=1000 | data[i]<1000) {",
        "$('td:eq('+i+')', row).html(data[i].toExponential(2));}}}"
      )
    ),
    extensions = "Buttons"
  )
}

network_filt <- function(thresh, weight = 0.25) {
  out1 <- 0.5 + (weight * (thresh - 0.5))
  out2 <- 0.5 - (weight * (thresh - 0.5))
  list(out1 = out1, out2 = out2)
}

scale_range <- function(x, from = c(1, 45), to = c(8, 14)) {
  to[1] + (from[2] - x) * diff(to) / diff(from)
}

map_range <- function (x) {
  ifelse(
    x <= 0.3, yes = 1, no = ifelse(x >= 0.7, yes = 4, no = round(1 + x * 3))
  )
}

add_newline <- function(string) {
  words <- strsplit(string, split = " ")[[1]]
  chunks <- split(words, f = ceiling(seq_along(words) / 2))
  paste(sapply(chunks, FUN = paste, collapse = " "), collapse = "\n")
}

data_sets <- c(
  "LCM Mass Spec - Human AD BA9 - Single-cell",
  "LCM Mass Spectrometry - Human AD BA9",
  "FACS-sorted ssRNAseq - Human AD BA9"
)
data_types <- c(
  "log2 LFQ Intensity", "log2 LFQ Intensity", "log2 CPM",
  "log2 Enrichment Score", "log2 Enrichment Score", "log2 Enrichment Score"
)

data_dir <- file.path("..", "data")
cache_dir <- file.path("..", "cache")
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}

gene_sets <- readRDS(file.path(data_dir, "gene_sets.rds"))
gene_anno <- readRDS(file.path(data_dir, "gene_anno.rds"))
map <- setNames(gene_anno$symbol, gene_anno$ensembl)

meta <- vector("list", length = length(data_sets))
deg <- vector("list", length = length(data_sets))
gse <- vector("list", length = length(data_sets))
res_deg <- vector("list", length = length(data_sets))
res_gse <- vector("list", length = length(data_sets))
res_gse_fmt <- vector("list", length = length(data_sets))

data_dir <- list.dirs(data_dir)
data_dir <- data_dir[c(2, 3, 4)]
for (i in seq_along(data_sets)) {
  meta[[i]] <- readRDS(file.path(data_dir[i], "meta.rds"))
  deg[[i]] <- readRDS(file.path(data_dir[i], "deg.rds"))
  gse[[i]] <- readRDS(file.path(data_dir[i], "gse.rds"))
  res_deg[[i]] <- readRDS(file.path(data_dir[i], "res_deg.rds"))
  res_gse[[i]] <- readRDS(file.path(data_dir[i], "res_gse.rds"))

  res_gse_fmt[[i]] <- res_gse[[i]]
  gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(res_gse_fmt[[i]])]
  gene_sets_sub <- gene_sets_sub[
    match(rownames(res_gse_fmt[[i]]), names(gene_sets_sub))
  ]
  id <- c()
  for (j in seq_along(gene_sets_sub)) {
    id <- c(id, gene_sets[[j]]@shortDescription)
  }
  rownames(res_gse_fmt[[i]]) <- id
  res_gse_fmt[[i]] <- res_gse_fmt[[i]][ , c(6, 1:5)]
}

top_anno <- vector("list", length = length(data_sets))
for (i in seq_along(data_sets)) {
  n <- length(unique(meta[[i]]$donor))
  anno_colour <- hue_pal()(n)
  top_anno[[i]] <- HeatmapAnnotation(
    Donor = meta[[i]]$donor,
    group = anno_block(
      gp = gpar(fill = "lightgray", col = "lightgray"),
      labels = unique(meta[[i]]$tau),
      labels_gp = gpar(col = "black", fontsize = 14),
      height = unit(6, "mm")
    ),
    col = list(Donor = setNames(anno_colour, unique(meta[[i]]$donor))),
    simple_anno_size = unit(4, "mm"),
    show_legend = c("Donor" = FALSE),
    annotation_name_gp = gpar(fontsize = 12)
  )
}

p_adj <- p.adjust(
  unlist(c(lapply(res_deg, `[[`, "P.value"), lapply(res_gse, `[[`, "P.value"))),
  method = "BH"
)
p_adj <- split(
  p_adj,
  f = rep(
    seq_along(c(res_deg, res_gse)),
    times = sapply(c(res_deg, res_gse), FUN = nrow)
  )
)
for (i in seq_along(res_deg)) {
  res_deg[[i]]$P.value.adj <- p_adj[[i]]
}
for (i in seq_along(res_gse)) {
  res_gse[[i]]$P.value.adj <- p_adj[[i + length(data_sets)]]
}

hubs <- vector("list", length = length(data_sets))
hubs[[2]] <- c(
  "NEFM", "APP", "SQSTM1", "HSP90AA1", "YWHAE", "WASF1", "CNTNAP1", "GOT2"
)
hubs[[3]] <- c(
  "NEFM", "APP", "SQSTM1", "HSP90AA1", "YWHAE", "WASF1", "CNTNAP1", "GOT2"
)

priority_gse <- vector("list", length = length(data_sets))
priority_gse[[1]] <- NULL
priority_gse[[2]] <- list(
  `NEFM Associated` = "NEFM",
  `APP Associated` = "APP",
  `SQSTM1 Associated` = "SQSTM1",
  `HSP90AA1 Associated` = "HSP90AA1",
  `YWHAE Associated` = "YWHAE",
  `WASF1 Associated` = "WASF1",
  `CNTNAP1 Associated` = "CNTNAP1",
  `GOT2 Associated`= "GOT2"
)
priority_gse[[3]] <- list(
  `NEFM Associated` = "NEFM",
  `APP Associated` = "APP",
  `SQSTM1 Associated` = "SQSTM1",
  `HSP90AA1 Associated` = "HSP90AA1",
  `YWHAE Associated` = "YWHAE",
  `WASF1 Associated` = "WASF1",
  `CNTNAP1 Associated` = "CNTNAP1",
  `GOT2 Associated`= "GOT2"
)

for (i in seq_along(priority_gse)) {
  for (j in seq_along(priority_gse[[i]])) {
    gene_anno_sub <- gene_anno[gene_anno$symbol %in% priority_gse[[i]][[j]], ]
    keep <- which(
      sapply(
        geneIds(gene_sets), FUN = function(x) any(x %in% gene_anno_sub$ensembl)
      )
    )
    priority_gse[[i]][[j]] <- names(keep)
  }
}


if (file.exists(file.path(cache_dir, "range_gse.rds"))) {
  range_gse <- readRDS(file.path(cache_dir, "range_gse.rds"))
  sig_gse <- readRDS(file.path(cache_dir, "sig_gse.rds"))
  top_gse <- readRDS(file.path(cache_dir, "top_gse.rds"))
  range_deg <- readRDS(file.path(cache_dir, "range_deg.rds"))
  sig_deg <- readRDS(file.path(cache_dir, "sig_deg.rds"))
  top_deg <- readRDS(file.path(cache_dir, "top_deg.rds"))
  pseudobulk_deg <- readRDS(file.path(cache_dir, "pseudobulk_deg.rds"))
} else {

  range_gse <- vector("list", length = length(data_sets))
  sig_gse <- vector("list", length = length(data_sets))
  top_gse <- vector("list", length = length(data_sets))
  n_gse <- vector("list", length = length(data_sets))
  pseudobulk_deg <- vector("list", length = length(data_sets))

  for (i in seq_along(data_sets)) {

    range_gse[[i]] <- c(
      min(gse[[i]]), (min(gse[[i]]) + max(gse[[i]])) / 2, max(gse[[i]])
    )

    sig <- ifelse(res_gse[[i]]$P.value.adj < 0.05, yes = 1, no = 0)
    direction <- ifelse(res_gse[[i]]$Coef > 0, yes = 1, no = -1)
    sig <- sig * direction

    names(sig) <- rownames(res_gse[[i]])
    sig <- sig[c(which(sig == 1), which(sig == -1), which(sig == 0))]
    sig <- ifelse(
      sig == -1, yes = "↓", no = ifelse(sig == 1, yes = "↑", no = "")
    )

    sig <- data.frame(CTRL = rep("", times = length(sig)), NFT = sig)
    sig_gse[[i]] <- vector("list", length = 2)
    for (j in seq_along(sig)) {
      sig_gse[[i]][[j]] <- replicate(nrow(meta[[i]]) / 2, expr = sig[ , j])
    }
    sig_gse[[i]] <- do.call(cbind, args = sig_gse[[i]])

    res_gse[[i]] <- res_gse[[i]][match(rownames(sig), rownames(res_gse[[i]])), ]
    rownames(sig_gse[[i]]) <- res_gse[[i]]$GeneSet

    top_gse[[i]] <- sapply(
      priority_gse[[i]],
      function (x) {
        patterns <- x
        matches <- sapply(
          patterns, function(y) {
            grep(
              paste0("\\b", y, "\\b"),
              rownames(res_gse[[i]][res_gse[[i]]$P.value.adj < 0.05, ]),
              value = TRUE
            )
          }
        )
        unique(unlist(matches))
      }
    )

    top_gse[[i]]$`Other Significant` <-
      rownames(res_gse[[i]][res_gse[[i]]$P.value.adj < 0.05, ])[
        rownames(res_gse[[i]][res_gse[[i]]$P.value.adj < 0.05, ]) %notin%
          unlist(top_gse[[i]])
      ]
    top_gse[[i]]$`All Pathways` <- rownames(res_gse[[i]])
  }

  range_deg <- vector("list", length = length(data_sets))
  sig_deg <- vector("list", length = length(data_sets))
  top_deg <- vector("list", length = length(data_sets))
  n_deg <- vector("list", length = length(data_sets))

  for (i in seq_along(data_sets)) {

    range_deg[[i]] <- c(
      min(deg[[i]]), (min(deg[[i]]) + max(deg[[i]])) / 2, max(deg[[i]])
    )

    sig <- ifelse(res_deg[[i]]$P.value.adj < 0.05, yes = 1, no = 0)
    direction <- ifelse(res_deg[[i]]$Coef > 0, yes = 1, no = -1)
    sig <- sig * direction

    names(sig) <- rownames(res_deg[[i]])
    sig <- sig[c(which(sig == 1), which(sig == -1), which(sig == 0))]
    sig <- ifelse(
      sig == -1, yes = "↓", no = ifelse(sig == 1, yes = "↑", no = "")
    )

    sig <- data.frame(CTRL = rep("", times = length(sig)), NFT = sig)
    sig_deg[[i]] <- vector("list", length = 2)
    for (j in seq_along(sig)) {
      sig_deg[[i]][[j]] <- replicate(nrow(meta[[i]]) / 2, expr = sig[ , j])
    }
    sig_deg[[i]] <- do.call(cbind, args = sig_deg[[i]])

    res_deg[[i]] <- res_deg[[i]][match(rownames(sig), rownames(res_deg[[i]])), ]
    rownames(sig_deg[[i]]) <- res_deg[[i]]$Gene

    top_deg[[i]] <- res_deg[[i]][res_deg[[i]]$P.value.adj < 0.05, ]$Gene
    n_deg[[i]] <- length(top_deg[[i]])

    pseudobulk_deg[[i]] <- SingleCellExperiment(list(logcounts = deg[[i]]))
    pseudobulk_deg[[i]] <- aggregateAcrossCells(
      pseudobulk_deg[[i]], meta[[i]]$tau,
      use_exprs_values = "logcounts", statistics = "mean"
    )

  }
  saveRDS(range_gse, file = file.path(cache_dir, "range_gse.rds"))
  saveRDS(sig_gse, file = file.path(cache_dir, "sig_gse.rds"))
  saveRDS(top_gse, file = file.path(cache_dir, "top_gse.rds"))
  saveRDS(range_deg, file = file.path(cache_dir, "range_deg.rds"))
  saveRDS(sig_deg, file = file.path(cache_dir, "sig_deg.rds"))
  saveRDS(top_deg, file = file.path(cache_dir, "top_deg.rds"))
  saveRDS(pseudobulk_deg, file = file.path(cache_dir, "pseudobulk_deg.rds"))
}

priority_gse[[1]] <- c(
  priority_gse[[1]], `Other Significant` = NA, `All Pathways` = NA
)
priority_gse[[2]] <- c(
  priority_gse[[2]], `Other Significant` = NA, `All Pathways` = NA
)
priority_gse[[3]] <- c(
  priority_gse[[3]], `Other Significant` = NA, `All Pathways` = NA
)

n_gse <- vector("list", length = length(data_sets))
n_gse_up <- vector("list", length = length(data_sets))
n_gse_dn <- vector("list", length = length(data_sets))
n_deg <- vector("list", length = length(data_sets))
n_deg_up <- vector("list", length = length(data_sets))
n_deg_dn <- vector("list", length = length(data_sets))
for (i in seq_along(n_gse)) {
  n_gse[[i]] <- res_gse[[i]][res_gse[[i]]$P.value.adj < 0.05, ]
  n_gse_up[[i]] <- nrow(n_gse[[i]][n_gse[[i]]$Coef > 0, ])
  n_gse_dn[[i]] <- nrow(n_gse[[i]][n_gse[[i]]$Coef < 0, ])
  n_gse[[i]] <- nrow(n_gse[[i]])
  n_deg[[i]] <- res_deg[[i]][res_deg[[i]]$P.value.adj < 0.05, ]
  n_deg_up[[i]] <- nrow(n_deg[[i]][n_deg[[i]]$Coef > 0, ])
  n_deg_dn[[i]] <- nrow(n_deg[[i]][n_deg[[i]]$Coef < 0, ])
  n_deg[[i]] <- nrow(n_deg[[i]])
}

venn_col <- brewer.pal(3, name = "Pastel2")[1:2]

res <- data.frame(
  a = c("LCM\nMass Spec", "FACS\nssRNAseq"),
  b = c(n_gse[[2]], n_gse[[3]]),
  c = c(n_deg[[2]], n_deg[[3]]),
  d = c(n_gse_up[[2]], n_gse_up[[3]]),
  e = c(n_gse_dn[[2]], n_gse_dn[[3]]),
  f = c(n_deg_up[[2]], n_deg_up[[3]]),
  g = c(n_deg_dn[[2]], n_deg_dn[[3]]),
  h = c(nrow(res_gse[[2]]), nrow(res_gse[[3]])),
  i = c(nrow(res_deg[[2]]), nrow(res_deg[[3]])),
  j = c(
    percent(n_gse[[2]] / nrow(res_gse[[2]])),
    percent(n_gse[[3]] / nrow(res_gse[[3]]))
  ),
  k = c(
    percent(n_deg[[2]] / nrow(res_deg[[2]])),
    percent(n_deg[[3]] / nrow(res_deg[[3]]))
  )
)
colnames(res) <- c(
  "Dataset",
  "Pathway\nSets\nDifferentially Enriched",
  "Genes/Proteins\nDifferential Expressed",
  "Pathway\nSets\nDE UP",
  "Pathway\nSets\nDE DOWN",
  "Genes/Proteins\nDE UP",
  "Genes/Proteins\nDE DOWN",
  "Pathway\nSets\nTested",
  "Genes/Proteins\nTested",
  "Percentage\nPathway\nSets\nDE",
  "Percentage\nGenes/Proteins\nDE"
)

venn_list_gse <- vector("list", length = 2)
names(venn_list_gse) <- res$Dataset
venn_list_gse[[1]] <- res_gse[[2]][res_gse[[2]]$P.value.adj < 0.05, ]$GeneSet
venn_list_gse[[2]] <- res_gse[[3]][res_gse[[3]]$P.value.adj < 0.05, ]$GeneSet

venn_list_deg <- vector("list", length = 2)
names(venn_list_deg) <- res$Dataset
venn_list_deg[[1]] <- res_deg[[2]][res_deg[[2]]$P.value.adj < 0.05, ]$Gene
venn_list_deg[[2]] <- res_deg[[3]][res_deg[[3]]$P.value.adj < 0.05, ]$Gene
```

Network Analysis
===

Row {.tabset}
---

### Upregulated in Tangle-bearing Neurons

```{r}
shinyApp(

  ui = fluidPage(
    fluidRow(
      column(
        width = 2,
        sliderInput(
          paste0("conn_up", 23),
          label = "Stringency:",
          min = 0,
          max = 1,
          value = 0.9,
          step = 0.1
        ),
      ),
      column(
        width = 2,
        radioButtons(
          paste0("shared_up", 23), label = NULL,
          c("All Features", "Shared Features Only"), inline = TRUE
        ),
        style = "padding-top: 35px;"
      ),
      column(
        width = 2,
        radioButtons(
          paste0("comp_up", 23), label = NULL,
          c("All Networks", "Main Network Only"), inline = TRUE
        ),
        style = "padding-top: 35px;"
      ),
      column(
        width = 2,
        sliderInput(
          paste0("degree_up", 23),
          label = "Connectivity:",
          min = 0,
          max = 1,
          value = 0.9,
          step = 0.1
        ),
      )
    ),
    fluidPage(
      div(
        style = "display:flex;",
        visNetworkOutput(paste0("vis_up", 23), height = "650px", width = "50%"),
        visNetworkOutput(paste0("vis2_up", 23), height = "650px", width = "50%")
      )
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("conn_up", 23)]]
      input[[paste0("shared_up", 23)]]
      input[[paste0("comp_up", 23)]]
      input[[paste0("degree_up", 23)]]
    }, {

      inverse <- round(1 - input[[paste0("conn_up", 23)]], digits = 2)

      de_gse <- rownames(res_gse[[2]][res_gse[[2]]$P.value.adj < 0.05, ])
      up_gse <- rownames(res_gse[[2]][res_gse[[2]]$Coef > 0, ])
      de_deg <- rownames(res_deg[[2]][res_deg[[2]]$P.value.adj < 0.05, ])
      up_deg <- rownames(res_deg[[2]][res_deg[[2]]$Coef > 0, ])

      net_data <- melt(geneIds(gene_sets[names(gene_sets) %in% de_gse]))
      names(net_data)[1] <- "ensembl"
      names(net_data)[2] <- "gene_set"

      gene_anno_sub <- gene_anno[gene_anno$ensembl %in% net_data$ensembl, ]
      net_data <- left_join(net_data, gene_anno_sub, by = "ensembl")
      net_data$gse_direction <- ifelse(
        net_data$gene_set %in% up_gse, yes = "up", no = "down"
      )
      net_data$gene_DE <- ifelse(
        net_data$ensembl %in% de_deg, yes = "yes", no = "no"
      )
      net_data$gene_direction <- ifelse(
        net_data$ensembl %in% up_deg, yes = "up", no = "down"
      )
      net_data$dataset <- "LCM Mass Spec"

      remove <- which(net_data$gse_direction != net_data$gene_direction)
      net_data <- net_data[-remove, ]
      net_data <- net_data[net_data$gene_DE == "yes", ]

      net_data <- net_data[which(net_data$gse_direction == "up"), ]
      net_data2 <- net_data
      net_data2_orig <- net_data2

      de_gse <- rownames(res_gse[[3]][res_gse[[3]]$P.value.adj < 0.05, ])
      up_gse <- rownames(res_gse[[3]][res_gse[[3]]$Coef > 0, ])
      de_deg <- rownames(res_deg[[3]][res_deg[[3]]$P.value.adj < 0.05, ])
      up_deg <- rownames(res_deg[[3]][res_deg[[3]]$Coef > 0, ])

      net_data <- melt(geneIds(gene_sets[names(gene_sets) %in% de_gse]))
      names(net_data)[1] <- "ensembl"
      names(net_data)[2] <- "gene_set"

      gene_anno_sub <- gene_anno[gene_anno$ensembl %in% net_data$ensembl, ]
      net_data <- left_join(net_data, gene_anno_sub, by = "ensembl")
      net_data$gse_direction <- ifelse(
        net_data$gene_set %in% up_gse, yes = "up", no = "down"
      )
      net_data$gene_DE <- ifelse(
        net_data$ensembl %in% de_deg, yes = "yes", no = "no"
      )
      net_data$gene_direction <- ifelse(
        net_data$ensembl %in% up_deg, yes = "up", no = "down"
      )
      net_data$dataset <- "FACS ssRNAseq"

      remove <- which(net_data$gse_direction != net_data$gene_direction)
      net_data <- net_data[-remove, ]
      net_data <- net_data[net_data$gene_DE == "yes", ]

      net_data <- net_data[which(net_data$gse_direction == "up"), ]
      net_data3 <- net_data
      net_data3_orig <- net_data3

      net_data <- rbind(net_data2, net_data3)
      net_data_table2 <- table(net_data2$symbol)
      net_data_table3 <- table(net_data3$symbol)

      gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(gse[[2]])]
      gene_sets_table2 <- table(unlist(geneIds(gene_sets_sub)))
      gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(gse[[3]])]
      gene_sets_table3 <- table(unlist(geneIds(gene_sets_sub)))

      gene_anno_sub <- gene_anno[
        gene_anno$ensembl %in% names(gene_sets_table2),
      ]
      gene_anno_sub <- gene_anno_sub[
        match(names(gene_sets_table2), gene_anno_sub$ensembl),
      ]
      names(gene_sets_table2) <- gene_anno_sub$symbol

      gene_anno_sub <- gene_anno[
        gene_anno$ensembl %in% names(gene_sets_table3),
      ]
      gene_anno_sub <- gene_anno_sub[
        match(names(gene_sets_table3), gene_anno_sub$ensembl),
      ]
      names(gene_sets_table3) <- gene_anno_sub$symbol

      gene_sets_table2 <- gene_sets_table2[
        names(gene_sets_table2) %in% names(net_data_table2)
      ]
      gene_sets_table3 <- gene_sets_table3[
        names(gene_sets_table3) %in% names(net_data_table3)
      ]
      gene_sets_table2 <- gene_sets_table2[
        match(names(net_data_table2), names(gene_sets_table2))
      ]
      gene_sets_table3 <- gene_sets_table3[
        match(names(net_data_table3), names(gene_sets_table3))
      ]

      pval2 <- res_deg[[2]]
      pval2 <- pval2[pval2$Gene %in% names(net_data_table2), ]
      pval2 <- pval2[match(names(net_data_table2), pval2$Gene), ]
      pval2 <- pval2$Coef
      pval2 <- (pval2 - min(pval2)) / (max(pval2) - min(pval2))
      pval2 <- rank(pval2)
      pval2 <- ((pval2 - min(pval2)) / (max(pval2) - min(pval2))) / 100000000000
      names(pval2) <- names(net_data_table2)

      pval3 <- res_deg[[3]]
      pval3 <- pval3[pval3$Gene %in% names(net_data_table3), ]
      pval3 <- pval3[match(names(net_data_table3), pval3$Gene), ]
      pval3 <- pval3$Coef
      pval3 <- (pval3 - min(pval3)) / (max(pval3) - min(pval3))
      pval3 <- rank(pval3)
      pval3 <- ((pval3 - min(pval3)) / (max(pval3) - min(pval3))) / 100000000000
      names(pval3) <- names(net_data_table3)

      dup2 <- c()
      for (gene in unique(net_data2$symbol)) {
        matched_gse <- net_data2[net_data2$symbol %in% gene, ]$gene_set
        net_data_sub <- net_data2[net_data2$gene_set %in% matched_gse, ]
        net_data_sub <- net_data_sub[net_data_sub$symbol %notin% gene, ]
        dup_table <- table(net_data_sub$symbol)
        if (length(dup_table) > 0) {
          dup2 <- c(dup2, mean(dup_table))
        } else {
          dup2 <- c(dup2, 1)
        }
      }
      names(dup2) <- unique(net_data2$symbol)
      dup2 <- dup2[match(names(net_data_table2), names(dup2))]

      dup3 <- c()
      set_size3 <- c()
      for (gene in unique(net_data3$symbol)) {
        matched_gse <- net_data3[net_data3$symbol %in% gene, ]$gene_set
        net_data_sub <- net_data3[net_data3$gene_set %in% matched_gse, ]
        net_data_sub <- net_data_sub[net_data_sub$symbol %notin% gene, ]
        dup_table <- table(net_data_sub$symbol)
        if (length(dup_table) > 0) {
          dup3 <- c(dup3, mean(dup_table))
        } else {
          dup3 <- c(dup3, 1)
        }
      }
      names(dup3) <- unique(net_data3$symbol)
      dup3 <- dup3[match(names(net_data_table3), names(dup3))]

      net_data_table2 <- (
        ((net_data_table2 / gene_sets_table2) * net_data_table2) / dup2
      ) + pval2
      net_data_table3 <- (
        ((net_data_table3 / gene_sets_table3) * net_data_table3) / dup3
      ) + pval3
      net_data_table2_scaled <- net_data_table2
      net_data_table3_scaled <- net_data_table3
      net_data_table2_scaled <- rank(net_data_table2_scaled)
      net_data_table3_scaled <- rank(net_data_table3_scaled)
      net_data_table2_scaled <- (
        (net_data_table2_scaled - min(net_data_table2_scaled)) /
          (max(net_data_table2_scaled) - min(net_data_table2_scaled))
      )
      net_data_table3_scaled <- (
        (net_data_table3_scaled - min(net_data_table3_scaled)) /
          (max(net_data_table3_scaled) - min(net_data_table3_scaled))
      )

      keep2 <- net_data_table2_scaled[
        net_data_table2_scaled > (
          inverse * ((length(net_data_table2) / length(net_data_table3)) * 1)
        ) + ((length(net_data_table2) / length(net_data_table3)) * 1)
      ]
      net_data2 <- net_data2[net_data2$symbol %in% names(keep2), ]

      hub_mat <- as.data.frame.matrix(table(net_data2[ , c(3, 2)]))
      hub_mat <- hub_mat > 0
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )

      degree <- degree(graph)
      degree2 <- degree

      hub_mat <- as.data.frame.matrix(table(net_data2[ , c(3, 2)]))
      hub_mat <- hub_mat > 0
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )
      graph <- subgraph(graph, which(degree(graph) > 0))

      keep2 <- c()
      for (i in seq_along(V(graph))) {
        neighbors <- c(V(graph)[i], neighbors(graph, V(graph)[i]))
        net_data_table_sub <- net_data_table2[
          names(net_data_table2) %in% names(neighbors)
        ]
        keep2 <- c(
          keep2,
          names(net_data_table_sub)[
            which(net_data_table_sub == max(net_data_table_sub))
          ]
        )
      }
      keep2_table <- table(keep2)
      keep2_ref <- c()
      for (name in names(keep2_table)) {
        neighbors <- c(
          V(graph)$name[which(V(graph)$name == name)],
          names(neighbors(graph, V(graph)$name[which(V(graph)$name == name)]))
        )
        keep2_ref <- c(keep2_ref, length(neighbors))
      }
      degree_sub <- degree2[names(degree2) %in% names(keep2_table)]
      degree_sub <- (keep2_table / degree_sub)
      keep2_table <- (keep2_table / keep2_ref)
      net_data_table_sub <- net_data_table2[
        names(net_data_table2) %in% names(keep2_table)
      ]
      keep2_table <- net_data_table_sub * keep2_table * degree_sub
      keep2_table_unscaled <- keep2_table
      if (length(keep2_table) > 1) {
        keep2_table <- rank(keep2_table)
        keep2_table <- (
          (keep2_table - min(keep2_table)) /
            (max(keep2_table) - min(keep2_table))
        )
      } else {
        keep2_table <- 1
      }
      if (length(keep2_table) > 1) {
        keep2 <- names(keep2_table)[
          keep2_table > network_filt(input[[paste0("conn_up", 23)]])[[1]]
        ]
      } else {
        keep2 <- names(keep2_table)
      }

      keep3 <- net_data_table3_scaled[
        net_data_table3_scaled > (
          input[[paste0("conn_up", 23)]] *
            ((length(net_data_table2) / length(net_data_table3)) * 1)
        ) +
          1 - ((length(net_data_table2) / length(net_data_table3)) * 1)
      ]
      net_data3 <- net_data3[net_data3$symbol %in% names(keep3), ]

      hub_mat <- as.data.frame.matrix(table(net_data3[ , c(3, 2)]))
      hub_mat <- hub_mat > 0
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )

      degree <- degree(graph)
      degree3 <- degree

      hub_mat <- as.data.frame.matrix(table(net_data3[ , c(3, 2)]))
      hub_mat <- hub_mat > 0
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )
      graph <- subgraph(graph, which(degree(graph) > 0))

      keep3 <- c()
      for (i in seq_along(V(graph))) {
        neighbors <- c(V(graph)[i], neighbors(graph, V(graph)[i]))
        net_data_table_sub <- net_data_table3[
          names(net_data_table3) %in% names(neighbors)
        ]
        keep3 <- c(
          keep3,
          names(net_data_table_sub)[
            which(net_data_table_sub == max(net_data_table_sub))
          ]
        )
      }
      keep3_table <- table(keep3)
      keep3_ref <- c()
      for (name in names(keep3_table)) {
        neighbors <- c(
          V(graph)$name[which(V(graph)$name == name)],
          names(neighbors(graph, V(graph)$name[which(V(graph)$name == name)]))
        )
        keep3_ref <- c(keep3_ref, length(neighbors))
      }
      degree_sub <- degree3[names(degree3) %in% names(keep3_table)]
      degree_sub <- (keep3_table / degree_sub)
      keep3_table <- (keep3_table / keep3_ref)
      net_data_table_sub <- net_data_table3[
        names(net_data_table3) %in% names(keep3_table)
      ]
      keep3_table <- net_data_table_sub * keep3_table * degree_sub
      keep3_table_unscaled <- keep3_table
      if (length(keep3_table) > 1) {
        keep3_table <- rank(keep3_table)
        keep3_table <- (
          (keep3_table - min(keep3_table)) /
            (max(keep3_table) - min(keep3_table))
        )
      } else {
        keep3_table <- 1
      }
      if (length(keep3_table) > 1) {
        keep3 <- names(keep3_table)[
          keep3_table > network_filt(input[[paste0("conn_up", 23)]])[[2]]
        ]
      } else {
        keep3 <- names(keep3_table)
      }

      if (input[[paste0("shared_up", 23)]] == "Shared Features Only") {
        net_data <- net_data[
          net_data$symbol %in% net_data2_orig$symbol &
            net_data$symbol %in% net_data3_orig$symbol,
        ]
      } else {
        net_data <- net_data[
          net_data$symbol %in% net_data2$symbol |
            net_data$symbol %in% net_data3$symbol,
        ]
      }

      hub_mat <- as.data.frame.matrix(table(net_data[ , c(3, 2)]))
      hub_mat <- hub_mat > 0
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )

      keep_idx <- which(V(graph)$name %in% c(keep2, keep3))
      keep_degree <- unique(
        unlist(
          neighborhood(graph, order = map_range(inverse), nodes = keep_idx)
        )
      )
      keep_degree <- unique(c(keep_idx, keep_degree))
      graph <- induced_subgraph(graph, vids = keep_degree)

      if (input[[paste0("shared_up", 23)]] == "All Features") {
        graph <- subgraph(graph, which(degree(graph) > 0))
      }

      if (input[[paste0("comp_up", 23)]] == "Main Network Only") {
        graph <- subgraph(graph, which(components(graph)$membership == 1))
      }

      edges <- get.edgelist(graph, names = FALSE)
      layout <- qgraph.layout.fruchtermanreingold(
        edges, vcount = vcount(graph), weights = E(graph)$weight,
        area = vcount(graph), repulse.rad = vcount(graph) / 4
      )
      vis <- toVisNetworkData(graph)

      net_data_nodup <- net_data[-which(duplicated(net_data$symbol)), ]
      net_data_nodup <- net_data_nodup[
        net_data_nodup$symbol %in% vis$nodes$id,
      ]
      net_data_nodup <- net_data_nodup[
        match(vis$nodes$id, net_data_nodup$symbol),
      ]
      vis$nodes$shape <- ifelse(
        net_data_nodup$symbol %in% c(keep2, keep3), yes = "circle", no = "box"
      )

      keep2_table <- keep2_table[names(keep2_table) %in% keep2]
      keep3_table <- keep3_table[names(keep3_table) %in% keep3]
      keep_table_comb <- c(keep2_table, keep3_table)
      keep_table_comb <- keep_table_comb[
        names(keep_table_comb) %in% vis$nodes$id
      ]
      if (length(keep_table_comb) > 1 & all(keep_table_comb != 1)) {
        keep_table_comb <- (
          (keep_table_comb - min(keep_table_comb)) /
            (max(keep_table_comb) - min(keep_table_comb))
        )
      } else if (length(keep_table_comb) == 1) {
        keep_table_comb <- 1
      }
      size <- c()
      for (gene in vis$nodes$id) {
        if (gene %in% names(keep_table_comb)) {
          keep_table_sub <- keep_table_comb[names(keep_table_comb) %in% gene]
          if (length(keep_table_sub > 0)) {
            keep_table_sub <- sum(keep_table_sub)
          }
          keep_table_sub <- (((keep_table_sub + 2) ^ 6) / 8) + 22
          size <- c(size, keep_table_sub)
        } else {
          size <- c(size, 30)
        }
      }
      vis$nodes$size <- size
      vis$edges$width <- (vis$edges$weight + 0) ^ 1.25

      if (length(names(V(graph))) > 0) {
        col <- case_when(
          names(V(graph)) %in% net_data2_orig$symbol &
            names(V(graph)) %in% net_data3_orig$symbol ~ "green",
          names(V(graph)) %in% net_data2_orig$symbol ~ "yellow",
          names(V(graph)) %in% net_data3_orig$symbol ~ "orange"
        )
        vis$nodes$group <- col

        f <- pmax((max(nchar(vis$nodes$id)) - nchar(vis$nodes$id)) / 2, 0)
        vis$nodes$label <- sprintf(
          "%-*s%s%*s", f, "", vis$nodes$id, ceiling(f), ""
        )
        vis$nodes$font.size <- vis$nodes$size

        title <- vector(length = length(vis$nodes$id))
        for (j in seq_along(title)) {
          net_data_sub <- net_data[net_data$symbol == vis$nodes$id[j], ]
          title[j] <- paste0(
            net_data_sub$gene_set, " - ",
            net_data_sub$dataset, collapse = "<br>"
          )
        }
        vis$nodes$title <- title

        vis <- visNetwork(vis$nodes, vis$edges, background = "black")
        vis <- visPhysics(vis, stabilization = FALSE)
        vis <- visEdges(
          vis, smooth = FALSE, color = list(opacity = 0.8, inherit = "both")
        )
        vis <- visIgraphLayout(
          vis, layout = "layout.norm", layoutMatrix = layout
        )
        vis <- visOptions(
          vis,
          highlightNearest = list(
            enabled = TRUE,
            degree = ifelse(
              map_range(inverse) - 1 < 1, yes = 1, no = map_range(inverse) - 1
            )
          ),
          nodesIdSelection = TRUE
        )
        vis <- visNodes(vis, font = list(color = "white"))
        vis <- visGroups(
          vis,
          groupname = "green",
          color = list(
            background = "rgba(0, 0, 0, 0.5)",
            border = "rgba(0, 255, 0, 1)",
            highlight = list(
              background = "rgba(0, 255, 0, 0.5)",
              border = "rgba(0, 255, 0, 1)"
            )
          )
        )
        vis <- visGroups(
          vis,
          groupname = "yellow",
          color = list(
            background = "rgba(0, 0, 0, 0.5)",
            border = "rgba(255, 255, 0, 1)",
            highlight = list(
              background = "rgba(255, 255, 0, 0.5)",
              border = "rgba(255, 255, 0, 1)"
            )
          )
        )
        vis <- visGroups(
          vis,
          groupname = "orange",
          color = list(
            background = "rgba(0, 0, 0, 0.5)",
            border = "rgba(255, 165, 0, 1)",
            highlight = list(
              background = "rgba(255, 165, 0, 0.5)",
              border = "rgba(255, 165, 0, 1)"
            )
          )
        )
        vis <- visInteraction(
          vis,
          tooltipStyle =
            "width: max-content;position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #ffffff;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);opacity: 0.9;"
        )
        lnodes <- data.frame(
          label = c(
            "Upregulated\nGenes/Proteins\n\n\nLegend",
            "Differentially\nExpressed",
            "Network\nHub",
            "LCM\nMass Spec",
            "FACS\nssRNAseq",
            "Both\nDatasets"
          ),
          color.background = c(
            "black", "black", "black", "black", "black", "black"
          ),
          color.border = c(
            "rgba(255, 255, 255, 1)", "rgba(255, 255, 255, 1)",
            "rgba(255, 255, 255, 1)", "rgba(255, 255, 0, 1)",
            "rgba(255, 165, 0, 1)", "rgba(0, 255, 0, 1)"
          ),
          shape = c("text", "box", "circle", "ellipse", "ellipse", "ellipse"),
          font.color = "white",
          font.size = c(15, 12, 12, 12, 12, 12)
        )
        vis <- visLegend(
          vis, addNodes = lnodes, position = "right",
          useGroups = FALSE, stepY = 60, zoom = FALSE
        )
      }
      output$vis_up23 <- renderVisNetwork({vis})

      net_data <- net_data[net_data$symbol %in% c(keep2, keep3), ]
      vertices <- data.frame(
        name = c(unique(net_data$symbol), unique(net_data$gene_set)),
        type = c(
          rep("deg", length(unique(net_data$symbol))),
          rep("gse", length(unique(net_data$gene_set)))
        )
      )
      edges <- as.data.frame(as.matrix(net_data[ , c(3, 2)]))
      colnames(edges) <- c("from", "to")

      graph <- graph_from_data_frame(
        edges, vertices = vertices, directed = FALSE
      )

      keep_idx <- which(V(graph)$name %in% c(keep2, keep3))
      keep_degree <- unique(
        unlist(
          neighborhood(graph, order = map_range(inverse), nodes = keep_idx)
        )
      )
      keep_degree <- unique(c(keep_idx, keep_degree))
      graph <- induced_subgraph(graph, vids = keep_degree)

      graph <- subgraph(graph, which(degree(graph) > 0))

      graph_sub <- V(graph)[which(V(graph)$type == "gse")]
      degree <- degree(graph, v = graph_sub)

      pval2 <- res_gse[[2]]
      pval2 <- pval2[pval2$GeneSet %in% names(degree), ]
      degree_sub <- degree[names(degree) %in% pval2$GeneSet]
      pval2 <- pval2[match(names(degree_sub), pval2$GeneSet), ]
      pval2 <- pval2$Coef
      names(pval2) <- names(degree_sub)

      pval3 <- res_gse[[3]]
      pval3 <- pval3[pval3$GeneSet %in% names(degree), ]
      degree_sub <- degree[names(degree) %in% pval3$GeneSet]
      pval3 <- pval3[match(names(degree_sub), pval3$GeneSet), ]
      pval3 <- pval3$Coef
      names(pval3) <- names(degree_sub)

      pval <- c(pval2, pval3)
      pval <- tapply(pval, INDEX = names(pval), FUN = sum)
      pval <- pval[match(names(degree), names(pval))]
      pval <- (pval - min(pval)) / (max(pval) - min(pval))
      pval <- rank(pval)
      pval <- ((pval - min(pval)) / (max(pval) - min(pval))) / 100000000000

      degree <- degree + pval
      degree <- (degree - min(degree)) / (max(degree) - min(degree))
      degree <- rank(degree)
      degree <- (degree - min(degree)) / (max(degree) - min(degree))

      graph_sub <- graph_sub[degree >= input[[paste0("degree_up", 23)]]]
      graph <- induced_subgraph(
        graph, vids = unlist(neighborhood(graph, order = 1, nodes = graph_sub))
      )

      edge_counts <- table(
        apply(
          get.edgelist(graph), MARGIN = 1,
          FUN = function(x) paste(sort(x), collapse = "-")
        )
      )
      E(graph)$weight <- edge_counts[
        apply(
          get.edgelist(graph), MARGIN = 1,
          FUN = function(x) paste(sort(x), collapse = "-")
        )
      ]
      edges <- get.edgelist(graph, names = FALSE)
      layout <- qgraph.layout.fruchtermanreingold(
        edges, vcount = vcount(graph), weights = E(graph)$weight,
        area = vcount(graph), repulse.rad = vcount(graph) / 4
      )
      vis2 <- toVisNetworkData(graph)

      vis2$nodes$shape <- ifelse(
        vis2$nodes$id %in% c(keep2, keep3), yes = "circle", no = "box"
      )

      keep2_table <- keep2_table[names(keep2_table) %in% keep2]
      keep3_table <- keep3_table[names(keep3_table) %in% keep3]
      keep_table_comb <- c(keep2_table, keep3_table)
      keep_table_comb <- keep_table_comb[
        names(keep_table_comb) %in% vis2$nodes$id
      ]
      if (length(keep_table_comb) > 1 & all(keep_table_comb != 1)) {
        keep_table_comb <- (
          (keep_table_comb - min(keep_table_comb)) /
            (max(keep_table_comb) - min(keep_table_comb))
        )
      } else if (length(keep_table_comb) == 1) {
        keep_table_comb <- 1
      }
      size <- c()
      for (gene in vis2$nodes$id) {
        if (gene %in% names(keep_table_comb)) {
          keep_table_sub <- keep_table_comb[names(keep_table_comb) %in% gene]
          if (length(keep_table_sub > 0)) {
            keep_table_sub <- sum(keep_table_sub)
          }
          keep_table_sub <- (((keep_table_sub + 2) ^ 6) / 16) + 11
          size <- c(size, keep_table_sub)
        } else {
          size <- c(size, 35)
        }
      }
      vis2$nodes$size <- size
      vis2$edges$width <- (vis2$edges$weight + 0.5) ^ 3.5

      if (length(names(V(graph))) > 0) {
        col <- case_when(
          names(V(graph)) %in% net_data2_orig$symbol &
            names(V(graph)) %in% net_data3_orig$symbol ~ "green",
          names(V(graph)) %in% net_data2_orig$symbol ~ "yellow",
          names(V(graph)) %in% net_data3_orig$symbol ~ "orange",
          names(V(graph)) %in% net_data2_orig$gene_set &
            names(V(graph)) %in% net_data3_orig$gene_set ~ "green",
          names(V(graph)) %in% net_data2_orig$gene_set ~ "yellow",
          names(V(graph)) %in% net_data3_orig$gene_set ~ "orange"
        )
        vis2$nodes$group <- col

        deg_string <- vis2$nodes$id[vis2$nodes$id %in% net_data$symbol]
        f <- pmax((max(nchar(net_data$symbol)) - nchar(deg_string)) / 2, 0)
        deg_string <- sprintf(
          "%-*s%s%*s", f, "", deg_string, ceiling(f), ""
        )
        gse_string <- vis2$nodes$id[vis2$nodes$id %in% net_data$gene_set]
        gse_string <- sapply(gse_string, FUN = add_newline)
        vis2$nodes$label <- c(deg_string, gse_string)
        vis2$nodes$font.size <- vis2$nodes$size

        vis2 <- visNetwork(vis2$nodes, vis2$edges, background = "black")
        vis2 <- visPhysics(vis2, stabilization = FALSE)
        vis2 <- visEdges(
          vis2, smooth = FALSE, color = list(opacity = 0.8, inherit = "both")
        )
        vis2 <- visIgraphLayout(
          vis2, layout = "layout.norm", layoutMatrix = layout
        )
        vis2 <- visOptions(
          vis2,
          highlightNearest = list(
            enabled = TRUE,
            degree = ifelse(
              map_range(inverse) - 1 < 1, yes = 1, no = map_range(inverse) - 1
            )
          ),
          nodesIdSelection = TRUE
        )
        vis2 <- visNodes(vis2, font = list(color = "white"))
        vis2 <- visGroups(
          vis2,
          groupname = "green",
          color = list(
            background = "rgba(0, 0, 0, 0.5)",
            border = "rgba(0, 255, 0, 1)",
            highlight = list(
              background = "rgba(0, 255, 0, 0.5)",
              border = "rgba(0, 255, 0, 1)"
            )
          )
        )
        vis2 <- visGroups(
          vis2,
          groupname = "yellow",
          color = list(
            background = "rgba(0, 0, 0, 0.5)",
            border = "rgba(255, 255, 0, 1)",
            highlight = list(
              background = "rgba(255, 255, 0, 0.5)",
              border = "rgba(255, 255, 0, 1)"
            )
          )
        )
        vis2 <- visGroups(
          vis2,
          groupname = "orange",
          color = list(
            background = "rgba(0, 0, 0, 0.5)",
            border = "rgba(255, 165, 0, 1)",
            highlight = list(
              background = "rgba(255, 165, 0, 0.5)",
              border = "rgba(255, 165, 0, 1)"
            )
          )
        )
        lnodes <- data.frame(
          label = c(
            "Upregulated\nPathways\n\n\nLegend",
            "Differentially\nEnriched",
            "Network\nHub",
            "LCM\nMass Spec",
            "FACS\nssRNAseq",
            "Both\nDatasets"
          ),
          color.background = c(
            "black", "black", "black", "black", "black", "black"
          ),
          color.border = c(
            "rgba(255, 255, 255, 1)", "rgba(255, 255, 255, 1)",
            "rgba(255, 255, 255, 1)", "rgba(255, 255, 0, 1)",
            "rgba(255, 165, 0, 1)", "rgba(0, 255, 0, 1)"
          ),
          shape = c("text", "box", "circle", "ellipse", "ellipse", "ellipse"),
          font.color = "white",
          font.size = c(15, 12, 12, 12, 12, 12)
        )
        vis2 <- visLegend(
          vis2, addNodes = lnodes, position = "right",
          useGroups = FALSE, stepY = 60, zoom = FALSE
        )
      }
      output$vis2_up23 <- renderVisNetwork({vis2})

    })
  }
)
```

### Downregulated in Tangle-bearing Neurons

```{r}
shinyApp(

  ui = fluidPage(
    fluidRow(
      column(
        width = 2,
        sliderInput(
          paste0("conn_dn", 23),
          label = "Stringency:",
          min = 0,
          max = 1,
          value = 0.1,
          step = 0.1
        ),
      ),
      column(
        width = 2,
        radioButtons(
          paste0("shared_dn", 23), label = NULL,
          c("All Features", "Shared Features Only"), inline = TRUE
        ),
        style = "padding-top: 35px;"
      ),
      column(
        width = 2,
        radioButtons(
          paste0("comp_dn", 23), label = NULL,
          c("All Networks", "Main Network Only"), inline = TRUE
        ),
        style = "padding-top: 35px;"
      ),
      column(
        width = 2,
        sliderInput(
          paste0("degree_dn", 23),
          label = "Connectivity:",
          min = 0,
          max = 1,
          value = 0.1,
          step = 0.1
        ),
      )
    ),
    fluidPage(
      div(
        style = "display:flex;",
        visNetworkOutput(paste0("vis_dn", 23), height = "650px", width = "50%"),
        visNetworkOutput(paste0("vis2_dn", 23), height = "650px", width = "50%")
      )
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("conn_dn", 23)]]
      input[[paste0("shared_dn", 23)]]
      input[[paste0("comp_dn", 23)]]
      input[[paste0("degree_dn", 23)]]
    }, {

      inverse <- round(1 - input[[paste0("conn_dn", 23)]], digits = 2)

      de_gse <- rownames(res_gse[[2]][res_gse[[2]]$P.value.adj < 0.05, ])
      up_gse <- rownames(res_gse[[2]][res_gse[[2]]$Coef > 0, ])
      de_deg <- rownames(res_deg[[2]][res_deg[[2]]$P.value.adj < 0.05, ])
      up_deg <- rownames(res_deg[[2]][res_deg[[2]]$Coef > 0, ])

      net_data <- melt(geneIds(gene_sets[names(gene_sets) %in% de_gse]))
      names(net_data)[1] <- "ensembl"
      names(net_data)[2] <- "gene_set"

      gene_anno_sub <- gene_anno[gene_anno$ensembl %in% net_data$ensembl, ]
      net_data <- left_join(net_data, gene_anno_sub, by = "ensembl")
      net_data$gse_direction <- ifelse(
        net_data$gene_set %in% up_gse, yes = "up", no = "down"
      )
      net_data$gene_DE <- ifelse(
        net_data$ensembl %in% de_deg, yes = "yes", no = "no"
      )
      net_data$gene_direction <- ifelse(
        net_data$ensembl %in% up_deg, yes = "up", no = "down"
      )
      net_data$dataset <- "LCM Mass Spec"

      remove <- which(net_data$gse_direction != net_data$gene_direction)
      net_data <- net_data[-remove, ]
      net_data <- net_data[net_data$gene_DE == "yes", ]

      net_data <- net_data[which(net_data$gse_direction == "down"), ]
      net_data2 <- net_data
      net_data2_orig <- net_data2

      de_gse <- rownames(res_gse[[3]][res_gse[[3]]$P.value.adj < 0.05, ])
      up_gse <- rownames(res_gse[[3]][res_gse[[3]]$Coef > 0, ])
      de_deg <- rownames(res_deg[[3]][res_deg[[3]]$P.value.adj < 0.05, ])
      up_deg <- rownames(res_deg[[3]][res_deg[[3]]$Coef > 0, ])

      net_data <- melt(geneIds(gene_sets[names(gene_sets) %in% de_gse]))
      names(net_data)[1] <- "ensembl"
      names(net_data)[2] <- "gene_set"

      gene_anno_sub <- gene_anno[gene_anno$ensembl %in% net_data$ensembl, ]
      net_data <- left_join(net_data, gene_anno_sub, by = "ensembl")
      net_data$gse_direction <- ifelse(
        net_data$gene_set %in% up_gse, yes = "up", no = "down"
      )
      net_data$gene_DE <- ifelse(
        net_data$ensembl %in% de_deg, yes = "yes", no = "no"
      )
      net_data$gene_direction <- ifelse(
        net_data$ensembl %in% up_deg, yes = "up", no = "down"
      )
      net_data$dataset <- "FACS ssRNAseq"

      remove <- which(net_data$gse_direction != net_data$gene_direction)
      net_data <- net_data[-remove, ]
      net_data <- net_data[net_data$gene_DE == "yes", ]

      net_data <- net_data[which(net_data$gse_direction == "down"), ]
      net_data3 <- net_data
      net_data3_orig <- net_data3

      net_data <- rbind(net_data2, net_data3)
      net_data_table2 <- table(net_data2$symbol)
      net_data_table3 <- table(net_data3$symbol)

      gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(gse[[2]])]
      gene_sets_table2 <- table(unlist(geneIds(gene_sets_sub)))
      gene_sets_sub <- gene_sets[names(gene_sets) %in% rownames(gse[[3]])]
      gene_sets_table3 <- table(unlist(geneIds(gene_sets_sub)))

      gene_anno_sub <- gene_anno[
        gene_anno$ensembl %in% names(gene_sets_table2),
      ]
      gene_anno_sub <- gene_anno_sub[
        match(names(gene_sets_table2), gene_anno_sub$ensembl),
      ]
      names(gene_sets_table2) <- gene_anno_sub$symbol

      gene_anno_sub <- gene_anno[
        gene_anno$ensembl %in% names(gene_sets_table3),
      ]
      gene_anno_sub <- gene_anno_sub[
        match(names(gene_sets_table3), gene_anno_sub$ensembl),
      ]
      names(gene_sets_table3) <- gene_anno_sub$symbol

      gene_sets_table2 <- gene_sets_table2[
        names(gene_sets_table2) %in% names(net_data_table2)
      ]
      gene_sets_table3 <- gene_sets_table3[
        names(gene_sets_table3) %in% names(net_data_table3)
      ]
      gene_sets_table2 <- gene_sets_table2[
        match(names(net_data_table2), names(gene_sets_table2))
      ]
      gene_sets_table3 <- gene_sets_table3[
        match(names(net_data_table3), names(gene_sets_table3))
      ]

      pval2 <- res_deg[[2]]
      pval2 <- pval2[pval2$Gene %in% names(net_data_table2), ]
      pval2 <- pval2[match(names(net_data_table2), pval2$Gene), ]
      pval2 <- pval2$Coef
      pval2 <- (pval2 - min(pval2)) / (max(pval2) - min(pval2))
      pval2 <- rank(pval2)
      pval2 <- ((pval2 - min(pval2)) / (max(pval2) - min(pval2))) / 100000000000
      names(pval2) <- names(net_data_table2)

      pval3 <- res_deg[[3]]
      pval3 <- pval3[pval3$Gene %in% names(net_data_table3), ]
      pval3 <- pval3[match(names(net_data_table3), pval3$Gene), ]
      pval3 <- pval3$Coef
      pval3 <- (pval3 - min(pval3)) / (max(pval3) - min(pval3))
      pval3 <- rank(pval3)
      pval3 <- ((pval3 - min(pval3)) / (max(pval3) - min(pval3))) / 100000000000
      names(pval3) <- names(net_data_table3)

      dup2 <- c()
      for (gene in unique(net_data2$symbol)) {
        matched_gse <- net_data2[net_data2$symbol %in% gene, ]$gene_set
        net_data_sub <- net_data2[net_data2$gene_set %in% matched_gse, ]
        net_data_sub <- net_data_sub[net_data_sub$symbol %notin% gene, ]
        dup_table <- table(net_data_sub$symbol)
        if (length(dup_table) > 0) {
          dup2 <- c(dup2, mean(dup_table))
        } else {
          dup2 <- c(dup2, 1)
        }
      }
      names(dup2) <- unique(net_data2$symbol)
      dup2 <- dup2[match(names(net_data_table2), names(dup2))]

      dup3 <- c()
      set_size3 <- c()
      for (gene in unique(net_data3$symbol)) {
        matched_gse <- net_data3[net_data3$symbol %in% gene, ]$gene_set
        net_data_sub <- net_data3[net_data3$gene_set %in% matched_gse, ]
        net_data_sub <- net_data_sub[net_data_sub$symbol %notin% gene, ]
        dup_table <- table(net_data_sub$symbol)
        if (length(dup_table) > 0) {
          dup3 <- c(dup3, mean(dup_table))
        } else {
          dup3 <- c(dup3, 1)
        }
      }
      names(dup3) <- unique(net_data3$symbol)
      dup3 <- dup3[match(names(net_data_table3), names(dup3))]

      net_data_table2 <- (
        ((net_data_table2 / gene_sets_table2) * net_data_table2) / dup2
      ) + pval2
      net_data_table3 <- (
        ((net_data_table3 / gene_sets_table3) * net_data_table3) / dup3
      ) + pval3
      net_data_table2_scaled <- net_data_table2
      net_data_table3_scaled <- net_data_table3
      net_data_table2_scaled <- rank(net_data_table2_scaled)
      net_data_table3_scaled <- rank(net_data_table3_scaled)
      net_data_table2_scaled <- (
        (net_data_table2_scaled - min(net_data_table2_scaled)) /
          (max(net_data_table2_scaled) - min(net_data_table2_scaled))
      )
      net_data_table3_scaled <- (
        (net_data_table3_scaled - min(net_data_table3_scaled)) /
          (max(net_data_table3_scaled) - min(net_data_table3_scaled))
      )

      keep2 <- net_data_table2_scaled[
        net_data_table2_scaled > (
          inverse * ((length(net_data_table2) / length(net_data_table3)) * 1)
        ) + ((length(net_data_table2) / length(net_data_table3)) * 1)
      ]
      net_data2 <- net_data2[net_data2$symbol %in% names(keep2), ]

      hub_mat <- as.data.frame.matrix(table(net_data2[ , c(3, 2)]))
      hub_mat <- hub_mat > 0
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )

      degree <- degree(graph)
      degree2 <- degree

      hub_mat <- as.data.frame.matrix(table(net_data2[ , c(3, 2)]))
      hub_mat <- hub_mat > 0
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )
      graph <- subgraph(graph, which(degree(graph) > 0))

      keep2 <- c()
      for (i in seq_along(V(graph))) {
        neighbors <- c(V(graph)[i], neighbors(graph, V(graph)[i]))
        net_data_table_sub <- net_data_table2[
          names(net_data_table2) %in% names(neighbors)
        ]
        keep2 <- c(
          keep2,
          names(net_data_table_sub)[
            which(net_data_table_sub == max(net_data_table_sub))
          ]
        )
      }
      keep2_table <- table(keep2)
      keep2_ref <- c()
      for (name in names(keep2_table)) {
        neighbors <- c(
          V(graph)$name[which(V(graph)$name == name)],
          names(neighbors(graph, V(graph)$name[which(V(graph)$name == name)]))
        )
        keep2_ref <- c(keep2_ref, length(neighbors))
      }
      degree_sub <- degree2[names(degree2) %in% names(keep2_table)]
      degree_sub <- (keep2_table / degree_sub)
      keep2_table <- (keep2_table / keep2_ref)
      net_data_table_sub <- net_data_table2[
        names(net_data_table2) %in% names(keep2_table)
      ]
      keep2_table <- net_data_table_sub * keep2_table * degree_sub
      keep2_table_unscaled <- keep2_table
      if (length(keep2_table) > 1) {
        keep2_table <- rank(keep2_table)
        keep2_table <- (
          (keep2_table - min(keep2_table)) /
            (max(keep2_table) - min(keep2_table))
        )
      } else {
        keep2_table <- 1
      }
      if (length(keep2_table) > 1) {
        keep2 <- names(keep2_table)[
          keep2_table > network_filt(input[[paste0("conn_dn", 23)]])[[1]]
        ]
      } else {
        keep2 <- names(keep2_table)
      }

      keep3 <- net_data_table3_scaled[
        net_data_table3_scaled > (
          input[[paste0("conn_dn", 23)]] *
            ((length(net_data_table2) / length(net_data_table3)) * 1)
        ) +
          1 - ((length(net_data_table2) / length(net_data_table3)) * 1)
      ]
      net_data3 <- net_data3[net_data3$symbol %in% names(keep3), ]

      hub_mat <- as.data.frame.matrix(table(net_data3[ , c(3, 2)]))
      hub_mat <- hub_mat > 0
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )

      degree <- degree(graph)
      degree3 <- degree

      hub_mat <- as.data.frame.matrix(table(net_data3[ , c(3, 2)]))
      hub_mat <- hub_mat > 0
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )
      graph <- subgraph(graph, which(degree(graph) > 0))

      keep3 <- c()
      for (i in seq_along(V(graph))) {
        neighbors <- c(V(graph)[i], neighbors(graph, V(graph)[i]))
        net_data_table_sub <- net_data_table3[
          names(net_data_table3) %in% names(neighbors)
        ]
        keep3 <- c(
          keep3,
          names(net_data_table_sub)[
            which(net_data_table_sub == max(net_data_table_sub))
          ]
        )
      }
      keep3_table <- table(keep3)
      keep3_ref <- c()
      for (name in names(keep3_table)) {
        neighbors <- c(
          V(graph)$name[which(V(graph)$name == name)],
          names(neighbors(graph, V(graph)$name[which(V(graph)$name == name)]))
        )
        keep3_ref <- c(keep3_ref, length(neighbors))
      }
      degree_sub <- degree3[names(degree3) %in% names(keep3_table)]
      degree_sub <- (keep3_table / degree_sub)
      keep3_table <- (keep3_table / keep3_ref)
      net_data_table_sub <- net_data_table3[
        names(net_data_table3) %in% names(keep3_table)
      ]
      keep3_table <- net_data_table_sub * keep3_table * degree_sub
      keep3_table_unscaled <- keep3_table
      if (length(keep3_table) > 1) {
        keep3_table <- rank(keep3_table)
        keep3_table <- (
          (keep3_table - min(keep3_table)) /
            (max(keep3_table) - min(keep3_table))
        )
      } else {
        keep3_table <- 1
      }
      if (length(keep3_table) > 1) {
        keep3 <- names(keep3_table)[
          keep3_table > network_filt(input[[paste0("conn_dn", 23)]])[[2]]
        ]
      } else {
        keep3 <- names(keep3_table)
      }

      if (input[[paste0("shared_dn", 23)]] == "Shared Features Only") {
        net_data <- net_data[
          net_data$symbol %in% net_data2_orig$symbol &
            net_data$symbol %in% net_data3_orig$symbol,
        ]
      } else {
        net_data <- net_data[
          net_data$symbol %in% net_data2$symbol |
            net_data$symbol %in% net_data3$symbol,
        ]
      }

      hub_mat <- as.data.frame.matrix(table(net_data[ , c(3, 2)]))
      hub_mat <- hub_mat > 0
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )

      keep_idx <- which(V(graph)$name %in% c(keep2, keep3))
      keep_degree <- unique(
        unlist(
          neighborhood(graph, order = map_range(inverse), nodes = keep_idx)
        )
      )
      keep_degree <- unique(c(keep_idx, keep_degree))
      graph <- induced_subgraph(graph, vids = keep_degree)

      if (input[[paste0("shared_dn", 23)]] == "All Features") {
        graph <- subgraph(graph, which(degree(graph) > 0))
      }

      if (input[[paste0("comp_dn", 23)]] == "Main Network Only") {
        graph <- subgraph(graph, which(components(graph)$membership == 1))
      }

      edges <- get.edgelist(graph, names = FALSE)
      layout <- qgraph.layout.fruchtermanreingold(
        edges, vcount = vcount(graph), weights = E(graph)$weight,
        area = vcount(graph), repulse.rad = vcount(graph) / 4
      )
      vis <- toVisNetworkData(graph)

      net_data_nodup <- net_data[-which(duplicated(net_data$symbol)), ]
      net_data_nodup <- net_data_nodup[
        net_data_nodup$symbol %in% vis$nodes$id,
      ]
      net_data_nodup <- net_data_nodup[
        match(vis$nodes$id, net_data_nodup$symbol),
      ]
      vis$nodes$shape <- ifelse(
        net_data_nodup$symbol %in% c(keep2, keep3), yes = "circle", no = "box"
      )

      keep2_table <- keep2_table[names(keep2_table) %in% keep2]
      keep3_table <- keep3_table[names(keep3_table) %in% keep3]
      keep_table_comb <- c(keep2_table, keep3_table)
      keep_table_comb <- keep_table_comb[
        names(keep_table_comb) %in% vis$nodes$id
      ]
      if (length(keep_table_comb) > 1 & all(keep_table_comb != 1)) {
        keep_table_comb <- (
          (keep_table_comb - min(keep_table_comb)) /
            (max(keep_table_comb) - min(keep_table_comb))
        )
      } else if (length(keep_table_comb) == 1) {
        keep_table_comb <- 1
      }
      size <- c()
      for (gene in vis$nodes$id) {
        if (gene %in% names(keep_table_comb)) {
          keep_table_sub <- keep_table_comb[names(keep_table_comb) %in% gene]
          if (length(keep_table_sub > 0)) {
            keep_table_sub <- sum(keep_table_sub)
          }
          keep_table_sub <- (((keep_table_sub + 2) ^ 5) / 4) + 22
          size <- c(size, keep_table_sub)
        } else {
          size <- c(size, 30)
        }
      }
      vis$nodes$size <- size
      vis$edges$width <- (vis$edges$weight + 1) ^ 1.25

      if (length(names(V(graph))) > 0) {
        col <- case_when(
          names(V(graph)) %in% net_data2_orig$symbol &
            names(V(graph)) %in% net_data3_orig$symbol ~ "green",
          names(V(graph)) %in% net_data2_orig$symbol ~ "yellow",
          names(V(graph)) %in% net_data3_orig$symbol ~ "orange"
        )
        vis$nodes$group <- col

        f <- pmax((max(nchar(vis$nodes$id)) - nchar(vis$nodes$id)) / 2, 0)
        vis$nodes$label <- sprintf(
          "%-*s%s%*s", f, "", vis$nodes$id, ceiling(f), ""
        )
        vis$nodes$font.size <- vis$nodes$size

        title <- vector(length = length(vis$nodes$id))
        for (j in seq_along(title)) {
          net_data_sub <- net_data[net_data$symbol == vis$nodes$id[j], ]
          title[j] <- paste0(
            net_data_sub$gene_set, " - ",
            net_data_sub$dataset, collapse = "<br>"
          )
        }
        vis$nodes$title <- title

        vis <- visNetwork(vis$nodes, vis$edges, background = "black")
        vis <- visPhysics(vis, stabilization = FALSE)
        vis <- visEdges(
          vis, smooth = FALSE, color = list(opacity = 0.8, inherit = "both")
        )
        vis <- visIgraphLayout(
          vis, layout = "layout.norm", layoutMatrix = layout
        )
        vis <- visOptions(
          vis,
          highlightNearest = list(
            enabled = TRUE,
            degree = ifelse(
              map_range(inverse) - 1 < 1, yes = 1, no = map_range(inverse) - 1
            )
          ),
          nodesIdSelection = TRUE
        )
        vis <- visNodes(vis, font = list(color = "white"))
        vis <- visGroups(
          vis,
          groupname = "green",
          color = list(
            background = "rgba(0, 0, 0, 0.5)",
            border = "rgba(0, 255, 0, 1)",
            highlight = list(
              background = "rgba(0, 255, 0, 0.5)",
              border = "rgba(0, 255, 0, 1)"
            )
          )
        )
        vis <- visGroups(
          vis,
          groupname = "yellow",
          color = list(
            background = "rgba(0, 0, 0, 0.5)",
            border = "rgba(255, 255, 0, 1)",
            highlight = list(
              background = "rgba(255, 255, 0, 0.5)",
              border = "rgba(255, 255, 0, 1)"
            )
          )
        )
        vis <- visGroups(
          vis,
          groupname = "orange",
          color = list(
            background = "rgba(0, 0, 0, 0.5)",
            border = "rgba(255, 165, 0, 1)",
            highlight = list(
              background = "rgba(255, 165, 0, 0.5)",
              border = "rgba(255, 165, 0, 1)"
            )
          )
        )
        vis <- visInteraction(
          vis,
          tooltipStyle =
            "width: max-content;position: fixed;visibility:hidden;padding: 5px;white-space: nowrap;font-family: verdana;font-size:14px;font-color:#000000;background-color: #ffffff;-moz-border-radius: 3px;-webkit-border-radius: 3px;border-radius: 3px;border: 1px solid #808074;box-shadow: 3px 3px 10px rgba(0, 0, 0, 0.2);opacity: 0.9;"
        )
        lnodes <- data.frame(
          label = c(
            "Downregulated\nGenes/Proteins\n\n\nLegend",
            "Differentially\nExpressed",
            "Network\nHub",
            "LCM\nMass Spec",
            "FACS\nssRNAseq",
            "Both\nDatasets"
          ),
          color.background = c(
            "black", "black", "black", "black", "black", "black"
          ),
          color.border = c(
            "rgba(255, 255, 255, 1)", "rgba(255, 255, 255, 1)",
            "rgba(255, 255, 255, 1)", "rgba(255, 255, 0, 1)",
            "rgba(255, 165, 0, 1)", "rgba(0, 255, 0, 1)"
          ),
          shape = c("text", "box", "circle", "ellipse", "ellipse", "ellipse"),
          font.color = "white",
          font.size = c(15, 12, 12, 12, 12, 12)
        )
        vis <- visLegend(
          vis, addNodes = lnodes, position = "right",
          useGroups = FALSE, stepY = 60, zoom = FALSE
        )
      }
      output$vis_dn23 <- renderVisNetwork({vis})

      net_data <- net_data[net_data$symbol %in% c(keep2, keep3), ]
      vertices <- data.frame(
        name = c(unique(net_data$symbol), unique(net_data$gene_set)),
        type = c(
          rep("deg", length(unique(net_data$symbol))),
          rep("gse", length(unique(net_data$gene_set)))
        )
      )
      edges <- as.data.frame(as.matrix(net_data[ , c(3, 2)]))
      colnames(edges) <- c("from", "to")

      graph <- graph_from_data_frame(
        edges, vertices = vertices, directed = FALSE
      )

      keep_idx <- which(V(graph)$name %in% c(keep2, keep3))
      keep_degree <- unique(
        unlist(
          neighborhood(graph, order = map_range(inverse), nodes = keep_idx)
        )
      )
      keep_degree <- unique(c(keep_idx, keep_degree))
      graph <- induced_subgraph(graph, vids = keep_degree)

      graph <- subgraph(graph, which(degree(graph) > 0))

      graph_sub <- V(graph)[which(V(graph)$type == "gse")]
      degree <- degree(graph, v = graph_sub)

      pval2 <- res_gse[[2]]
      pval2 <- pval2[pval2$GeneSet %in% names(degree), ]
      degree_sub <- degree[names(degree) %in% pval2$GeneSet]
      pval2 <- pval2[match(names(degree_sub), pval2$GeneSet), ]
      pval2 <- pval2$Coef
      names(pval2) <- names(degree_sub)

      pval3 <- res_gse[[3]]
      pval3 <- pval3[pval3$GeneSet %in% names(degree), ]
      degree_sub <- degree[names(degree) %in% pval3$GeneSet]
      pval3 <- pval3[match(names(degree_sub), pval3$GeneSet), ]
      pval3 <- pval3$Coef
      names(pval3) <- names(degree_sub)

      pval <- c(pval2, pval3)
      pval <- tapply(pval, INDEX = names(pval), FUN = sum)
      pval <- pval[match(names(degree), names(pval))]
      pval <- (pval - min(pval)) / (max(pval) - min(pval))
      pval <- rank(pval)
      pval <- ((pval - min(pval)) / (max(pval) - min(pval))) / 100000000000

      degree <- degree + pval
      degree <- (degree - min(degree)) / (max(degree) - min(degree))
      degree <- rank(degree)
      degree <- (degree - min(degree)) / (max(degree) - min(degree))

      graph_sub <- graph_sub[degree >= input[[paste0("degree_dn", 23)]]]
      graph <- induced_subgraph(
        graph, vids = unlist(neighborhood(graph, order = 1, nodes = graph_sub))
      )

      edge_counts <- table(
        apply(
          get.edgelist(graph), MARGIN = 1,
          FUN = function(x) paste(sort(x), collapse = "-")
        )
      )
      E(graph)$weight <- edge_counts[
        apply(
          get.edgelist(graph), MARGIN = 1,
          FUN = function(x) paste(sort(x), collapse = "-")
        )
      ]
      edges <- get.edgelist(graph, names = FALSE)
      layout <- qgraph.layout.fruchtermanreingold(
        edges, vcount = vcount(graph), weights = E(graph)$weight,
        area = vcount(graph), repulse.rad = vcount(graph) / 4
      )
      vis2 <- toVisNetworkData(graph)

      vis2$nodes$shape <- ifelse(
        vis2$nodes$id %in% c(keep2, keep3), yes = "circle", no = "box"
      )

      keep2_table <- keep2_table[names(keep2_table) %in% keep2]
      keep3_table <- keep3_table[names(keep3_table) %in% keep3]
      keep_table_comb <- c(keep2_table, keep3_table)
      keep_table_comb <- keep_table_comb[
        names(keep_table_comb) %in% vis2$nodes$id
      ]
      if (length(keep_table_comb) > 1 & all(keep_table_comb != 1)) {
        keep_table_comb <- (
          (keep_table_comb - min(keep_table_comb)) /
            (max(keep_table_comb) - min(keep_table_comb))
        )
      } else if (length(keep_table_comb) == 1) {
        keep_table_comb <- 1
      }
      size <- c()
      for (gene in vis2$nodes$id) {
        if (gene %in% names(keep_table_comb)) {
          keep_table_sub <- keep_table_comb[names(keep_table_comb) %in% gene]
          if (length(keep_table_sub > 0)) {
            keep_table_sub <- sum(keep_table_sub)
          }
          keep_table_sub <- (((keep_table_sub + 2) ^ 6) / 16) + 11
          size <- c(size, keep_table_sub)
        } else {
          size <- c(size, 35)
        }
      }
      vis2$nodes$size <- size
      vis2$edges$width <- (vis2$edges$weight + 0.5) ^ 2

      if (length(names(V(graph))) > 0) {
        col <- case_when(
          names(V(graph)) %in% net_data2_orig$symbol &
            names(V(graph)) %in% net_data3_orig$symbol ~ "green",
          names(V(graph)) %in% net_data2_orig$symbol ~ "yellow",
          names(V(graph)) %in% net_data3_orig$symbol ~ "orange",
          names(V(graph)) %in% net_data2_orig$gene_set &
            names(V(graph)) %in% net_data3_orig$gene_set ~ "green",
          names(V(graph)) %in% net_data2_orig$gene_set ~ "yellow",
          names(V(graph)) %in% net_data3_orig$gene_set ~ "orange"
        )
        vis2$nodes$group <- col

        deg_string <- vis2$nodes$id[vis2$nodes$id %in% net_data$symbol]
        f <- pmax((max(nchar(net_data$symbol)) - nchar(deg_string)) / 2, 0)
        deg_string <- sprintf(
          "%-*s%s%*s", f, "", deg_string, ceiling(f), ""
        )
        gse_string <- vis2$nodes$id[vis2$nodes$id %in% net_data$gene_set]
        gse_string <- sapply(gse_string, FUN = add_newline)
        vis2$nodes$label <- c(deg_string, gse_string)
        vis2$nodes$font.size <- vis2$nodes$size

        vis2 <- visNetwork(vis2$nodes, vis2$edges, background = "black")
        vis2 <- visPhysics(vis2, stabilization = FALSE)
        vis2 <- visEdges(
          vis2, smooth = FALSE, color = list(opacity = 0.8, inherit = "both")
        )
        vis2 <- visIgraphLayout(
          vis2, layout = "layout.norm", layoutMatrix = layout
        )
        vis2 <- visOptions(
          vis2,
          highlightNearest = list(
            enabled = TRUE,
            degree = ifelse(
              map_range(inverse) - 1 < 1, yes = 1, no = map_range(inverse) - 1
            )
          ),
          nodesIdSelection = TRUE
        )
        vis2 <- visNodes(vis2, font = list(color = "white"))
        vis2 <- visGroups(
          vis2,
          groupname = "green",
          color = list(
            background = "rgba(0, 0, 0, 0.5)",
            border = "rgba(0, 255, 0, 1)",
            highlight = list(
              background = "rgba(0, 255, 0, 0.5)",
              border = "rgba(0, 255, 0, 1)"
            )
          )
        )
        vis2 <- visGroups(
          vis2,
          groupname = "yellow",
          color = list(
            background = "rgba(0, 0, 0, 0.5)",
            border = "rgba(255, 255, 0, 1)",
            highlight = list(
              background = "rgba(255, 255, 0, 0.5)",
              border = "rgba(255, 255, 0, 1)"
            )
          )
        )
        vis2 <- visGroups(
          vis2,
          groupname = "orange",
          color = list(
            background = "rgba(0, 0, 0, 0.5)",
            border = "rgba(255, 165, 0, 1)",
            highlight = list(
              background = "rgba(255, 165, 0, 0.5)",
              border = "rgba(255, 165, 0, 1)"
            )
          )
        )
        lnodes <- data.frame(
          label = c(
            "Downregulated\nPathways\n\n\nLegend",
            "Differentially\nEnriched",
            "Network\nHub",
            "LCM\nMass Spec",
            "FACS\nssRNAseq",
            "Both\nDatasets"
          ),
          color.background = c(
            "black", "black", "black", "black", "black", "black"
          ),
          color.border = c(
            "rgba(255, 255, 255, 1)", "rgba(255, 255, 255, 1)",
            "rgba(255, 255, 255, 1)", "rgba(255, 255, 0, 1)",
            "rgba(255, 165, 0, 1)", "rgba(0, 255, 0, 1)"
          ),
          shape = c("text", "box", "circle", "ellipse", "ellipse", "ellipse"),
          font.color = "white",
          font.size = c(15, 12, 12, 12, 12, 12)
        )
        vis2 <- visLegend(
          vis2, addNodes = lnodes, position = "right",
          useGroups = FALSE, stepY = 60, zoom = FALSE
        )
      }
      output$vis2_dn23 <- renderVisNetwork({vis2})

    })
  }
)
```

Pathways - LCM Mass Spec
===

```{r}
idx <- 2
sub_idx <- 0
```

Row {.tabset}
---

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_21_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_21_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = "Click row to see proteins. Drag bottom-right corner to resize.",
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 21)]]
      input[[paste0("scaling", 21)]]
      input[[paste0("init", 21)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 21)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 21)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 21)]] == "Scaled") {
        if (length(input[[paste0("gse", 21)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Arrows Indicate Significance ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 21)]]) > 0) {
        if (input[[paste0("scaling", 21)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 21), click_action = click_21_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 21), click_action = click_21_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_22_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_22_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 22)]]
      input[[paste0("scaling", 22)]]
      input[[paste0("init", 22)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 22)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 22)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 22)]] == "Scaled") {
        if (length(input[[paste0("gse", 22)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[2], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 22)]]) > 0) {
        if (input[[paste0("scaling", 22)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 22), click_action = click_22_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 22), click_action = click_22_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_23_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_23_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 23)]]
      input[[paste0("scaling", 23)]]
      input[[paste0("init", 23)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 23)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 23)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 23)]] == "Scaled") {
        if (length(input[[paste0("gse", 23)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[3], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 23)]]) > 0) {
        if (input[[paste0("scaling", 23)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 23), click_action = click_23_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 23), click_action = click_23_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_24_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_24_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 24)]]
      input[[paste0("scaling", 24)]]
      input[[paste0("init", 24)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 24)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 24)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 24)]] == "Scaled") {
        if (length(input[[paste0("gse", 24)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[4], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 24)]]) > 0) {
        if (input[[paste0("scaling", 24)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 24), click_action = click_24_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 24), click_action = click_24_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_25_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_25_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 25)]]
      input[[paste0("scaling", 25)]]
      input[[paste0("init", 25)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 25)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 25)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 25)]] == "Scaled") {
        if (length(input[[paste0("gse", 25)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[5], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 25)]]) > 0) {
        if (input[[paste0("scaling", 25)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 25), click_action = click_25_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 25), click_action = click_25_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_26_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_26_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 26)]]
      input[[paste0("scaling", 26)]]
      input[[paste0("init", 26)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 26)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 26)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 26)]] == "Scaled") {
        if (length(input[[paste0("gse", 26)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[6], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 26)]]) > 0) {
        if (input[[paste0("scaling", 26)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 26), click_action = click_26_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 26), click_action = click_26_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_27_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_27_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 27)]]
      input[[paste0("scaling", 27)]]
      input[[paste0("init", 27)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 27)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 27)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 27)]] == "Scaled") {
        if (length(input[[paste0("gse", 27)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[7], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 27)]]) > 0) {
        if (input[[paste0("scaling", 27)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 27), click_action = click_27_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 27), click_action = click_27_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_28_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_28_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 28)]]
      input[[paste0("scaling", 28)]]
      input[[paste0("init", 28)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 28)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 28)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 28)]] == "Scaled") {
        if (length(input[[paste0("gse", 28)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[8], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 28)]]) > 0) {
        if (input[[paste0("scaling", 28)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 28), click_action = click_28_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 28), click_action = click_28_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_29_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_29_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 29)]]
      input[[paste0("scaling", 29)]]
      input[[paste0("init", 29)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 29)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 29)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 29)]] == "Scaled") {
        if (length(input[[paste0("gse", 29)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[9], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 29)]]) > 0) {
        if (input[[paste0("scaling", 29)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 29), click_action = click_29_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 29), click_action = click_29_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_210_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_210_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:10],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 210)]]
      input[[paste0("scaling", 210)]]
      input[[paste0("init", 210)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 210)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 210)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 210)]] == "Scaled") {
        if (length(input[[paste0("gse", 210)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[10], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 210)]]) > 0) {
        if (input[[paste0("scaling", 210)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 210), click_action = click_210_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 210), click_action = click_210_unscaled
          )
        }
      }
    })
  },
)
```

Pathways - FACS ssRNAseq
===

```{r}
idx <- 3
sub_idx <- 0
```

Row {.tabset}
---

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_31_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_31_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = "Click row to see genes. Drag bottom-right corner to resize.",
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected genes."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 31)]]
      input[[paste0("scaling", 31)]]
      input[[paste0("init", 31)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 31)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 31)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 31)]] == "Scaled") {
        if (length(input[[paste0("gse", 31)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Arrows Indicate Significance ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 31)]]) > 0) {
        if (input[[paste0("scaling", 31)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 31), click_action = click_31_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 31), click_action = click_31_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_32_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_32_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see genes. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected genes."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 32)]]
      input[[paste0("scaling", 32)]]
      input[[paste0("init", 32)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 32)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 32)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 32)]] == "Scaled") {
        if (length(input[[paste0("gse", 32)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[2], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 32)]]) > 0) {
        if (input[[paste0("scaling", 32)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 32), click_action = click_32_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 32), click_action = click_32_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_33_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_33_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see genes. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected genes."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 33)]]
      input[[paste0("scaling", 33)]]
      input[[paste0("init", 33)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 33)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 33)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 33)]] == "Scaled") {
        if (length(input[[paste0("gse", 33)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[3], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 33)]]) > 0) {
        if (input[[paste0("scaling", 33)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 33), click_action = click_33_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 33), click_action = click_33_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_34_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_34_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see genes. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected genes."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 34)]]
      input[[paste0("scaling", 34)]]
      input[[paste0("init", 34)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 34)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 34)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 34)]] == "Scaled") {
        if (length(input[[paste0("gse", 34)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[4], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 34)]]) > 0) {
        if (input[[paste0("scaling", 34)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 34), click_action = click_34_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 34), click_action = click_34_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_35_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_35_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see genes. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected genes."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 35)]]
      input[[paste0("scaling", 35)]]
      input[[paste0("init", 35)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 35)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 35)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 35)]] == "Scaled") {
        if (length(input[[paste0("gse", 35)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[5], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 35)]]) > 0) {
        if (input[[paste0("scaling", 35)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 35), click_action = click_35_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 35), click_action = click_35_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_36_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_36_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see genes. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected genes."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 36)]]
      input[[paste0("scaling", 36)]]
      input[[paste0("init", 36)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 36)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 36)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 36)]] == "Scaled") {
        if (length(input[[paste0("gse", 36)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[6], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 36)]]) > 0) {
        if (input[[paste0("scaling", 36)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 36), click_action = click_36_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 36), click_action = click_36_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_37_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_37_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see genes. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected genes."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 37)]]
      input[[paste0("scaling", 37)]]
      input[[paste0("init", 37)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 37)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 37)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 37)]] == "Scaled") {
        if (length(input[[paste0("gse", 37)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[7], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 37)]]) > 0) {
        if (input[[paste0("scaling", 37)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 37), click_action = click_37_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 37), click_action = click_37_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_38_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_38_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see genes. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected genes."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 38)]]
      input[[paste0("scaling", 38)]]
      input[[paste0("init", 38)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 38)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 38)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 38)]] == "Scaled") {
        if (length(input[[paste0("gse", 38)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[8], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 38)]]) > 0) {
        if (input[[paste0("scaling", 38)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 38), click_action = click_38_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 38), click_action = click_38_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_39_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_39_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:30],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see genes. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected genes."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 39)]]
      input[[paste0("scaling", 39)]]
      input[[paste0("init", 39)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 39)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 39)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 39)]] == "Scaled") {
        if (length(input[[paste0("gse", 39)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[9], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 39)]]) > 0) {
        if (input[[paste0("scaling", 39)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 39), click_action = click_39_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 39), click_action = click_39_unscaled
          )
        }
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse[[idx]])[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_310_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_310_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:10],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see genes. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected genes."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 310)]]
      input[[paste0("scaling", 310)]]
      input[[paste0("init", 310)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 310)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 310)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 310)]] == "Scaled") {
        if (length(input[[paste0("gse", 310)]]) > 0) {
          mat <- t(
            apply(
              mat, MARGIN = 1,
              FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
            )
          )
        }
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[10], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = scale_range(nrow(mat))),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = scale_range(nrow(mat)) + 1)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat))))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (length(input[[paste0("gse", 310)]]) > 0) {
        if (input[[paste0("scaling", 310)]] == "Scaled") {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 310), click_action = click_310_scaled
          )
        } else {
          makeInteractiveComplexHeatmap(
            input, output, session, ht, res = 80,
            heatmap_id = paste0("ht", 310), click_action = click_310_unscaled
          )
        }
      }
    })
  },
)
```

Gene / Protein Scatter Plots
===

Row
---

### LCM Mass Spec

```{r}
df <- as.data.frame(logcounts(pseudobulk_deg[[2]]))
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(df), ]
gene_anno_sub <- gene_anno_sub[match(rownames(df), gene_anno_sub$ensembl), ]
rownames(df) <- gene_anno_sub$symbol
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Protein"
df <- melt(df, id.vars = "Protein")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Network Hub` <- ifelse(df$Protein %in% hubs[[2]], yes = "Yes", "No")

ggplotly(
  ggplot(df, aes(CTRL, NFT, label = Protein)) +
    geom_point(aes(color = `Network Hub`)) +
    geom_abline(intercept = 0, slope = 1) +
    xlab(paste(data_types[[2]], "in CTRL Neurons", sep = " ")) +
    ylab(paste(data_types[[2]], "in NFT Neurons", sep = " ")) +
    ggtitle(paste("LCM Mass Spec -", nrow(deg[[2]]), "Proteins", sep = " ")) +
    theme_light() +
    theme(plot.title = element_text(hjust = 0.5))
)
```

### FACS ssRNAseq

```{r}
df <- as.data.frame(logcounts(pseudobulk_deg[[3]]))
gene_anno_sub <- gene_anno[gene_anno$ensembl %in% rownames(df), ]
gene_anno_sub <- gene_anno_sub[match(rownames(df), gene_anno_sub$ensembl), ]
rownames(df) <- gene_anno_sub$symbol
df <- cbind(rownames(df), df)
colnames(df)[1] <- "Gene"
df <- melt(df, id.vars = "Gene")
df <- pivot_wider(df, names_from = variable, values_from = value)
df$`Network Hub` <- ifelse(df$Gene %in% hubs[[3]], yes = "Yes", "No")

ggplotly(
  ggplot(df, aes(CTRL, NFT, label = Gene)) +
    geom_point(aes(color = `Network Hub`)) +
    geom_abline(intercept = 0, slope = 1) +
    xlab(paste(data_types[[3]], "in CTRL Neurons", sep = " ")) +
    ylab(paste(data_types[[3]], "in NFT Neurons", sep = " ")) +
    ggtitle(paste("FACS ssRNAseq -", nrow(deg[[3]]), "Genes", sep = " ")) +
    theme_light() +
    theme(plot.title = element_text(hjust = 0.5))
)
```

Search Genes / Proteins
===

Row {.tabset}
---

### LCM Mass Spec

```{r}
shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("deg", "_deg2"),
      label = "Type to Search Proteins:",
      choices = res_deg[[2]]$Gene,
      selected = top_deg[[2]][1:10],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 2,
        radioButtons(
          paste0("scaling", "_deg2"), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      )
    ),
    fluidPage(
      div(
        style = "display:flex;",
        plotOutput("heatmap2", height = "660px"),
        sankeyNetworkOutput("sankey2", height = "685px")
      )
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("deg", "_deg2")]]
      input[[paste0("scaling", "_deg2")]]
    }, {

      features2 <- input[[paste0("deg", "_deg2")]]
      sig2 <- sig_deg[[2]][
        rownames(sig_deg[[2]]) %in% input[[paste0("deg", "_deg2")]], ,
        drop = FALSE
      ]
      features2 <- features2[features2 %in% rownames(sig2)]
      if (nrow(sig2) > 1) {
        sig2 <- sig2[match(features2, rownames(sig2)), , drop = FALSE]
      }

      gene_anno_sub2 <- gene_anno[gene_anno$symbol %in% features2, ]
      gene_anno_sub2 <- gene_anno_sub2[
        match(features2, gene_anno_sub2$symbol),
      ]

      mat2 <- deg[[2]][
        rownames(deg[[2]]) %in% gene_anno_sub2$ensembl, , drop = FALSE
      ]
      if (nrow(mat2) > 1) {
        mat2 <- mat2[
          match(gene_anno_sub2$ensembl, rownames(mat2)), , drop = FALSE
        ]
      }
      rownames(mat2) <- gene_anno_sub2$symbol

      gene_anno_sub2 <- gene_anno[gene_anno$ensembl %in% rownames(deg[[2]]), ]
      add_anno2 <- input[[paste0("deg", "_deg2")]][
        input[[paste0("deg", "_deg2")]] %notin% gene_anno_sub2$symbol
      ]
      add2 <- matrix(NA, nrow = length(add_anno2), ncol = ncol(deg[[2]]))
      rownames(add2) <- add_anno2
      mat2 <- rbind(mat2, add2)
      sig2 <- rbind(sig2, add2)
      sig2[is.na(sig2)] <- ""
      mat2 <- mat2[match(input[[paste0("deg", "_deg2")]], rownames(mat2)), ]
      sig2 <- sig2[match(input[[paste0("deg", "_deg2")]], rownames(sig2)), ]

      if (length(features2) == 1) {
        mat2 <- t(as.matrix(mat2))
        sig2 <- t(as.matrix(sig2))
        rownames(mat2) <- features2
        rownames(sig2) <- features2
      }

      if (input[[paste0("scaling", "_deg2")]] == "Scaled") {
        if (length(features2) > 0) {
          mat2 <- t(
            apply(
              mat2,
              MARGIN = 1,
              FUN = function (x) {
                (
                  (
                    2 * (x - min(x, na.rm = TRUE)) /
                      (max(x, na.rm = TRUE) - min(x, na.rm = TRUE))
                  ) - 1
                )
              }
            )
          )
        }
        col2 <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title2 <- paste0("Scaled ", data_types[2])
      } else {
        col2 <- colorRamp2(range_deg[[2]], colors = c("blue", "white", "red"))
        title2 <- data_types[2]
      }

      output$heatmap2 <- renderPlot({
        draw(
          Heatmap(
            mat2,
            col = col2,
            na_col = "darkgrey",
            column_title = paste0(
              data_sets[2], "\n",
              "Significant Genes (Adj. P < 0.05)"
            ),
            cluster_columns = FALSE,
            cluster_rows = FALSE,
            column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
            column_gap = unit(1, units = "mm"),
            rect_gp = gpar(col = "black", lwd = 0.25),
            row_names_gp = gpar(fontsize = scale_range(nrow(mat2))),
            column_title_gp = gpar(fontsize = 14),
            row_title_rot = 0,
            top_annotation = top_anno[[2]],
            show_column_names = FALSE,
            cell_fun = function(j, i, x, y, width, height, fill) {
              grid.text(sig2[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat2))))
            },
            heatmap_legend_param = list(
              title = title2, direction = "horizontal",
              title_position = "topcenter", labels_gp = gpar(fontsize = 10),
              grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
            )
          ),
          heatmap_legend_side = "top"
        )
      }, res = 96)

    })

    output$sankey2 <- renderSankeyNetwork({

      links_list2 <- vector(
        "list", length = length(input[[paste0("deg", "_deg2")]])
      )
      for (i in seq_along(input[[paste0("deg", "_deg2")]])) {
        gse_sub2 <- res_gse[[2]][res_gse[[2]]$P.value.adj < 0.05, ]
        deg_sub2 <- res_deg[[2]][
          res_deg[[2]]$Gene %in% input[[paste0("deg", "_deg2")]][i],
        ]$Coef
        if (deg_sub2 > 0) {
          gse_sub2 <- gse_sub2[gse_sub2$Coef > 0, ]
        } else {
          gse_sub2 <- gse_sub2[gse_sub2$Coef < 0, ]
        }
        gene_ids2 <- geneIds(gene_sets)[names(gene_sets) %in% gse_sub2$GeneSet]
        gene_ids2 <- gene_ids2[order(names(gene_ids2))]

        links_orig2 <- data.frame(
          source = rep(names(gene_ids2), times = lengths(gene_ids2)),
          target = gene_ids2 %>% map_dfr(as_tibble)
        )
        links2 <- links_orig2
        links_orig2[] <- map[unlist(links_orig2)]
        links2$value <- links_orig2$value
        colnames(links2) <- c("source", "target")
        links2 <- links2[
          links2$target %in% input[[paste0("deg", "_deg2")]][i],
        ]
        links2$value <- rep(1, times = nrow(links2))
        links_list2[[i]] <- links2
      }
      links_comb2 <- do.call(rbind, args = links_list2)

      nodes2 <- data.frame(
        name = c(
          as.character(links_comb2$source), as.character(links_comb2$target)
        ) %>% unique()
      )
      links_comb2$IDsource <- match(links_comb2$source, nodes2$name) - 1
      links_comb2$IDtarget <- match(links_comb2$target, nodes2$name) - 1

      if (nrow(nodes2) > 0) {
        select_length2 <- length(
          which(input[[paste0("deg", "_deg2")]] %in% nodes2$name)
        )
        nodes2$group <- c(
          rep("target", times = nrow(nodes2) - select_length2),
          nodes2$name[(nrow(nodes2) - (select_length2 - 1)):nrow(nodes2)]
        )
        colour2 <- c(hue_pal()(select_length2), "#D3D3D3")
        colour2 <- paste0(
          'd3.scaleOrdinal().range(["', paste(colour2, collapse = '", "'), '"])'
        )

        sankeyNetwork(
          Links = links_comb2, Nodes = nodes2, Source = "IDsource",
          Target = "IDtarget", Value = "value", NodeID = "name",
          sinksRight = FALSE, fontSize = 20, NodeGroup = "group",
          LinkGroup = "target", colourScale = colour2,
          margin = list(right = -400, left = -400, top = 100)
        )
      }
    })

  }
)
```

### FACS ssRNAseq

```{r}
shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("deg", "_deg3"),
      label = "Type to Search Genes:",
      choices = res_deg[[3]]$Gene,
      selected = top_deg[[3]][1:5],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 2,
        radioButtons(
          paste0("scaling", "_deg3"), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      )
    ),
    fluidPage(
      div(
        style = "display:flex;",
        plotOutput("heatmap3", height = "660px"),
        sankeyNetworkOutput("sankey3", height = "685px")
      )
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("deg", "_deg3")]]
      input[[paste0("scaling", "_deg3")]]
    }, {

      features3 <- input[[paste0("deg", "_deg3")]]
      sig3 <- sig_deg[[3]][
        rownames(sig_deg[[3]]) %in% input[[paste0("deg", "_deg3")]], ,
        drop = FALSE
      ]
      features3 <- features3[features3 %in% rownames(sig3)]
      if (nrow(sig3) > 1) {
        sig3 <- sig3[match(features3, rownames(sig3)), , drop = FALSE]
      }

      gene_anno_sub3 <- gene_anno[gene_anno$symbol %in% features3, ]
      gene_anno_sub3 <- gene_anno_sub3[
        match(features3, gene_anno_sub3$symbol),
      ]

      mat3 <- deg[[3]][
        rownames(deg[[3]]) %in% gene_anno_sub3$ensembl, , drop = FALSE
      ]
      if (nrow(mat3) > 1) {
        mat3 <- mat3[
          match(gene_anno_sub3$ensembl, rownames(mat3)), , drop = FALSE
        ]
      }
      rownames(mat3) <- gene_anno_sub3$symbol

      gene_anno_sub3 <- gene_anno[gene_anno$ensembl %in% rownames(deg[[3]]), ]
      add_anno3 <- input[[paste0("deg", "_deg3")]][
        input[[paste0("deg", "_deg3")]] %notin% gene_anno_sub3$symbol
      ]
      add3 <- matrix(NA, nrow = length(add_anno3), ncol = ncol(deg[[3]]))
      rownames(add3) <- add_anno3
      mat3 <- rbind(mat3, add3)
      sig3 <- rbind(sig3, add3)
      sig3[is.na(sig3)] <- ""
      mat3 <- mat3[match(input[[paste0("deg", "_deg3")]], rownames(mat3)), ]
      sig3 <- sig3[match(input[[paste0("deg", "_deg3")]], rownames(sig3)), ]

      if (length(features3) == 1) {
        mat3 <- t(as.matrix(mat3))
        sig3 <- t(as.matrix(sig3))
        rownames(mat3) <- features3
        rownames(sig3) <- features3
      }

      if (input[[paste0("scaling", "_deg3")]] == "Scaled") {
        if (length(features3) > 0) {
          mat3 <- t(
            apply(
              mat3,
              MARGIN = 1,
              FUN = function (x) {
                (
                  (
                    2 * (x - min(x, na.rm = TRUE)) /
                      (max(x, na.rm = TRUE) - min(x, na.rm = TRUE))
                  ) - 1
                )
              }
            )
          )
        }
        col3 <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title3 <- paste0("Scaled ", data_types[3])
      } else {
        col3 <- colorRamp2(range_deg[[3]], colors = c("blue", "white", "red"))
        title3 <- data_types[3]
      }

      output$heatmap3 <- renderPlot({
        draw(
          Heatmap(
            mat3,
            col = col3,
            na_col = "darkgrey",
            column_title = paste0(
              data_sets[3], "\n",
              "Significant Proteins (Adj. P < 0.05)"
            ),
            cluster_columns = FALSE,
            cluster_rows = FALSE,
            column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
            column_gap = unit(1, units = "mm"),
            rect_gp = gpar(col = "black", lwd = 0.25),
            row_names_gp = gpar(fontsize = scale_range(nrow(mat3))),
            column_title_gp = gpar(fontsize = 14),
            row_title_rot = 0,
            top_annotation = top_anno[[3]],
            show_column_names = FALSE,
            cell_fun = function(j, i, x, y, width, height, fill) {
              grid.text(sig3[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat3))))
            },
            heatmap_legend_param = list(
              title = title3, direction = "horizontal",
              title_position = "topcenter", labels_gp = gpar(fontsize = 10),
              grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
            )
          ),
          heatmap_legend_side = "top"
        )
      }, res = 96)

    })

    output$sankey3 <- renderSankeyNetwork({

      links_list3 <- vector(
        "list", length = length(input[[paste0("deg", "_deg3")]])
      )
      for (i in seq_along(input[[paste0("deg", "_deg3")]])) {
        gse_sub3 <- res_gse[[3]][res_gse[[3]]$P.value.adj < 0.05, ]
        deg_sub3 <- res_deg[[3]][
          res_deg[[3]]$Gene %in% input[[paste0("deg", "_deg3")]][i],
        ]$Coef
        if (deg_sub3 > 0) {
          gse_sub3 <- gse_sub3[gse_sub3$Coef > 0, ]
        } else {
          gse_sub3 <- gse_sub3[gse_sub3$Coef < 0, ]
        }
        gene_ids3 <- geneIds(gene_sets)[names(gene_sets) %in% gse_sub3$GeneSet]
        gene_ids3 <- gene_ids3[order(names(gene_ids3))]

        links_orig3 <- data.frame(
          source = rep(names(gene_ids3), times = lengths(gene_ids3)),
          target = gene_ids3 %>% map_dfr(as_tibble)
        )
        links3 <- links_orig3
        links_orig3[] <- map[unlist(links_orig3)]
        links3$value <- links_orig3$value
        colnames(links3) <- c("source", "target")
        links3 <- links3[
          links3$target %in% input[[paste0("deg", "_deg3")]][i],
        ]
        links3$value <- rep(1, times = nrow(links3))
        links_list3[[i]] <- links3
      }
      links_comb3 <- do.call(rbind, args = links_list3)

      nodes3 <- data.frame(
        name = c(
          as.character(links_comb3$source), as.character(links_comb3$target)
        ) %>% unique()
      )
      links_comb3$IDsource <- match(links_comb3$source, nodes3$name) - 1
      links_comb3$IDtarget <- match(links_comb3$target, nodes3$name) - 1

      if (nrow(nodes3) > 0) {
        select_length3 <- length(
          which(input[[paste0("deg", "_deg3")]] %in% nodes3$name)
        )
        nodes3$group <- c(
          rep("target", times = nrow(nodes3) - select_length3),
          nodes3$name[(nrow(nodes3) - (select_length3 - 1)):nrow(nodes3)]
        )
        colour3 <- c(hue_pal()(select_length3), "#D3D3D3")
        colour3 <- paste0(
          'd3.scaleOrdinal().range(["', paste(colour3, collapse = '", "'), '"])'
        )

        sankeyNetwork(
          Links = links_comb3, Nodes = nodes3, Source = "IDsource",
          Target = "IDtarget", Value = "value", NodeID = "name",
          sinksRight = FALSE, fontSize = 20, NodeGroup = "group",
          LinkGroup = "target", colourScale = colour3,
          margin = list(right = -400, left = -400, top = 100)
        )
      }
    })

  }
)
```

### Both Datasets

```{r}
shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("deg", "_deg"),
      label = "Type to Search Genes/Proteins:",
      choices = unique(c(top_deg[[2]], top_deg[[3]])),
      selected = intersect(top_deg[[2]], top_deg[[3]])[1:25],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 2,
        radioButtons(
          paste0("scaling", "_deg"), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      )
    ),
    fluidPage(
      div(
        style = "display:flex;",
        plotOutput("heatmap2c", height = "660px"),
        plotOutput("heatmap3c", height = "660px")
      )
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("deg", "_deg")]]
      input[[paste0("scaling", "_deg")]]
    }, {

      features2c <- input[[paste0("deg", "_deg")]]
      sig2c <- sig_deg[[2]][
        rownames(sig_deg[[2]]) %in% input[[paste0("deg", "_deg")]], ,
        drop = FALSE
      ]
      features2c <- features2c[features2c %in% rownames(sig2c)]
      if (nrow(sig2c) > 1) {
        sig2c <- sig2c[match(features2c, rownames(sig2c)), , drop = FALSE]
      }

      gene_anno_sub2c <- gene_anno[gene_anno$symbol %in% features2c, ]
      gene_anno_sub2c <- gene_anno_sub2c[
        match(features2c, gene_anno_sub2c$symbol),
      ]

      mat2c <- deg[[2]][
        rownames(deg[[2]]) %in% gene_anno_sub2c$ensembl, , drop = FALSE
      ]
      if (nrow(mat2c) > 1) {
        mat2c <- mat2c[
          match(gene_anno_sub2c$ensembl, rownames(mat2c)), , drop = FALSE
        ]
      }
      rownames(mat2c) <- gene_anno_sub2c$symbol

      gene_anno_sub2c <- gene_anno[gene_anno$ensembl %in% rownames(deg[[2]]), ]
      add_anno2c <- input[[paste0("deg", "_deg")]][
        input[[paste0("deg", "_deg")]] %notin% gene_anno_sub2c$symbol
      ]
      add2c <- matrix(NA, nrow = length(add_anno2c), ncol = ncol(deg[[2]]))
      rownames(add2c) <- add_anno2c
      mat2c <- rbind(mat2c, add2c)
      sig2c <- rbind(sig2c, add2c)
      sig2c[is.na(sig2c)] <- ""
      mat2c <- mat2c[match(input[[paste0("deg", "_deg")]], rownames(mat2c)), ]
      sig2c <- sig2c[match(input[[paste0("deg", "_deg")]], rownames(sig2c)), ]

      if (length(features2c) == 1) {
        mat2c <- t(as.matrix(mat2c))
        sig2c <- t(as.matrix(sig2c))
        rownames(mat2c) <- features2c
        rownames(sig2c) <- features2c
      }

      if (input[[paste0("scaling", "_deg")]] == "Scaled") {
        if (length(features2c) > 0) {
          mat2c <- t(
            apply(
              mat2c,
              MARGIN = 1,
              FUN = function (x) {
                (
                  (
                    2 * (x - min(x, na.rm = TRUE)) /
                      (max(x, na.rm = TRUE) - min(x, na.rm = TRUE))
                  ) - 1
                )
              }
            )
          )
        }
        col2c <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title2c <- paste0("Scaled ", data_types[2])
      } else {
        col2c <- colorRamp2(range_deg[[2]], colors = c("blue", "white", "red"))
        title2c <- data_types[2]
      }

      output$heatmap2c <- renderPlot({
        draw(
          Heatmap(
            mat2c,
            col = col2c,
            na_col = "darkgrey",
            column_title = paste0(
              data_sets[2], "\n",
              "Significant Proteins (Adj. P < 0.05)"
            ),
            cluster_columns = FALSE,
            cluster_rows = FALSE,
            column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
            column_gap = unit(1, units = "mm"),
            rect_gp = gpar(col = "black", lwd = 0.25),
            row_names_gp = gpar(fontsize = scale_range(nrow(mat2c))),
            column_title_gp = gpar(fontsize = 14),
            row_title_rot = 0,
            top_annotation = top_anno[[2]],
            show_column_names = FALSE,
            cell_fun = function(j, i, x, y, width, height, fill) {
              grid.text(sig2c[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat2c))))
            },
            heatmap_legend_param = list(
              title = title2c, direction = "horizontal",
              title_position = "topcenter", labels_gp = gpar(fontsize = 10),
              grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
            )
          ),
          heatmap_legend_side = "top"
        )
      }, res = 96)

      features3c <- input[[paste0("deg", "_deg")]]
      sig3c <- sig_deg[[3]][
        rownames(sig_deg[[3]]) %in% input[[paste0("deg", "_deg")]], ,
        drop = FALSE
      ]
      features3c <- features3c[features3c %in% rownames(sig3c)]
      if (nrow(sig3c) > 1) {
        sig3c <- sig3c[match(features3c, rownames(sig3c)), , drop = FALSE]
      }

      gene_anno_sub3c <- gene_anno[
        gene_anno$symbol %in% features3c,
      ]
      gene_anno_sub3c <- gene_anno_sub3c[
        match(features3c, gene_anno_sub3c$symbol),
      ]

      mat3c <- deg[[3]][
        rownames(deg[[3]]) %in% gene_anno_sub3c$ensembl, , drop = FALSE
      ]
      if (nrow(mat3c) > 1) {
        mat3c <- mat3c[
          match(gene_anno_sub3c$ensembl, rownames(mat3c)), , drop = FALSE
        ]
      }
      rownames(mat3c) <- gene_anno_sub3c$symbol

      gene_anno_sub3c <- gene_anno[
        gene_anno$symbol %in% input[[paste0("deg", "_deg")]],
      ]
      gene_anno_sub3c <- gene_anno_sub3c[
        match(input[[paste0("deg", "_deg")]], gene_anno_sub3c$symbol),
      ]

      gene_anno_sub3c <- gene_anno[gene_anno$ensembl %in% rownames(deg[[3]]), ]
      add_anno3c <- input[[paste0("deg", "_deg")]][
        input[[paste0("deg", "_deg")]] %notin% gene_anno_sub3c$symbol
      ]
      add3c <- matrix(NA, nrow = length(add_anno3c), ncol = ncol(deg[[3]]))
      rownames(add3c) <- add_anno3c
      mat3c <- rbind(mat3c, add3c)
      sig3c <- rbind(sig3c, add3c)
      sig3c[is.na(sig3c)] <- ""
      mat3c <- mat3c[match(input[[paste0("deg", "_deg")]], rownames(mat3c)), ]
      sig3c <- sig3c[match(input[[paste0("deg", "_deg")]], rownames(sig3c)), ]

      if (length(features3c) == 1) {
        mat3c <- t(as.matrix(mat3c))
        sig3c <- t(as.matrix(sig3c))
        rownames(mat3c) <- features3c
        rownames(sig3c) <- features3c
      }

      if (input[[paste0("scaling", "_deg")]] == "Scaled") {
        if (length(features3c) > 0) {
          mat3c <- t(
            apply(
              mat3c,
              MARGIN = 1,
              FUN = function (x) {
                (
                  (
                    2 * (x - min(x, na.rm = TRUE)) /
                      (max(x, na.rm = TRUE) - min(x, na.rm = TRUE))
                  ) - 1
                )
              }
            )
          )
        }
        col3c <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title3c <- paste0("Scaled ", data_types[3])
      } else {
        col3c <- colorRamp2(range_deg[[3]], colors = c("blue", "white", "red"))
        title3c <- data_types[3]
      }

      output$heatmap3c <- renderPlot({
        draw(
          Heatmap(
            mat3c,
            col = col3c,
            na_col = "darkgrey",
            column_title = paste0(
              data_sets[3], "\n",
              "Significant Genes (Adj. P < 0.05)"
            ),
            cluster_columns = FALSE,
            cluster_rows = FALSE,
            column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
            column_gap = unit(1, units = "mm"),
            rect_gp = gpar(col = "black", lwd = 0.25),
            row_names_gp = gpar(fontsize = scale_range(nrow(mat3c))),
            column_title_gp = gpar(fontsize = 14),
            row_title_rot = 0,
            top_annotation = top_anno[[3]],
            show_column_names = FALSE,
            cell_fun = function(j, i, x, y, width, height, fill) {
              grid.text(sig3c[i, j], x = x, y = y, gp = gpar(fontsize = scale_range(nrow(mat3c))))
            },
            heatmap_legend_param = list(
              title = title3c, direction = "horizontal",
              title_position = "topcenter", labels_gp = gpar(fontsize = 10),
              grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
            )
          ),
          heatmap_legend_side = "top"
        )
      }, res = 96)

    })
  }
)
```

Data Tables
===

Row {.tabset}
---

### LCM Mass Spec Proteins - UP

```{r}
datatable_download_exp_15(res_deg[[2]][res_deg[[2]]$Coef > 0, ])
```

### LCM Mass Spec Proteins - DOWN

```{r}
datatable_download_exp_15(res_deg[[2]][res_deg[[2]]$Coef < 0, ])
```

### FACS ssRNAseq Genes - UP

```{r}
datatable_download_exp_15(res_deg[[3]][res_deg[[3]]$Coef > 0, ])
```

### FACS ssRNAseq Genes - DOWN

```{r}
datatable_download_exp_15(res_deg[[3]][res_deg[[3]]$Coef < 0, ])
```

### LCM Mass Spec Pathways - UP

```{r}
datatable_download_exp_15(res_gse_fmt[[2]][res_gse_fmt[[2]]$Coef > 0, ])
```

### LCM Mass Spec Pathways - DOWN

```{r}
datatable_download_exp_15(res_gse_fmt[[2]][res_gse_fmt[[2]]$Coef < 0, ])
```

### FACS ssRNAseq Pathways - UP

```{r}
datatable_download_exp_15(res_gse_fmt[[3]][res_gse_fmt[[3]]$Coef > 0, ])
```

### FACS ssRNAseq Pathways - DOWN

```{r}
datatable_download_exp_15(res_gse_fmt[[3]][res_gse_fmt[[3]]$Coef < 0, ])
```


```{r, include = FALSE}
rm(
  deg, edges, gene_anno, gene_anno_sub, gene_sets, gene_sets_sub, graph, gse,
  hub_mat, layout, meta, n_deg, n_gse, net_data, net_data_nodup, net_data_sub,
  net_data2, net_data2_orig, net_data3, net_data3_orig, p_adj, range_deg,
  range_gse, res_deg, res_gse, res_gse_fmt, sig_deg, sig_gse, top_anno, top_deg,
  top_gse, vis, input, mat, sig, add, add_anno, mat2, mat3, sig2, sig3, add2,
  add3, add_anno2, add_anno3, gene_anno_sub2, gene_anno_sub3, gse_sub2,
  gene_ids2, links2, links_orig2, links_comb2, links_list2, nodes2, gse_sub3,
  gene_ids3, links3, links_orig3, links_comb3, links_list3, nodes3, deg_sub2,
  deg_sub3, mat2c, mat3c, sig2c, sig3c, add2c, add3c, add_anno2c, add_anno3c,
  gene_anno_sub2c, gene_anno_sub3c, df, hubs, pseudobulk_deg, n_gse, n_gse_up,
  n_gse_dn, n_deg, n_deg_up, n_deg_dn, res, venn_list_gse, venn_list_deg,
  vertices
)
gc()
```
