---
title: "Tangle Bearing Neurons Viewer"
output:
  flexdashboard::flex_dashboard:
    source_code: embed
runtime: shiny
---

```{r global, include = FALSE}
#    This file is part of tangle-bearing-neurons.
#    Copyright (C) 2024  Emir Turkes, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

library(conflicted)
library(scales)
library(ComplexHeatmap)
library(circlize)
library(DT)
library(GSEABase)

`%notin%` <- Negate(`%in%`)

data_sets <- c(
  "Laser-capture Microdissection Proteomics - Human AD BA9",
  "Single-cell Transcriptomics - Human AD BA9"
)
data_types <- c(
  "log2 LFQ Intensity", "log2 CPM",
  "log2 Enrichment Score", "log2 Enrichment Score"
)

data_dir <- file.path("..", "data")

gene_sets <- readRDS(file.path(data_dir, "gene_sets.rds"))
gene_anno <- readRDS(file.path(data_dir, "gene_anno.rds"))

meta <- vector("list", length = length(data_sets))
deg <- vector("list", length = length(data_sets))
gse <- vector("list", length = length(data_sets))
res_deg <- vector("list", length = length(data_sets))
res_gse <- vector("list", length = length(data_sets))

data_dir <- list.dirs(data_dir)
data_dir <- data_dir[c(2, 3)]
for (i in seq_along(data_sets)) {
  meta[[i]] <- readRDS(file.path(data_dir[i], "meta.rds"))
  deg[[i]] <- readRDS(file.path(data_dir[i], "deg.rds"))
  gse[[i]] <- readRDS(file.path(data_dir[i], "gse.rds"))
  res_deg[[i]] <- readRDS(file.path(data_dir[i], "res_deg.rds"))
  res_gse[[i]] <- readRDS(file.path(data_dir[i], "res_gse.rds"))
}

top_anno <- vector("list", length = length(data_sets))
for (i in seq_along(data_sets)) {
  n <- length(unique(meta[[i]]$donor))
  anno_colour <- hue_pal()(n)
  top_anno[[i]] <- HeatmapAnnotation(
    Donor = meta[[i]]$donor,
    group = anno_block(
      gp = gpar(fill = "lightgray", col = "lightgray"),
      labels = unique(meta[[i]]$tau),
      labels_gp = gpar(col = "black", fontsize = 14),
      height = unit(6, "mm")
    ),
    col = list(Donor = setNames(anno_colour, unique(meta[[i]]$donor))),
    simple_anno_size = unit(4, "mm"),
    show_legend = c("Donor" = FALSE),
    annotation_name_gp = gpar(fontsize = 12)
  )
}

p_adj <- p.adjust(
  unlist(c(lapply(res_deg, `[[`, "P.value"), lapply(res_gse, `[[`, "P.value"))),
  method = "BH"
)
p_adj <- split(
  p_adj,
  f = rep(
    seq_along(c(res_deg, res_gse)),
    times = sapply(c(res_deg, res_gse), FUN = nrow)
  )
)
for (i in seq_along(res_deg)) {
  res_deg[[i]]$P.value.adj <- p_adj[[i]]
}
for (i in seq_along(res_gse)) {
  res_gse[[i]]$P.value.adj <- p_adj[[i + length(data_sets)]]
}

priority_gse <- c(
  "tau-protein", "tau protein", "neurofi", "microtubule", "dynein", "cytoskele",
  "laminin", "inclusion", "aggrep", "death", "necro", "folding", "folded",
  "synap", "axon", "dendrit", "senescen", "apopto", "exocyt", "endocyt",
  "phagocyt", "proteasom", "lysosom", "endoplasmic", "Golgi", "Hsp",
  "heat shock", "ESCRT", "ubiqui", "heparan", "chondroitin", "sulfotransferase",
  "glycosaminoglycan", "transcription", "translation", "amyloid",
  "apolipoprotein", "lipoprotein", "fatty", "cholesterol", "insulin", "pruning",
  "phosphatidylserine", "complement", "T cell", "B cell", "immun", "inflammat",
  "stress", "cortisol", "interleukin", "leukocyte", "myeloid", "NF-kappaB",
  "toll-like", "macrophage", "astrocyte", "oligodendrocyte", "Schwann", "myelin"
)

range_gse <- vector("list", length = length(data_sets))
sig_gse <- vector("list", length = length(data_sets))
top_gse <- vector("list", length = length(data_sets))
n_gse <- vector("list", length = length(data_sets))

for (i in seq_along(data_sets)) {

  range_gse[[i]] <- c(
    min(gse[[i]]), (min(gse[[i]]) + max(gse[[i]])) / 2, max(gse[[i]])
  )

  sig <- ifelse(res_gse[[i]]$P.value.adj < 0.05, yes = 1, no = 0)
  direction <- ifelse(res_gse[[i]]$Coef > 0, yes = 1, no = -1)
  sig <- sig * direction

  names(sig) <- rownames(res_gse[[i]])
  sig <- sig[c(which(sig == 1), which(sig == -1), which(sig == 0))]
  sig <- ifelse(sig == -1, yes = "↓", no = ifelse(sig == 1, yes = "↑", no = ""))

  sig <- data.frame(CTRL = rep("", times = length(sig)), NFT = sig)
  sig_gse[[i]] <- vector("list", length = 2)
  for (j in seq_along(sig)) {
    sig_gse[[i]][[j]] <- replicate(nrow(meta[[i]]) / 2, expr = sig[ , j])
  }
  sig_gse[[i]] <- do.call(cbind, args = sig_gse[[i]])

  res_gse[[i]] <- res_gse[[i]][match(rownames(sig), rownames(res_gse[[i]])), ]
  rownames(sig_gse[[i]]) <- res_gse[[i]]$GeneSet

  n_gse[[i]] <- c(
    length(which(sig[ , 2] == "↑")), length(which(sig[ , 2] == "↓"))
  )

  top_gse[[i]] <- sig_gse[[i]][
    unique(
      unlist(sapply(priority_gse, FUN = grep, x = rownames(sig_gse[[i]])))
    ),
  ]
  sig <- sig[rownames(sig) %in% rownames(top_gse[[i]]), ]
  sig <- sig[match(rownames(top_gse[[i]]), rownames(sig)), ]
  top_gse[[i]] <- top_gse[[i]][
    c(which(sig[ , 2] == "↑")[1:8], which(sig[ , 2] == "↓")[1:8]),
  ]

}

range_deg <- vector("list", length = length(data_sets))
sig_deg <- vector("list", length = length(data_sets))
top_deg <- vector("list", length = length(data_sets))
n_deg <- vector("list", length = length(data_sets))

for (i in seq_along(data_sets)) {

  range_deg[[i]] <- c(
    min(deg[[i]]), (min(deg[[i]]) + max(deg[[i]])) / 2, max(deg[[i]])
  )

  sig <- ifelse(res_deg[[i]]$P.value.adj < 0.05, yes = 1, no = 0)
  direction <- ifelse(res_deg[[i]]$Coef > 0, yes = 1, no = -1)
  sig <- sig * direction

  names(sig) <- rownames(res_deg[[i]])
  sig <- sig[c(which(sig == 1), which(sig == -1), which(sig == 0))]
  sig <- ifelse(sig == -1, yes = "↓", no = ifelse(sig == 1, yes = "↑", no = ""))

  sig <- data.frame(CTRL = rep("", times = length(sig)), NFT = sig)
  sig_deg[[i]] <- vector("list", length = 2)
  for (j in seq_along(sig)) {
    sig_deg[[i]][[j]] <- replicate(nrow(meta[[i]]) / 2, expr = sig[ , j])
  }
  sig_deg[[i]] <- do.call(cbind, args = sig_deg[[i]])

  res_deg[[i]] <- res_deg[[i]][match(rownames(sig), rownames(res_deg[[i]])), ]
  rownames(sig_deg[[i]]) <- rownames(res_deg[[i]])

  n_deg[[i]] <- c(
    length(which(sig[ , 2] == "↑")), length(which(sig[ , 2] == "↓"))
  )

  priority_deg <- top_gse[[i]][
    unique(unlist(sapply(priority_gse, FUN = grep, x = rownames(top_gse[[i]])))),
  ]
  priority_deg <- gene_sets[names(gene_sets) %in% rownames(priority_deg)]
  priority_deg <- unique(unlist(geneIds(priority_deg)))
  top_deg[[i]] <- sig_deg[[i]][rownames(sig_deg[[i]]) %in% priority_deg, ]
  sig <- sig[rownames(sig) %in% rownames(top_deg[[i]]), ]
  sig <- sig[match(rownames(top_deg[[i]]), rownames(sig)), ]
  top_deg[[i]] <- top_deg[[i]][
    c(which(sig[ , 2] == "↑"), which(sig[ , 2] == "↓")),
  ]

  rownames(sig_deg[[i]]) <- res_deg[[i]]$Gene
  rownames(top_deg[[i]]) <- res_deg[[i]][
    rownames(res_deg[[i]]) %in% rownames(top_deg[[i]]),
  ]$Gene

}

datatable_download_exp_15 <- function(dt) {
  datatable(
    dt,
    options = list(
      pageLength = 15,
      scrollX = TRUE,
      dom = "Blfrtip",
      buttons = list(
        "copy", "print",
        list(
          extend = "collection", buttons = c("csv", "excel", "pdf"),
          text = "Download"
        )
      ),
      rowCallback = JS(
        "function(row, data) {",
        "for (i = 1; i < data.length; i++) {",
        "if (data[i]>=1000 | data[i]<1000) {",
        "$('td:eq('+i+')', row).html(data[i].toExponential(2));}}}"
      )
    ),
    extensions = "Buttons"
  )
}
```

Laser-capture Microdissection Proteomics - Human AD BA9
===

```{r}
idx <- 1
```

Row {.tabset}
---

### Differentially Enriched Pathways

```{r}
selectizeInput(
  paste0("gse", idx),
  label = "Type to Search Pathways:",
  choices = res_gse[[idx]]$GeneSet,
  selected = rownames(top_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", idx), label = NULL, c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", idx)]], , drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          n_gse[[idx]][1], " upregulated, ", n_gse[[idx]][2], " downregulated ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1400,
  height = 615,
  res = 96
)
```

</div>
</div>

### Enrichment Stats

```{r}
datatable_download_exp_15(res_gse[[idx]])
```

### Differentially Expressed Genes

```{r}
selectizeInput(
  paste0("deg", idx),
  label = "Type to Search Genes:",
  choices = res_deg[[idx]]$Gene,
  selected = rownames(top_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1

    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("deg", idx)]], gene_anno_sub$symbol),
    ]
    sig <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
    }

    mat <- deg[[idx]][
      rownames(deg[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE]
    }
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[idx]], colors = c("blue", "white", "red")),
        column_title = paste0(
          data_sets[idx], "\n",
          n_deg[[idx]][1], " upregulated, ", n_deg[[idx]][2], " downregulated ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = data_types[idx], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 1

    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("deg", idx)]], gene_anno_sub$symbol),
    ]
    sig <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
    }

    mat <- deg[[idx]][
      rownames(deg[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE]
    }
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = paste0(
          data_sets[idx], "\n",
          n_deg[[idx]][1], " upregulated, ", n_deg[[idx]][2], " downregulated ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[idx]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Protein Stats

```{r}
datatable_download_exp_15(res_deg[[idx]])
```

### Pathway Genes

```{r}
selectizeInput(
  paste0("path_deg", idx),
  label = "Type to Search Pathways:",
  choices = rownames(res_gse[[idx]]),
  selected = rownames(top_gse[[idx]])[1],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 1

    genes <- unlist(
      geneIds(gene_sets[names(gene_sets) %in% input[[paste0("path_deg", idx)]]])
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- genes[genes %notin% rownames(deg[[idx]])]
    add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
    }

    mat <- deg[[idx]][
      rownames(deg[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[idx]]))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig <- rbind(sig, add)
    sig[is.na(sig)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          data_sets[idx], ": ",
          gene_sets[[input[[paste0("path_deg", idx)]]]]@shortDescription, "\n",
          input[[paste0("path_deg", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 7))
        },
        heatmap_legend_param = list(
          title = data_types[idx], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 1

    genes <- unlist(
      geneIds(gene_sets[names(gene_sets) %in% input[[paste0("path_deg", idx)]]])
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
    }

    mat <- deg[[idx]][
      rownames(deg[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[idx]]))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig <- rbind(sig, add)
    sig[is.na(sig)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          data_sets[idx], ": ",
          gene_sets[[input[[paste0("path_deg", idx)]]]]@shortDescription, "\n",
          input[[paste0("path_deg", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 7))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[idx]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

Single-cell Transcriptomics - Human AD BA9
===

```{r}
idx <- 2
```

Row {.tabset}
---

### Differentially Enriched Pathways

```{r}
selectizeInput(
  paste0("gse", idx),
  label = "Type to Search Pathways:",
  choices = res_gse[[idx]]$GeneSet,
  selected = rownames(top_gse[[idx]]),
  multiple = TRUE,
  width = "100%"
)

radioButtons(
  paste0("scaling", idx), label = NULL, c("Scaled", "Unscaled"), inline = TRUE
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2

    sig <- sig_gse[[idx]][
      rownames(sig_gse[[idx]]) %in% input[[paste0("gse", idx)]], , drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[
        match(input[[paste0("gse", idx)]], rownames(sig)), , drop = FALSE
      ]
    }

    mat <- gse[[idx]][
      rownames(gse[[idx]]) %in% rownames(sig), , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
    }

    if (input[[paste0("scaling", idx)]] == "Scaled") {
      mat <- t(
        apply(
          mat, MARGIN = 1,
          FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
        )
      )
      col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
      title <- paste0("Scaled ", data_types[idx + 2])
    } else {
      col <- colorRamp2(range_gse[[idx]], colors = c("blue", "white", "red"))
      title <- data_types[idx + 2]
    }

    draw(
      Heatmap(
        mat,
        col = col,
        column_title = paste0(
          data_sets[idx], "\n",
          n_gse[[idx]][1], " upregulated, ", n_gse[[idx]][2], " downregulated ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        row_names_max_width = max_text_width(
          rownames(mat), gp = gpar(fontsize = 14)
        ),
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = title, direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 1400,
  height = 615,
  res = 96
)
```

</div>
</div>

### Enrichment Stats

```{r}
datatable_download_exp_15(res_gse[[idx]])
```

### Differentially Expressed Genes

```{r}
selectizeInput(
  paste0("deg", idx),
  label = "Type to Search Genes:",
  choices = res_deg[[idx]]$Gene,
  selected = rownames(top_deg[[idx]]),
  multiple = TRUE,
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2

    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("deg", idx)]], gene_anno_sub$symbol),
    ]
    sig <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
    }

    mat <- deg[[idx]][
      rownames(deg[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE]
    }
    rownames(mat) <- gene_anno_sub$symbol

    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[idx]], colors = c("blue", "white", "red")),
        column_title = paste0(
          data_sets[idx], "\n",
          n_deg[[idx]][1], " upregulated, ", n_deg[[idx]][2], " downregulated ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 7))
        },
        heatmap_legend_param = list(
          title = data_types[idx], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 2

    gene_anno_sub <- gene_anno[
      gene_anno$symbol %in% input[[paste0("deg", idx)]],
    ]
    gene_anno_sub <- gene_anno_sub[
      match(input[[paste0("deg", idx)]], gene_anno_sub$symbol),
    ]
    sig <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
    }

    mat <- deg[[idx]][
      rownames(deg[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE]
    }
    rownames(mat) <- gene_anno_sub$symbol

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        column_title = paste0(
          data_sets[idx], "\n",
          n_deg[[idx]][1], " upregulated, ", n_deg[[idx]][2], " downregulated ",
          "(Adj. P < 0.05)"
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 7))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[idx]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

### Gene Stats

```{r}
datatable_download_exp_15(res_deg[[idx]])
```

### Pathway Genes

```{r}
selectizeInput(
  paste0("path_deg", idx),
  label = "Type to Search Pathways:",
  choices = rownames(res_gse[[idx]]),
  selected = rownames(top_gse[[idx]])[2],
  width = "100%"
)
```

<div style='display:flex; flex-direction:row; justify-content:space-evenly;'>
<div>

```{r}
renderPlot(
  {
    idx <- 2

    genes <- unlist(
      geneIds(gene_sets[names(gene_sets) %in% input[[paste0("path_deg", idx)]]])
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- genes[genes %notin% rownames(deg[[idx]])]
    add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
    }

    mat <- deg[[idx]][
      rownames(deg[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[idx]]))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig <- rbind(sig, add)
    sig[is.na(sig)] <- ""

    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[idx]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          data_sets[idx], ": ",
          gene_sets[[input[[paste0("path_deg", idx)]]]]@shortDescription, "\n",
          input[[paste0("path_deg", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 7))
        },
        heatmap_legend_param = list(
          title = data_types[idx], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
<div>

```{r}
renderPlot(
  {
    idx <- 2

    genes <- unlist(
      geneIds(gene_sets[names(gene_sets) %in% input[[paste0("path_deg", idx)]]])
    )
    gene_anno_sub <- gene_anno[gene_anno$ensembl %in% genes, ]
    gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
    add_anno <- gene_anno_sub[
      gene_anno_sub$symbol %notin% rownames(sig_deg[[idx]]),
    ]
    gene_anno_sub <- gene_anno_sub[
      gene_anno_sub$symbol %in% rownames(sig_deg[[idx]]),
    ]
    sig <- sig_deg[[idx]][
      rownames(sig_deg[[idx]]) %in% gene_anno_sub$symbol, , drop = FALSE
    ]
    if (nrow(sig) > 1) {
      sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
    }

    mat <- deg[[idx]][
      rownames(deg[[idx]]) %in% gene_anno_sub$ensembl, , drop = FALSE
    ]
    if (nrow(mat) > 1) {
      mat <- mat[
        match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
      ]
    }
    if (nrow(mat) > 0) {
      rownames(mat) <- gene_anno_sub$symbol
    }

    add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[idx]]))
    rownames(add) <- add_anno$symbol
    mat <- rbind(mat, add)
    sig <- rbind(sig, add)
    sig[is.na(sig)] <- ""

    mat <- t(
      apply(
        mat, MARGIN = 1,
        FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
      )
    )
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          data_sets[idx], ": ",
          gene_sets[[input[[paste0("path_deg", idx)]]]]@shortDescription, "\n",
          input[[paste0("path_deg", idx)]]
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[idx]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[idx]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 7))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[idx]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  },
  width = 750,
  height = 700,
  res = 96
)
```

</div>
</div>

```{r, include = FALSE}
rm(data_dir, n, anno_colour, p_adj, direction, priority_gse, priority_deg)
gc()
```
