---
title: "Tangle Bearing Neurons Viewer"
output:
  flexdashboard::flex_dashboard:
    source_code: embed
runtime: shiny
---

```{r global, include = FALSE}
#    This file is part of tangle-bearing-neurons.
#    Copyright (C) 2024  Emir Turkes, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

library(conflicted)
library(scales)
library(InteractiveComplexHeatmap)
library(circlize)
library(GSEABase)

`%notin%` <- Negate(`%in%`)

data_sets <- c(
  "LCM Mass Spec - Human AD BA9 - Single-cell",
  "LCM Mass Spec - Human AD BA9 - Pooled-cells",
  "FACS-sorted ssRNAseq - Human AD BA9"
)
data_types <- c(
  "log2 LFQ Intensity", "log2 LFQ Intensity", "log2 CPM",
  "log2 Enrichment Score", "log2 Enrichment Score", "log2 Enrichment Score"
)

data_dir <- file.path("..", "data")
cache_dir <- file.path("..", "cache")
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}

gene_sets <- readRDS(file.path(data_dir, "gene_sets.rds"))
gene_anno <- readRDS(file.path(data_dir, "gene_anno.rds"))

meta <- vector("list", length = length(data_sets))
deg <- vector("list", length = length(data_sets))
gse <- vector("list", length = length(data_sets))
res_deg <- vector("list", length = length(data_sets))
res_gse <- vector("list", length = length(data_sets))

data_dir <- list.dirs(data_dir)
data_dir <- data_dir[c(2, 3, 4)]
for (i in seq_along(data_sets)) {
  meta[[i]] <- readRDS(file.path(data_dir[i], "meta.rds"))
  deg[[i]] <- readRDS(file.path(data_dir[i], "deg.rds"))
  gse[[i]] <- readRDS(file.path(data_dir[i], "gse.rds"))
  res_deg[[i]] <- readRDS(file.path(data_dir[i], "res_deg.rds"))
  res_gse[[i]] <- readRDS(file.path(data_dir[i], "res_gse.rds"))
}

top_anno <- vector("list", length = length(data_sets))
for (i in seq_along(data_sets)) {
  n <- length(unique(meta[[i]]$donor))
  anno_colour <- hue_pal()(n)
  top_anno[[i]] <- HeatmapAnnotation(
    Donor = meta[[i]]$donor,
    group = anno_block(
      gp = gpar(fill = "lightgray", col = "lightgray"),
      labels = unique(meta[[i]]$tau),
      labels_gp = gpar(col = "black", fontsize = 14),
      height = unit(6, "mm")
    ),
    col = list(Donor = setNames(anno_colour, unique(meta[[i]]$donor))),
    simple_anno_size = unit(4, "mm"),
    show_legend = c("Donor" = FALSE),
    annotation_name_gp = gpar(fontsize = 12)
  )
}

p_adj <- p.adjust(
  unlist(c(lapply(res_deg, `[[`, "P.value"), lapply(res_gse, `[[`, "P.value"))),
  method = "BH"
)
p_adj <- split(
  p_adj,
  f = rep(
    seq_along(c(res_deg, res_gse)),
    times = sapply(c(res_deg, res_gse), FUN = nrow)
  )
)
for (i in seq_along(res_deg)) {
  res_deg[[i]]$P.value.adj <- p_adj[[i]]
}
for (i in seq_along(res_gse)) {
  res_gse[[i]]$P.value.adj <- p_adj[[i + length(data_sets)]]
}

priority_gse <- c(
  `Tau & Neurofibrillary Tangle` =
    "tau-protein | tau protein | neurofi | inclusion | aggrep",
  `Microtubule & Cytoskeletal` =
    "microtubule | dynein | cytoskele | laminin",
  `Death, Stress, & Ageing` =
    "death | necro | senescen | stress",
  `Protein Folding` =
    "folding | folded",
  `Synapse, Axon, & Dendrite` =
    "synap | axon | dendrit",
  `Exocytosis & Endocytosis` =
    "exocyt | endocyt",
  `Apoptosis & Autophagy`=
    "apopto | autophag | phagocyt",
  `Proteasomal & Lysosomal`=
    "proteasom | lysosom | endoplasmic | Golgi",
  `Glycosaminoglycan`=
    "heparan | chondroitin | sulfotransferase | glycosaminoglycan",
  `Transcription & Translation` =
    "transcription | translation",
  `Amyloid & Lipoprotein` =
    "amyloid | lipoprotein",
  `Lipid Metabolism`=
    "fatty | cholesterol | insulin",
  `Synaptic Pruning` =
    "pruning | phosphatidylserine | complement",
  `T Cell, B Cell, & Other Immune Cell`=
    "T cell | B cell | leukocyte | myeloid | lymphocyte",
  `Inflammatory Response` =
    "immun | inflammat | cortisol | interleukin | NF-kappaB | toll-like",
  `Macrophage & Astrocyte` =
    "macrophage | astrocyte",
  `Oligodendrocyte & Myelin` =
    "oligodendrocyte | Schwann | myelin",
  `Developmental`=
    "development | differentiation | morphogenesis"
)

if (file.exists(file.path(cache_dir, "range_gse.rds"))) {
  range_gse <- readRDS(file.path(cache_dir, "range_gse.rds"))
  sig_gse <- readRDS(file.path(cache_dir, "sig_gse.rds"))
  top_gse <- readRDS(file.path(cache_dir, "top_gse.rds"))
  n_gse <- readRDS(file.path(cache_dir, "n_gse.rds"))
  range_deg <- readRDS(file.path(cache_dir, "range_deg.rds"))
  sig_deg <- readRDS(file.path(cache_dir, "sig_deg.rds"))
  top_deg <- readRDS(file.path(cache_dir, "top_deg.rds"))
  n_deg <- readRDS(file.path(cache_dir, "n_deg.rds"))
} else {

  ref <- sapply(
    priority_gse,
    function (x) {
      patterns <- unlist(strsplit(x, " \\| "))
      matches <- sapply(
        patterns, function(y) {
          grep(paste0(".*", y, ".*"), names(gene_sets), value = TRUE)
        }
      )
      unique(unlist(matches))
    }
  )

  range_gse <- vector("list", length = length(data_sets))
  sig_gse <- vector("list", length = length(data_sets))
  top_gse <- vector("list", length = length(data_sets))
  n_gse <- vector("list", length = length(data_sets))

  for (i in seq_along(data_sets)) {

    range_gse[[i]] <- c(
      min(gse[[i]]), (min(gse[[i]]) + max(gse[[i]])) / 2, max(gse[[i]])
    )

    sig <- ifelse(res_gse[[i]]$P.value.adj < 0.05, yes = 1, no = 0)
    direction <- ifelse(res_gse[[i]]$Coef > 0, yes = 1, no = -1)
    sig <- sig * direction

    names(sig) <- rownames(res_gse[[i]])
    sig <- sig[c(which(sig == 1), which(sig == -1), which(sig == 0))]
    sig <- ifelse(
      sig == -1, yes = "↓", no = ifelse(sig == 1, yes = "↑", no = "")
    )

    sig <- data.frame(CTRL = rep("", times = length(sig)), NFT = sig)
    sig_gse[[i]] <- vector("list", length = 2)
    for (j in seq_along(sig)) {
      sig_gse[[i]][[j]] <- replicate(nrow(meta[[i]]) / 2, expr = sig[ , j])
    }
    sig_gse[[i]] <- do.call(cbind, args = sig_gse[[i]])

    res_gse[[i]] <- res_gse[[i]][match(rownames(sig), rownames(res_gse[[i]])), ]
    rownames(sig_gse[[i]]) <- res_gse[[i]]$GeneSet

    top_gse[[i]] <- sapply(
      priority_gse,
      function (x) {
        patterns <- unlist(strsplit(x, " \\| "))
        matches <- sapply(
          patterns, function(y) {
            grep(
              paste0(".*", y, ".*"),
              rownames(res_gse[[i]][res_gse[[i]]$P.value.adj < 0.05, ]),
              value = TRUE
            )
          }
        )
        unique(unlist(matches))
      }
    )

    n_gse[[i]] <- Map(
      function(x, y) {
        ratio <- round(length(x) / length(y), digits = 2) * 100
        ifelse(is.na(ratio), yes = 0, no = ratio)
      },
      top_gse[[i]], ref
    )

  }

  range_deg <- vector("list", length = length(data_sets))
  sig_deg <- vector("list", length = length(data_sets))
  top_deg <- vector("list", length = length(data_sets))
  n_deg <- vector("list", length = length(data_sets))

  for (i in seq_along(data_sets)) {

    range_deg[[i]] <- c(
      min(deg[[i]]), (min(deg[[i]]) + max(deg[[i]])) / 2, max(deg[[i]])
    )

    sig <- ifelse(res_deg[[i]]$P.value.adj < 0.05, yes = 1, no = 0)
    direction <- ifelse(res_deg[[i]]$Coef > 0, yes = 1, no = -1)
    sig <- sig * direction

    names(sig) <- rownames(res_deg[[i]])
    sig <- sig[c(which(sig == 1), which(sig == -1), which(sig == 0))]
    sig <- ifelse(
      sig == -1, yes = "↓", no = ifelse(sig == 1, yes = "↑", no = "")
    )

    sig <- data.frame(CTRL = rep("", times = length(sig)), NFT = sig)
    sig_deg[[i]] <- vector("list", length = 2)
    for (j in seq_along(sig)) {
      sig_deg[[i]][[j]] <- replicate(nrow(meta[[i]]) / 2, expr = sig[ , j])
    }
    sig_deg[[i]] <- do.call(cbind, args = sig_deg[[i]])

    res_deg[[i]] <- res_deg[[i]][match(rownames(sig), rownames(res_deg[[i]])), ]
    rownames(sig_deg[[i]]) <- res_deg[[i]]$Gene

  }
  saveRDS(range_gse, file = file.path(cache_dir, "range_gse.rds"))
  saveRDS(sig_gse, file = file.path(cache_dir, "sig_gse.rds"))
  saveRDS(top_gse, file = file.path(cache_dir, "top_gse.rds"))
  saveRDS(n_gse, file = file.path(cache_dir, "n_gse.rds"))
  saveRDS(range_deg, file = file.path(cache_dir, "range_deg.rds"))
  saveRDS(sig_deg, file = file.path(cache_dir, "sig_deg.rds"))
  saveRDS(top_deg, file = file.path(cache_dir, "top_deg.rds"))
  saveRDS(n_deg, file = file.path(cache_dir, "n_deg.rds"))
}
```

Laser-capture Microdissection Mass Spec - Human AD BA9 - Single-cell
===

```{r}
idx <- 1
sub_idx <- 0
```

Row {.tabset}
---

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%",
    " Of Tested Pathways Significant)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_11_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 96)
}

click_11_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 96)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    radioButtons(
      paste0("scaling", comb_idx), label = NULL,
      c("Scaled", "Unscaled"), inline = TRUE
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = "Click row to See proteins. Drag bottom-right corner to resize.",
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1200, height1 = 600, response = "click",
      output_ui = plotOutput("heatmap", width = 600, height = 610)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 11)]]
      input[[paste0("scaling", 11)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 11)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 11)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 11)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[1], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 14),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 11)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 96,
          heatmap_id = paste0("ht", 11), click_action = click_11_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 96,
          heatmap_id = paste0("ht", 11), click_action = click_11_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_12_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 96)
}

click_12_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 14),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 96)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    radioButtons(
      paste0("scaling", comb_idx), label = NULL,
      c("Scaled", "Unscaled"), inline = TRUE
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize.
        Resizing also fixes legend position."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1200, height1 = 600, response = "click",
      output_ui = plotOutput("heatmap", width = 600, height = 610)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 12)]]
      input[[paste0("scaling", 12)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 12)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 12)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 12)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[1], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 14),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 12))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 12)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 96,
          heatmap_id = paste0("ht", 12), click_action = click_12_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 96,
          heatmap_id = paste0("ht", 12), click_action = click_12_unscaled
        )
      }
    })
  },
)
```

```{r, include = FALSE}
rm(p_adj, anno_colour, cache_dir, data_dir, priority_gse)
gc()
```
