---
title: "Tangle Bearing Neurons Viewer"
output:
  flexdashboard::flex_dashboard:
runtime: shiny
---

```{r global, include = FALSE}
#    This file is part of tangle-bearing-neurons.
#    Copyright (C) 2024  Emir Turkes, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

library(conflicted)
conflicts_prefer(igraph::degree, base::intersect)
library(scales)
library(InteractiveComplexHeatmap)
library(circlize)
library(GSEABase)
library(reshape2)
library(dplyr)
library(igraph)
library(qgraph)
library(visNetwork)
library(purrr)
library(networkD3)

`%notin%` <- Negate(`%in%`)

pdf(NULL)

data_sets <- c(
  "LCM Mass Spec - Human AD BA9 - Single-cell",
  "LCM Mass Spectrometry - Human AD BA9",
  "FACS-sorted ssRNAseq - Human AD BA9"
)
data_types <- c(
  "log2 LFQ Intensity", "log2 LFQ Intensity", "log2 CPM",
  "log2 Enrichment Score", "log2 Enrichment Score", "log2 Enrichment Score"
)

data_dir <- file.path("..", "data")
cache_dir <- file.path("..", "cache")
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}

gene_sets <- readRDS(file.path(data_dir, "gene_sets.rds"))
gene_anno <- readRDS(file.path(data_dir, "gene_anno.rds"))
map <- setNames(gene_anno$symbol, gene_anno$ensembl)

meta <- vector("list", length = length(data_sets))
deg <- vector("list", length = length(data_sets))
gse <- vector("list", length = length(data_sets))
res_deg <- vector("list", length = length(data_sets))
res_gse <- vector("list", length = length(data_sets))

data_dir <- list.dirs(data_dir)
data_dir <- data_dir[c(2, 3, 4)]
for (i in seq_along(data_sets)) {
  meta[[i]] <- readRDS(file.path(data_dir[i], "meta.rds"))
  deg[[i]] <- readRDS(file.path(data_dir[i], "deg.rds"))
  gse[[i]] <- readRDS(file.path(data_dir[i], "gse.rds"))
  res_deg[[i]] <- readRDS(file.path(data_dir[i], "res_deg.rds"))
  res_gse[[i]] <- readRDS(file.path(data_dir[i], "res_gse.rds"))
}

top_anno <- vector("list", length = length(data_sets))
for (i in seq_along(data_sets)) {
  n <- length(unique(meta[[i]]$donor))
  anno_colour <- hue_pal()(n)
  top_anno[[i]] <- HeatmapAnnotation(
    Donor = meta[[i]]$donor,
    group = anno_block(
      gp = gpar(fill = "lightgray", col = "lightgray"),
      labels = unique(meta[[i]]$tau),
      labels_gp = gpar(col = "black", fontsize = 14),
      height = unit(6, "mm")
    ),
    col = list(Donor = setNames(anno_colour, unique(meta[[i]]$donor))),
    simple_anno_size = unit(4, "mm"),
    show_legend = c("Donor" = FALSE),
    annotation_name_gp = gpar(fontsize = 12)
  )
}

p_adj <- p.adjust(
  unlist(c(lapply(res_deg, `[[`, "P.value"), lapply(res_gse, `[[`, "P.value"))),
  method = "BH"
)
p_adj <- split(
  p_adj,
  f = rep(
    seq_along(c(res_deg, res_gse)),
    times = sapply(c(res_deg, res_gse), FUN = nrow)
  )
)
for (i in seq_along(res_deg)) {
  res_deg[[i]]$P.value.adj <- p_adj[[i]]
}
for (i in seq_along(res_gse)) {
  res_gse[[i]]$P.value.adj <- p_adj[[i + length(data_sets)]]
}

priority_gse <- c(
  `Tau & Neurofibrillary Tangle` =
    "tau-protein | tau protein | neurofi | inclusion | aggrep",
  `Microtubule & Cytoskeletal` =
    "microtubule | dynein | cytoskele | laminin",
  `Death, Stress, & Ageing` =
    "death | necro | senescen | stress",
  `Protein Folding` =
    "folding | folded",
  `Synapse, Axon, & Dendrite` =
    "synap | axon | dendrit",
  `Exocytosis & Endocytosis` =
    "exocyt | endocyt",
  `Apoptosis & Autophagy`=
    "apopto | autophag | phagocyt",
  `Proteasomal & Lysosomal`=
    "proteasom | lysosom | endoplasmic | Golgi",
  `Glycosaminoglycan`=
    "heparan | chondroitin | sulfotransferase | glycosaminoglycan",
  `Transcription & Translation` =
    "transcription | translation | dna | dbp | rna | rbp",
  `Amyloid & Lipoprotein` =
    "amyloid | lipoprotein",
  `Lipid Metabolism`=
    "fatty | cholesterol | insulin",
  `Synaptic Pruning` =
    "pruning | phosphatidylserine | complement",
  `T Cell, B Cell, & Other Immune Cell`=
    "T cell | B cell | leukocyte | myeloid | lymphocyte",
  `Inflammatory Response` =
    "immun | inflammat | cortisol | interleukin | NF-kappaB | toll-like",
  `Macrophage & Astrocyte` =
    "macrophage | astrocyte",
  `Oligodendrocyte & Myelin` =
    "oligodendrocyte | Schwann | myelin"
)

if (file.exists(file.path(cache_dir, "range_gse.rds"))) {
  range_gse <- readRDS(file.path(cache_dir, "range_gse.rds"))
  sig_gse <- readRDS(file.path(cache_dir, "sig_gse.rds"))
  top_gse <- readRDS(file.path(cache_dir, "top_gse.rds"))
  n_gse <- readRDS(file.path(cache_dir, "n_gse.rds"))
  range_deg <- readRDS(file.path(cache_dir, "range_deg.rds"))
  sig_deg <- readRDS(file.path(cache_dir, "sig_deg.rds"))
  top_deg <- readRDS(file.path(cache_dir, "top_deg.rds"))
  n_deg <- readRDS(file.path(cache_dir, "n_deg.rds"))
} else {

  ref <- sapply(
    priority_gse,
    function (x) {
      patterns <- unlist(strsplit(x, " \\| "))
      matches <- sapply(
        patterns, function(y) {
          grep(paste0(".*", y, ".*"), names(gene_sets), value = TRUE)
        }
      )
      unique(unlist(matches))
    }
  )

  range_gse <- vector("list", length = length(data_sets))
  sig_gse <- vector("list", length = length(data_sets))
  top_gse <- vector("list", length = length(data_sets))
  n_gse <- vector("list", length = length(data_sets))

  for (i in seq_along(data_sets)) {

    range_gse[[i]] <- c(
      min(gse[[i]]), (min(gse[[i]]) + max(gse[[i]])) / 2, max(gse[[i]])
    )

    sig <- ifelse(res_gse[[i]]$P.value.adj < 0.05, yes = 1, no = 0)
    direction <- ifelse(res_gse[[i]]$Coef > 0, yes = 1, no = -1)
    sig <- sig * direction

    names(sig) <- rownames(res_gse[[i]])
    sig <- sig[c(which(sig == 1), which(sig == -1), which(sig == 0))]
    sig <- ifelse(
      sig == -1, yes = "↓", no = ifelse(sig == 1, yes = "↑", no = "")
    )

    sig <- data.frame(CTRL = rep("", times = length(sig)), NFT = sig)
    sig_gse[[i]] <- vector("list", length = 2)
    for (j in seq_along(sig)) {
      sig_gse[[i]][[j]] <- replicate(nrow(meta[[i]]) / 2, expr = sig[ , j])
    }
    sig_gse[[i]] <- do.call(cbind, args = sig_gse[[i]])

    res_gse[[i]] <- res_gse[[i]][match(rownames(sig), rownames(res_gse[[i]])), ]
    rownames(sig_gse[[i]]) <- res_gse[[i]]$GeneSet

    top_gse[[i]] <- sapply(
      priority_gse,
      function (x) {
        patterns <- unlist(strsplit(x, " \\| "))
        matches <- sapply(
          patterns, function(y) {
            grep(
              paste0(".*", y, ".*"),
              rownames(res_gse[[i]][res_gse[[i]]$P.value.adj < 0.05, ]),
              value = TRUE
            )
          }
        )
        unique(unlist(matches))
      }
    )

    n_gse[[i]] <- Map(
      function(x, y) {
        ratio <- round(length(x) / length(y), digits = 2) * 100
        ifelse(is.na(ratio), yes = 0, no = ratio)
      },
      top_gse[[i]], ref
    )

    top_gse[[i]]$`All Other` <-
      rownames(res_gse[[i]][res_gse[[i]]$P.value.adj < 0.05, ])[
        rownames(res_gse[[i]][res_gse[[i]]$P.value.adj < 0.05, ]) %notin%
          unlist(top_gse[[i]])
      ]
  }

  range_deg <- vector("list", length = length(data_sets))
  sig_deg <- vector("list", length = length(data_sets))
  top_deg <- vector("list", length = length(data_sets))
  n_deg <- vector("list", length = length(data_sets))

  for (i in seq_along(data_sets)) {

    range_deg[[i]] <- c(
      min(deg[[i]]), (min(deg[[i]]) + max(deg[[i]])) / 2, max(deg[[i]])
    )

    sig <- ifelse(res_deg[[i]]$P.value.adj < 0.05, yes = 1, no = 0)
    direction <- ifelse(res_deg[[i]]$Coef > 0, yes = 1, no = -1)
    sig <- sig * direction

    names(sig) <- rownames(res_deg[[i]])
    sig <- sig[c(which(sig == 1), which(sig == -1), which(sig == 0))]
    sig <- ifelse(
      sig == -1, yes = "↓", no = ifelse(sig == 1, yes = "↑", no = "")
    )

    sig <- data.frame(CTRL = rep("", times = length(sig)), NFT = sig)
    sig_deg[[i]] <- vector("list", length = 2)
    for (j in seq_along(sig)) {
      sig_deg[[i]][[j]] <- replicate(nrow(meta[[i]]) / 2, expr = sig[ , j])
    }
    sig_deg[[i]] <- do.call(cbind, args = sig_deg[[i]])

    res_deg[[i]] <- res_deg[[i]][match(rownames(sig), rownames(res_deg[[i]])), ]
    rownames(sig_deg[[i]]) <- res_deg[[i]]$Gene

    top_deg[[i]] <- res_deg[[i]][res_deg[[i]]$P.value.adj < 0.05, ]$Gene
    n_deg[[i]] <- length(top_deg[[i]])

  }
  saveRDS(range_gse, file = file.path(cache_dir, "range_gse.rds"))
  saveRDS(sig_gse, file = file.path(cache_dir, "sig_gse.rds"))
  saveRDS(top_gse, file = file.path(cache_dir, "top_gse.rds"))
  saveRDS(n_gse, file = file.path(cache_dir, "n_gse.rds"))
  saveRDS(range_deg, file = file.path(cache_dir, "range_deg.rds"))
  saveRDS(sig_deg, file = file.path(cache_dir, "sig_deg.rds"))
  saveRDS(top_deg, file = file.path(cache_dir, "top_deg.rds"))
  saveRDS(n_deg, file = file.path(cache_dir, "n_deg.rds"))
}

priority_gse <- c(priority_gse, `All Other` = NA)
```

Laser-capture Microdissection Mass Spectrometry - Human AD BA9
===

```{r}
idx <- 2
sub_idx <- 0
```

Row {.tabset}
---

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%",
    " Of Tested Pathways Significant)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_21_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_21_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = "Click row to See proteins. Drag bottom-right corner to resize.",
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 21)]]
      input[[paste0("scaling", 21)]]
      input[[paste0("init", 21)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 21)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 21)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 21)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[2], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 21)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 21), click_action = click_21_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 21), click_action = click_21_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_22_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_22_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 22)]]
      input[[paste0("scaling", 22)]]
      input[[paste0("init", 22)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 22)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 22)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 22)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[2], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 22)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 22), click_action = click_22_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 22), click_action = click_22_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_23_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_23_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 23)]]
      input[[paste0("scaling", 23)]]
      input[[paste0("init", 23)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 23)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 23)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 23)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[3], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 23)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 23), click_action = click_23_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 23), click_action = click_23_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_24_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_24_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 24)]]
      input[[paste0("scaling", 24)]]
      input[[paste0("init", 24)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 24)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 24)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 24)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[4], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 24)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 24), click_action = click_24_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 24), click_action = click_24_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_25_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_25_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 25)]]
      input[[paste0("scaling", 25)]]
      input[[paste0("init", 25)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 25)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 25)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 25)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[5], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 25)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 25), click_action = click_25_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 25), click_action = click_25_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_26_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_26_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 26)]]
      input[[paste0("scaling", 26)]]
      input[[paste0("init", 26)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 26)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 26)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 26)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[6], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 26)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 26), click_action = click_26_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 26), click_action = click_26_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_27_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_27_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 27)]]
      input[[paste0("scaling", 27)]]
      input[[paste0("init", 27)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 27)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 27)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 27)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[7], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 27)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 27), click_action = click_27_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 27), click_action = click_27_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_28_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_28_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 28)]]
      input[[paste0("scaling", 28)]]
      input[[paste0("init", 28)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 28)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 28)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 28)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[8], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 28)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 28), click_action = click_28_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 28), click_action = click_28_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_29_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_29_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 29)]]
      input[[paste0("scaling", 29)]]
      input[[paste0("init", 29)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 29)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 29)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 29)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[9], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 29)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 29), click_action = click_29_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 29), click_action = click_29_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_210_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_210_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 210)]]
      input[[paste0("scaling", 210)]]
      input[[paste0("init", 210)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 210)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 210)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 210)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[10], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 210)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 210), click_action = click_210_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 210), click_action = click_210_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_211_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[11]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_211_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[11]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 211)]]
      input[[paste0("scaling", 211)]]
      input[[paste0("init", 211)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 211)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 211)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 211)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[11], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 211)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 211), click_action = click_211_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 211), click_action = click_211_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_212_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[12]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_212_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[12]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 212)]]
      input[[paste0("scaling", 212)]]
      input[[paste0("init", 212)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 212)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 212)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 212)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[12], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 212)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 212), click_action = click_212_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 212), click_action = click_212_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_213_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[13]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_213_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[13]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 213)]]
      input[[paste0("scaling", 213)]]
      input[[paste0("init", 213)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 213)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 213)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 213)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[13], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 213)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 213), click_action = click_213_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 213), click_action = click_213_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_214_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[14]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_214_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[14]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 214)]]
      input[[paste0("scaling", 214)]]
      input[[paste0("init", 214)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 214)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 214)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 214)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[14], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 214)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 214), click_action = click_214_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 214), click_action = click_214_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_215_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[15]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_215_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[15]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 215)]]
      input[[paste0("scaling", 215)]]
      input[[paste0("init", 215)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 215)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 215)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 215)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[15], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 215)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 215), click_action = click_215_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 215), click_action = click_215_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_216_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[16]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_216_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[16]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 216)]]
      input[[paste0("scaling", 216)]]
      input[[paste0("init", 216)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 216)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 216)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 216)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[16], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 216)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 216), click_action = click_216_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 216), click_action = click_216_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_217_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[17]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_217_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[17]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 217)]]
      input[[paste0("scaling", 217)]]
      input[[paste0("init", 217)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 217)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 217)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 217)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[17], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 217)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 217), click_action = click_217_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 217), click_action = click_217_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse)[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_218_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[18]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_218_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[18]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 218)]]
      input[[paste0("scaling", 218)]]
      input[[paste0("init", 218)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 218)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 218)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 218)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[18], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 218)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 218), click_action = click_218_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 218), click_action = click_218_unscaled
        )
      }
    })
  },
)
```

FACS-sorted Single-soma RNAseq - Human AD BA9
===

```{r}
idx <- 3
sub_idx <- 0
```

Row {.tabset}
---

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%",
    " Of Tested Pathways Significant)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_31_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_31_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = "Click row to See proteins. Drag bottom-right corner to resize.",
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 31)]]
      input[[paste0("scaling", 31)]]
      input[[paste0("init", 31)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 31)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 31)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 31)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[3], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 31)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 31), click_action = click_31_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 31), click_action = click_31_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_32_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_32_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 32)]]
      input[[paste0("scaling", 32)]]
      input[[paste0("init", 32)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 32)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 32)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 32)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[3], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 32)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 32), click_action = click_32_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 32), click_action = click_32_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_33_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_33_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 33)]]
      input[[paste0("scaling", 33)]]
      input[[paste0("init", 33)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 33)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 33)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 33)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[3], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 33)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 33), click_action = click_33_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 33), click_action = click_33_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_34_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_34_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 34)]]
      input[[paste0("scaling", 34)]]
      input[[paste0("init", 34)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 34)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 34)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 34)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[4], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 34)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 34), click_action = click_34_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 34), click_action = click_34_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_35_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_35_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 35)]]
      input[[paste0("scaling", 35)]]
      input[[paste0("init", 35)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 35)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 35)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 35)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[5], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 35)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 35), click_action = click_35_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 35), click_action = click_35_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_36_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_36_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 36)]]
      input[[paste0("scaling", 36)]]
      input[[paste0("init", 36)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 36)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 36)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 36)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[6], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 36)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 36), click_action = click_36_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 36), click_action = click_36_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_37_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_37_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 37)]]
      input[[paste0("scaling", 37)]]
      input[[paste0("init", 37)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 37)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 37)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 37)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[7], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 37)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 37), click_action = click_37_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 37), click_action = click_37_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_38_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_38_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 38)]]
      input[[paste0("scaling", 38)]]
      input[[paste0("init", 38)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 38)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 38)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 38)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[8], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 38)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 38), click_action = click_38_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 38), click_action = click_38_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_39_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_39_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 39)]]
      input[[paste0("scaling", 39)]]
      input[[paste0("init", 39)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 39)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 39)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 39)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[9], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 39)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 39), click_action = click_39_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 39), click_action = click_39_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_310_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_310_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 310)]]
      input[[paste0("scaling", 310)]]
      input[[paste0("init", 310)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 310)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 310)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 310)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[10], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 310)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 310), click_action = click_310_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 310), click_action = click_310_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_311_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[11]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_311_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[11]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 311)]]
      input[[paste0("scaling", 311)]]
      input[[paste0("init", 311)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 311)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 311)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 311)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[11], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 311)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 311), click_action = click_311_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 311), click_action = click_311_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_312_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[12]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_312_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[12]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 312)]]
      input[[paste0("scaling", 312)]]
      input[[paste0("init", 312)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 312)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 312)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 312)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[12], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 312)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 312), click_action = click_312_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 312), click_action = click_312_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_313_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[13]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_313_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[13]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 313)]]
      input[[paste0("scaling", 313)]]
      input[[paste0("init", 313)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 313)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 313)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 313)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[13], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 313)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 313), click_action = click_313_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 313), click_action = click_313_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_314_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[14]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_314_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[14]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 314)]]
      input[[paste0("scaling", 314)]]
      input[[paste0("init", 314)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 314)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 314)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 314)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[14], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 314)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 314), click_action = click_314_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 314), click_action = click_314_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_315_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[15]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_315_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[15]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 315)]]
      input[[paste0("scaling", 315)]]
      input[[paste0("init", 315)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 315)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 315)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 315)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[15], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 315)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 315), click_action = click_315_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 315), click_action = click_315_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_316_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[16]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_316_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[16]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 316)]]
      input[[paste0("scaling", 316)]]
      input[[paste0("init", 316)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 316)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 316)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 316)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[16], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 316)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 316), click_action = click_316_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 316), click_action = click_316_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_317_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[17]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_317_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[17]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 317)]]
      input[[paste0("scaling", 317)]]
      input[[paste0("init", 317)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 317)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 317)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 317)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[17], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 317)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 317), click_action = click_317_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 217), click_action = click_317_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(paste0("### ", names(priority_gse)[sub_idx + 1]))
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_317_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[18]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_318_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[18]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = "padding-left: 100px;"
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 517, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 525)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 318)]]
      input[[paste0("scaling", 318)]]
      input[[paste0("init", 318)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 318)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 318)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 318)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[18], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 318)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 318), click_action = click_318_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 218), click_action = click_318_unscaled
        )
      }
    })
  },
)
```

Network Analysis
===

Row {.tabset}
---

### Shared - UP

```{r}
shinyApp(

  ui = fluidPage(
    fluidRow(
      column(
        width = 2,
        sliderInput(
          paste0("conn_up", 23),
          label = "Connectivity:",
          min = 0,
          max = 5,
          value = 1,
          step = 0.1
        ),
      ),
      column(
        width = 2,
        radioButtons(
          paste0("shared_up", 23), label = NULL,
          c("All", "Shared Only"), inline = TRUE
        ),
        style = "padding-top: 35px;"
      )
    ),
    visNetworkOutput(paste0("vis_up", 23), height = "660px")
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("conn_up", 23)]]
      input[[paste0("shared_up", 23)]]
    }, {

      de_gse <- rownames(res_gse[[2]][res_gse[[2]]$P.value.adj < 0.05, ])
      up_gse <- rownames(res_gse[[2]][res_gse[[2]]$Coef > 0, ])
      de_deg <- rownames(res_deg[[2]][res_deg[[2]]$P.value.adj < 0.05, ])
      up_deg <- rownames(res_deg[[2]][res_deg[[2]]$Coef > 0, ])

      net_data <- melt(geneIds(gene_sets[names(gene_sets) %in% de_gse]))
      names(net_data)[1] <- "ensembl"
      names(net_data)[2] <- "gene_set"

      gene_anno_sub <- gene_anno[gene_anno$ensembl %in% net_data$ensembl, ]
      net_data <- left_join(net_data, gene_anno_sub, by = "ensembl")
      net_data$gse_direction <- ifelse(
        net_data$gene_set %in% up_gse, yes = "up", no = "down"
      )
      net_data$gene_DE <- ifelse(
        net_data$ensembl %in% de_deg, yes = "yes", no = "no"
      )
      net_data$gene_direction <- ifelse(
        net_data$ensembl %in% up_deg, yes = "up", no = "down"
      )
      net_data$dataset <- 2

      remove <- which(net_data$gse_direction != net_data$gene_direction)
      net_data <- net_data[-remove, ]
      net_data <- net_data[net_data$gene_DE == "yes", ]

      net_data2 <- net_data

      de_gse <- rownames(res_gse[[3]][res_gse[[3]]$P.value.adj < 0.05, ])
      up_gse <- rownames(res_gse[[3]][res_gse[[3]]$Coef > 0, ])
      de_deg <- rownames(res_deg[[3]][res_deg[[3]]$P.value.adj < 0.05, ])
      up_deg <- rownames(res_deg[[3]][res_deg[[3]]$Coef > 0, ])

      net_data <- melt(geneIds(gene_sets[names(gene_sets) %in% de_gse]))
      names(net_data)[1] <- "ensembl"
      names(net_data)[2] <- "gene_set"

      gene_anno_sub <- gene_anno[gene_anno$ensembl %in% net_data$ensembl, ]
      net_data <- left_join(net_data, gene_anno_sub, by = "ensembl")
      net_data$gse_direction <- ifelse(
        net_data$gene_set %in% up_gse, yes = "up", no = "down"
      )
      net_data$gene_DE <- ifelse(
        net_data$ensembl %in% de_deg, yes = "yes", no = "no"
      )
      net_data$gene_direction <- ifelse(
        net_data$ensembl %in% up_deg, yes = "up", no = "down"
      )
      net_data$dataset <- 3

      remove <- which(net_data$gse_direction != net_data$gene_direction)
      net_data <- net_data[-remove, ]
      net_data <- net_data[net_data$gene_DE == "yes", ]

      net_data3 <- net_data

      net_data <- rbind(net_data2, net_data3)

      if (input[[paste0("shared_up", 23)]] == "Shared Only") {
        net_data <- net_data[
          net_data$ensembl %in% net_data2$ensembl &
            net_data$ensembl %in% net_data3$ensembl,
        ]
      }

      net_data_up <- net_data[which(net_data$gse_direction == "up"), ]
      hub_mat <- as.data.frame.matrix(table(net_data_up[ , c(3, 2)]))
      hub_mat <- hub_mat == 1
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )

      net_data_table <- table(net_data_up$symbol)
      gene_sets_table <- table(unlist(geneIds(gene_sets)))
      gene_anno_sub <- gene_anno[
        gene_anno$ensembl %in% names(gene_sets_table),
      ]
      gene_anno_sub <- gene_anno_sub[
        match(names(gene_sets_table), gene_anno_sub$ensembl),
      ]
      names(gene_sets_table) <- gene_anno_sub$symbol
      gene_sets_table <- gene_sets_table[
        names(gene_sets_table) %in% names(net_data_table)
      ]
      gene_sets_table <- gene_sets_table[
        match(names(net_data_table), names(gene_sets_table))
      ]

      net_data_table <- (net_data_table / gene_sets_table) * net_data_table

      remove <- net_data_table[
        net_data_table < input[[paste0("conn_up", 23)]]
      ]
      if (length(remove) > 0) {
        remove <- which(net_data_up$symbol %in% names(remove))
        net_data_up <- net_data_up[-remove, ]
        remove <- V(graph)$name %notin% net_data_up$symbol
        graph <- delete.vertices(graph, v = remove)
      }

      if (length(names(V(graph))) < 0) {
        return
      }

      graph <- subgraph(graph, which(degree(graph) > 0))

      clust <- cluster_edge_betweenness(graph)
      keep <- c()
      for (membership in seq_along(unique(clust$membership))) {
        sub <- clust$names[which(clust$membership == membership)]
        sub <- net_data_table[names(net_data_table) %in% sub]
        keep <- c(keep, names(sub[which.max(sub)]))
      }
      keep_idx <- which(V(graph)$name %in% keep)

      keep2 <- unique(unlist(neighborhood(graph, order = 1, nodes = keep_idx)))
      keep2 <- unique(c(keep_idx, keep2))
      graph <- induced_subgraph(graph, vids = keep2)

      edges <- get.edgelist(graph, names = FALSE)
      layout <- qgraph.layout.fruchtermanreingold(
        edges, vcount = vcount(graph), weights = E(graph)$weight,
        area = vcount(graph), repulse.rad = vcount(graph) / 4
      )
      vis_up <- toVisNetworkData(graph)

      net_data_up_nodup <- net_data_up[-which(duplicated(net_data_up$symbol)), ]
      net_data_up_nodup <- net_data_up_nodup[
        net_data_up_nodup$symbol %in% vis_up$nodes$id,
      ]
      net_data_up_nodup <- net_data_up_nodup[
        match(vis_up$nodes$id, net_data_up_nodup$symbol),
      ]
      vis_up$nodes$shape <- ifelse(
        net_data_up_nodup$symbol %in% keep, yes = "circle", no = "box"
      )

      net_data_table <- net_data_table[
        names(net_data_table) %in% vis_up$nodes$id
      ]
      vis_up$nodes$size <- (net_data_table + 5) ^ 1.8
      vis_up$edges$width <- (vis_up$edges$weight * 2) ^ 1.5

      if (length(names(V(graph))) > 0) {
        col <- case_when(
          names(V(graph)) %in% net_data2$symbol &
            names(V(graph)) %in% net_data3$symbol ~ "green",
          names(V(graph)) %in% net_data2$symbol ~ "yellow",
          names(V(graph)) %in% net_data3$symbol ~ "orange"
        )
        vis_up$nodes$color.border <- col
        vis_up$nodes$color.highlight <- col

        f <- pmax((max(nchar(vis_up$nodes$id)) - nchar(vis_up$nodes$id)) / 2, 0)
        vis_up$nodes$label <- sprintf(
          "%-*s%s%*s", f, "", vis_up$nodes$id, ceiling(f), ""
        )
        vis_up$nodes$font.size <- vis_up$nodes$size

        title <- vector(length = length(vis_up$nodes$id))
        for (j in seq_along(title)) {
          net_data_sub <- net_data_up[
            net_data_up$symbol == vis_up$nodes$id[j],
          ]
          title[j] <- paste(net_data_sub$gene_set, collapse = "<br>")
        }
        vis_up$nodes$title <- title

        vis_up <- visNetwork(vis_up$nodes, vis_up$edges, background = "black")
        vis_up <- visPhysics(vis_up, stabilization = FALSE)
        vis_up <- visEdges(
          vis_up, smooth = FALSE, color = list(opacity = 0.8, inherit = "both")
        )
        vis_up <- visIgraphLayout(vis_up, "layout.norm", layoutMatrix = layout)
        vis_up <- visOptions(
          vis_up,
          highlightNearest = list(
            enabled = TRUE, degree = 1, labelOnly = FALSE
          ),
          nodesIdSelection = TRUE
        )
        vis_up <- visNodes(
          vis_up, color = list(background = "rgba(0, 0, 0, 0)"),
          font = list(color = "white")
        )
      }

      output$vis_up23 <- renderVisNetwork({vis_up})
    })
  }
)
```

### Shared - DOWN

```{r}
shinyApp(

  ui = fluidPage(
    fluidRow(
      column(
        width = 2,
        sliderInput(
          paste0("conn_dn", 23),
          label = "Connectivity:",
          min = 0,
          max = 5,
          value = 1,
          step = 0.1
        ),
      ),
      column(
        width = 2,
        radioButtons(
          paste0("shared_dn", 23), label = NULL,
          c("All", "Shared Only"), inline = TRUE
        ),
        style = "padding-top: 35px;"
      )
    ),
    visNetworkOutput(paste0("vis_dn", 23), height = "660px")
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("conn_dn", 23)]]
      input[[paste0("shared_dn", 23)]]
    }, {

      de_gse <- rownames(res_gse[[2]][res_gse[[2]]$P.value.adj < 0.05, ])
      up_gse <- rownames(res_gse[[2]][res_gse[[2]]$Coef > 0, ])
      de_deg <- rownames(res_deg[[2]][res_deg[[2]]$P.value.adj < 0.05, ])
      up_deg <- rownames(res_deg[[2]][res_deg[[2]]$Coef > 0, ])

      net_data <- melt(geneIds(gene_sets[names(gene_sets) %in% de_gse]))
      names(net_data)[1] <- "ensembl"
      names(net_data)[2] <- "gene_set"

      gene_anno_sub <- gene_anno[gene_anno$ensembl %in% net_data$ensembl, ]
      net_data <- left_join(net_data, gene_anno_sub, by = "ensembl")
      net_data$gse_direction <- ifelse(
        net_data$gene_set %in% up_gse, yes = "up", no = "down"
      )
      net_data$gene_DE <- ifelse(
        net_data$ensembl %in% de_deg, yes = "yes", no = "no"
      )
      net_data$gene_direction <- ifelse(
        net_data$ensembl %in% up_deg, yes = "up", no = "down"
      )
      net_data$dataset <- 2

      remove <- which(net_data$gse_direction != net_data$gene_direction)
      net_data <- net_data[-remove, ]
      net_data <- net_data[net_data$gene_DE == "yes", ]

      net_data2 <- net_data

      de_gse <- rownames(res_gse[[3]][res_gse[[3]]$P.value.adj < 0.05, ])
      up_gse <- rownames(res_gse[[3]][res_gse[[3]]$Coef > 0, ])
      de_deg <- rownames(res_deg[[3]][res_deg[[3]]$P.value.adj < 0.05, ])
      up_deg <- rownames(res_deg[[3]][res_deg[[3]]$Coef > 0, ])

      net_data <- melt(geneIds(gene_sets[names(gene_sets) %in% de_gse]))
      names(net_data)[1] <- "ensembl"
      names(net_data)[2] <- "gene_set"

      gene_anno_sub <- gene_anno[gene_anno$ensembl %in% net_data$ensembl, ]
      net_data <- left_join(net_data, gene_anno_sub, by = "ensembl")
      net_data$gse_direction <- ifelse(
        net_data$gene_set %in% up_gse, yes = "up", no = "down"
      )
      net_data$gene_DE <- ifelse(
        net_data$ensembl %in% de_deg, yes = "yes", no = "no"
      )
      net_data$gene_direction <- ifelse(
        net_data$ensembl %in% up_deg, yes = "up", no = "down"
      )
      net_data$dataset <- 3

      remove <- which(net_data$gse_direction != net_data$gene_direction)
      net_data <- net_data[-remove, ]
      net_data <- net_data[net_data$gene_DE == "yes", ]

      net_data3 <- net_data

      net_data <- rbind(net_data2, net_data3)

      if (input[[paste0("shared_dn", 23)]] == "Shared Only") {
        net_data <- net_data[
          net_data$ensembl %in% net_data2$ensembl &
            net_data$ensembl %in% net_data3$ensembl,
        ]
      }

      net_data_dn <- net_data[which(net_data$gse_direction == "down"), ]
      hub_mat <- as.data.frame.matrix(table(net_data_dn[ , c(3, 2)]))
      hub_mat <- hub_mat == 1
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )

      net_data_table <- table(net_data_dn$symbol)
      gene_sets_table <- table(unlist(geneIds(gene_sets)))
      gene_anno_sub <- gene_anno[
        gene_anno$ensembl %in% names(gene_sets_table),
      ]
      gene_anno_sub <- gene_anno_sub[
        match(names(gene_sets_table), gene_anno_sub$ensembl),
      ]
      names(gene_sets_table) <- gene_anno_sub$symbol
      gene_sets_table <- gene_sets_table[
        names(gene_sets_table) %in% names(net_data_table)
      ]
      gene_sets_table <- gene_sets_table[
        match(names(net_data_table), names(gene_sets_table))
      ]

      net_data_table <- (net_data_table / gene_sets_table) * net_data_table

      remove <- net_data_table[
        net_data_table < input[[paste0("conn_dn", 23)]]
      ]
      if (length(remove) > 0) {
        remove <- which(net_data_dn$symbol %in% names(remove))
        net_data_dn <- net_data_dn[-remove, ]
        remove <- V(graph)$name %notin% net_data_dn$symbol
        graph <- delete.vertices(graph, v = remove)
      }

      if (length(names(V(graph))) < 0) {
        return
      }

      graph <- subgraph(graph, which(degree(graph) > 0))

      clust <- cluster_edge_betweenness(graph)
      keep <- c()
      for (membership in seq_along(unique(clust$membership))) {
        sub <- clust$names[which(clust$membership == membership)]
        sub <- net_data_table[names(net_data_table) %in% sub]
        keep <- c(keep, names(sub[which.max(sub)]))
      }
      keep_idx <- which(V(graph)$name %in% keep)

      keep2 <- unique(unlist(neighborhood(graph, order = 1, nodes = keep_idx)))
      keep2 <- unique(c(keep_idx, keep2))
      graph <- induced_subgraph(graph, vids = keep2)

      edges <- get.edgelist(graph, names = FALSE)
      layout <- qgraph.layout.fruchtermanreingold(
        edges, vcount = vcount(graph), weights = E(graph)$weight,
        area = vcount(graph), repulse.rad = vcount(graph) / 4
      )
      vis_dn <- toVisNetworkData(graph)

      net_data_dn_nodup <- net_data_dn[-which(duplicated(net_data_dn$symbol)), ]
      net_data_dn_nodup <- net_data_dn_nodup[
        net_data_dn_nodup$symbol %in% vis_dn$nodes$id,
      ]
      net_data_dn_nodup <- net_data_dn_nodup[
        match(vis_dn$nodes$id, net_data_dn_nodup$symbol),
      ]
      vis_dn$nodes$shape <- ifelse(
        net_data_dn_nodup$symbol %in% keep, yes = "circle", no = "box"
      )

      net_data_table <- net_data_table[
        names(net_data_table) %in% vis_dn$nodes$id
      ]
      vis_dn$nodes$size <- (net_data_table + 5) ^ 1.8
      vis_dn$edges$width <- (vis_dn$edges$weight * 2) ^ 1.5

      if (length(names(V(graph))) > 0) {
        col <- case_when(
          names(V(graph)) %in% net_data2$symbol &
            names(V(graph)) %in% net_data3$symbol ~ "green",
          names(V(graph)) %in% net_data2$symbol ~ "yellow",
          names(V(graph)) %in% net_data3$symbol ~ "orange"
        )
        vis_dn$nodes$color.border <- col
        vis_dn$nodes$color.highlight <- col

        f <- pmax((max(nchar(vis_dn$nodes$id)) - nchar(vis_dn$nodes$id)) / 2, 0)
        vis_dn$nodes$label <- sprintf(
          "%-*s%s%*s", f, "", vis_dn$nodes$id, ceiling(f), ""
        )
        vis_dn$nodes$font.size <- vis_dn$nodes$size

        title <- vector(length = length(vis_dn$nodes$id))
        for (j in seq_along(title)) {
          net_data_sub <- net_data_dn[
            net_data_dn$symbol == vis_dn$nodes$id[j],
          ]
          title[j] <- paste(net_data_sub$gene_set, collapse = "<br>")
        }
        vis_dn$nodes$title <- title

        vis_dn <- visNetwork(vis_dn$nodes, vis_dn$edges, background = "black")
        vis_dn <- visPhysics(vis_dn, stabilization = FALSE)
        vis_dn <- visEdges(
          vis_dn, smooth = FALSE, color = list(opacity = 0.8, inherit = "both")
        )
        vis_dn <- visIgraphLayout(vis_dn, "layout.norm", layoutMatrix = layout)
        vis_dn <- visOptions(
          vis_dn,
          highlightNearest = list(
            enabled = TRUE, degree = 1, labelOnly = FALSE
          ),
          nodesIdSelection = TRUE
        )
        vis_dn <- visNodes(
          vis_dn, color = list(background = "rgba(0, 0, 0, 0)"),
          font = list(color = "white")
        )
      }

      output$vis_dn23 <- renderVisNetwork({vis_dn})
    })
  }
)
```

### LCM Mass Spec - UP

```{r}
shinyApp(

  ui = fluidPage(
    fluidRow(
      column(
        width = 2,
        sliderInput(
          paste0("conn_up", 2),
          label = "Connectivity:",
          min = 0,
          max = 5,
          value = 1,
          step = 0.1
        )
      )
    ),
    visNetworkOutput(paste0("vis_up", 2), height = "660px")
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("conn_up", 2)]]
    }, {

      de_gse <- rownames(res_gse[[2]][res_gse[[2]]$P.value.adj < 0.05, ])
      up_gse <- rownames(res_gse[[2]][res_gse[[2]]$Coef > 0, ])
      de_deg <- rownames(res_deg[[2]][res_deg[[2]]$P.value.adj < 0.05, ])
      up_deg <- rownames(res_deg[[2]][res_deg[[2]]$Coef > 0, ])

      net_data <- melt(geneIds(gene_sets[names(gene_sets) %in% de_gse]))
      names(net_data)[1] <- "ensembl"
      names(net_data)[2] <- "gene_set"

      gene_anno_sub <- gene_anno[gene_anno$ensembl %in% net_data$ensembl, ]
      net_data <- left_join(net_data, gene_anno_sub, by = "ensembl")
      net_data$gse_direction <- ifelse(
        net_data$gene_set %in% up_gse, yes = "up", no = "down"
      )
      net_data$gene_DE <- ifelse(
        net_data$ensembl %in% de_deg, yes = "yes", no = "no"
      )
      net_data$gene_direction <- ifelse(
        net_data$ensembl %in% up_deg, yes = "up", no = "down"
      )

      remove <- which(net_data$gse_direction != net_data$gene_direction)
      net_data <- net_data[-remove, ]
      net_data <- net_data[net_data$gene_DE == "yes", ]

      net_data_up <- net_data[which(net_data$gse_direction == "up"), ]
      hub_mat <- as.data.frame.matrix(table(net_data_up[ , c(3, 2)]))
      hub_mat <- hub_mat == 1
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )

      net_data_table <- table(net_data_up$symbol)
      gene_sets_table <- table(unlist(geneIds(gene_sets)))
      gene_anno_sub <- gene_anno[
        gene_anno$ensembl %in% names(gene_sets_table),
      ]
      gene_anno_sub <- gene_anno_sub[
        match(names(gene_sets_table), gene_anno_sub$ensembl),
      ]
      names(gene_sets_table) <- gene_anno_sub$symbol
      gene_sets_table <- gene_sets_table[
        names(gene_sets_table) %in% names(net_data_table)
      ]
      gene_sets_table <- gene_sets_table[
        match(names(net_data_table), names(gene_sets_table))
      ]

      net_data_table <- (net_data_table / gene_sets_table) * net_data_table

      remove <- net_data_table[
        net_data_table < input[[paste0("conn_up", 2)]]
      ]
      if (length(remove) > 0) {
        remove <- which(net_data_up$symbol %in% names(remove))
        net_data_up <- net_data_up[-remove, ]
        remove <- V(graph)$name %notin% net_data_up$symbol
        graph <- delete.vertices(graph, v = remove)
      }

      if (length(names(V(graph))) < 0) {
        return
      }

      graph <- subgraph(graph, which(degree(graph) > 0))

      clust <- cluster_edge_betweenness(graph)
      keep <- c()
      for (membership in seq_along(unique(clust$membership))) {
        sub <- clust$names[which(clust$membership == membership)]
        sub <- net_data_table[names(net_data_table) %in% sub]
        keep <- c(keep, names(sub[which.max(sub)]))
      }
      keep_idx <- which(V(graph)$name %in% keep)

      keep2 <- unique(unlist(neighborhood(graph, order = 1, nodes = keep_idx)))
      keep2 <- unique(c(keep_idx, keep2))
      graph <- induced_subgraph(graph, vids = keep2)

      edges <- get.edgelist(graph, names = FALSE)
      layout <- qgraph.layout.fruchtermanreingold(
        edges, vcount = vcount(graph), weights = E(graph)$weight,
        area = vcount(graph), repulse.rad = vcount(graph) / 4
      )
      vis_up <- toVisNetworkData(graph)

      net_data_up_nodup <- net_data_up[-which(duplicated(net_data_up$symbol)), ]
      net_data_up_nodup <- net_data_up_nodup[
        net_data_up_nodup$symbol %in% vis_up$nodes$id,
      ]
      net_data_up_nodup <- net_data_up_nodup[
        match(vis_up$nodes$id, net_data_up_nodup$symbol),
      ]
      vis_up$nodes$shape <- ifelse(
        net_data_up_nodup$symbol %in% keep, yes = "circle", no = "box"
      )

      net_data_table <- net_data_table[
        names(net_data_table) %in% vis_up$nodes$id
      ]
      vis_up$nodes$size <- (net_data_table + 5) ^ 1.8
      vis_up$edges$width <- (vis_up$edges$weight * 2) ^ 1.5

      net_res <- res_deg[[2]][res_deg[[2]]$Gene %in% vis_up$nodes$id, ]
      net_res <- net_res[match(vis_up$nodes$id, net_res$Gene), ]
      if (length(names(V(graph))) > 0) {
        col <- colorRampPalette(c("gray25", "red"))
        col <- col(100)[
          as.numeric(
            cut(
              rescale(log10(1 / net_res$P.value.adj), to = c(0, 1)),
              breaks = 100
            )
          )
        ]
        vis_up$nodes$color.border <- col
        vis_up$nodes$color.highlight <- col

        f <- pmax((max(nchar(vis_up$nodes$id)) - nchar(vis_up$nodes$id)) / 2, 0)
        vis_up$nodes$label <- sprintf(
          "%-*s%s%*s", f, "", vis_up$nodes$id, ceiling(f), ""
        )
        vis_up$nodes$font.size <- vis_up$nodes$size

        title <- vector(length = length(vis_up$nodes$id))
        for (j in seq_along(title)) {
          net_data_sub <- net_data_up[
            net_data_up$symbol == vis_up$nodes$id[j],
          ]
          title[j] <- paste(net_data_sub$gene_set, collapse = "<br>")
        }
        vis_up$nodes$title <- title

        vis_up <- visNetwork(vis_up$nodes, vis_up$edges, background = "black")
        vis_up <- visPhysics(vis_up, stabilization = FALSE)
        vis_up <- visEdges(
          vis_up, smooth = FALSE, color = list(opacity = 0.8, inherit = "both")
        )
        vis_up <- visIgraphLayout(vis_up, "layout.norm", layoutMatrix = layout)
        vis_up <- visOptions(
          vis_up,
          highlightNearest = list(
            enabled = TRUE, degree = 1, labelOnly = FALSE
          ),
          nodesIdSelection = TRUE
        )
        vis_up <- visNodes(
          vis_up, color = list(background = "rgba(0, 0, 0, 0)"),
          font = list(color = "white")
        )
      }

      output$vis_up2 <- renderVisNetwork({vis_up})
    })
  }
)
```

### LCM Mass Spec - DOWN

```{r}
shinyApp(

  ui = fluidPage(
    fluidRow(
      column(
        width = 2,
        sliderInput(
          paste0("conn_dn", 2),
          label = "Connectivity:",
          min = 0,
          max = 5,
          value = 1,
          step = 0.1
        )
      )
    ),
    visNetworkOutput(paste0("vis_dn", 2), height = "660px")
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("conn_dn", 2)]]
      # input[[paste0("init_dn", 2)]]
    }, {

      de_gse <- rownames(res_gse[[2]][res_gse[[2]]$P.value.adj < 0.05, ])
      up_gse <- rownames(res_gse[[2]][res_gse[[2]]$Coef > 0, ])
      de_deg <- rownames(res_deg[[2]][res_deg[[2]]$P.value.adj < 0.05, ])
      up_deg <- rownames(res_deg[[2]][res_deg[[2]]$Coef > 0, ])

      net_data <- melt(geneIds(gene_sets[names(gene_sets) %in% de_gse]))
      names(net_data)[1] <- "ensembl"
      names(net_data)[2] <- "gene_set"

      gene_anno_sub <- gene_anno[gene_anno$ensembl %in% net_data$ensembl, ]
      net_data <- left_join(net_data, gene_anno_sub, by = "ensembl")
      net_data$gse_direction <- ifelse(
        net_data$gene_set %in% up_gse, yes = "up", no = "down"
      )
      net_data$gene_DE <- ifelse(
        net_data$ensembl %in% de_deg, yes = "yes", no = "no"
      )
      net_data$gene_direction <- ifelse(
        net_data$ensembl %in% up_deg, yes = "up", no = "down"
      )

      remove <- which(net_data$gse_direction != net_data$gene_direction)
      net_data <- net_data[-remove, ]
      net_data <- net_data[net_data$gene_DE == "yes", ]

      net_data_dn <- net_data[which(net_data$gse_direction == "down"), ]
      hub_mat <- as.data.frame.matrix(table(net_data_dn[ , c(3, 2)]))
      hub_mat <- hub_mat == 1
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )

      net_data_table <- table(net_data_dn$symbol)
      gene_sets_table <- table(unlist(geneIds(gene_sets)))
      gene_anno_sub <- gene_anno[
        gene_anno$ensembl %in% names(gene_sets_table),
      ]
      gene_anno_sub <- gene_anno_sub[
        match(names(gene_sets_table), gene_anno_sub$ensembl),
      ]
      names(gene_sets_table) <- gene_anno_sub$symbol
      gene_sets_table <- gene_sets_table[
        names(gene_sets_table) %in% names(net_data_table)
      ]
      gene_sets_table <- gene_sets_table[
        match(names(net_data_table), names(gene_sets_table))
      ]
      net_data_table <- (net_data_table / gene_sets_table) * net_data_table

      remove <- net_data_table[
        net_data_table < input[[paste0("conn_dn", 2)]]
      ]
      if (length(remove) > 0) {
        remove <- which(net_data_dn$symbol %in% names(remove))
        net_data_dn <- net_data_dn[-remove, ]
        remove <- V(graph)$name %notin% net_data_dn$symbol
        graph <- delete.vertices(graph, v = remove)
      }

      if (length(names(V(graph))) < 0) {
        return
      }

      graph <- subgraph(graph, which(degree(graph) > 0))

      clust <- cluster_edge_betweenness(graph)
      keep <- c()
      for (membership in seq_along(unique(clust$membership))) {
        sub <- clust$names[which(clust$membership == membership)]
        sub <- net_data_table[names(net_data_table) %in% sub]
        keep <- c(keep, names(sub[which.max(sub)]))
      }
      keep_idx <- which(V(graph)$name %in% keep)

      keep2 <- unique(unlist(neighborhood(graph, order = 1, nodes = keep_idx)))
      keep2 <- unique(c(keep_idx, keep2))
      graph <- induced_subgraph(graph, vids = keep2)

      edges <- get.edgelist(graph, names = FALSE)
      layout <- qgraph.layout.fruchtermanreingold(
        edges, vcount = vcount(graph), weights = E(graph)$weight,
        area = vcount(graph), repulse.rad = vcount(graph) / 4
      )
      vis_dn <- toVisNetworkData(graph)

      net_data_dn_nodup <- net_data_dn[-which(duplicated(net_data_dn$symbol)), ]
      net_data_dn_nodup <- net_data_dn_nodup[
        net_data_dn_nodup$symbol %in% vis_dn$nodes$id,
      ]
      net_data_dn_nodup <- net_data_dn_nodup[
        match(vis_dn$nodes$id, net_data_dn_nodup$symbol),
      ]
      vis_dn$nodes$shape <- ifelse(
        net_data_dn_nodup$symbol %in% keep, yes = "circle", no = "box"
      )

      net_data_table <- net_data_table[
        names(net_data_table) %in% vis_dn$nodes$id
      ]
      vis_dn$nodes$size <- (net_data_table + 5) ^ 1.8
      vis_dn$edges$width <- (vis_dn$edges$weight * 2) ^ 1.5

      net_res <- res_deg[[2]][res_deg[[2]]$Gene %in% vis_dn$nodes$id, ]
      net_res <- net_res[match(vis_dn$nodes$id, net_res$Gene), ]
      if (length(names(V(graph))) > 0) {
        col <- colorRampPalette(c("gray25", "blue"))
        col <- col(100)[
          as.numeric(
            cut(
              rescale(log10(1 / net_res$P.value.adj), to = c(0, 1)),
              breaks = 100
            )
          )
        ]
        vis_dn$nodes$color.border <- col
        vis_dn$nodes$color.highlight <- col

        f <- pmax((max(nchar(vis_dn$nodes$id)) - nchar(vis_dn$nodes$id)) / 2, 0)
        vis_dn$nodes$label <- sprintf(
          "%-*s%s%*s", f, "", vis_dn$nodes$id, ceiling(f), ""
        )
        vis_dn$nodes$font.size <- vis_dn$nodes$size

        title <- vector(length = length(vis_dn$nodes$id))
        for (j in seq_along(title)) {
          net_data_sub <- net_data_dn[
            net_data_dn$symbol == vis_dn$nodes$id[j],
          ]
          title[j] <- paste(net_data_sub$gene_set, collapse = "<br>")
        }
        vis_dn$nodes$title <- title

        vis_dn <- visNetwork(vis_dn$nodes, vis_dn$edges, background = "black")
        vis_dn <- visPhysics(vis_dn, stabilization = FALSE)
        vis_dn <- visEdges(
          vis_dn, smooth = FALSE, color = list(opacity = 0.8, inherit = "both")
        )
        vis_dn <- visIgraphLayout(vis_dn, "layout.norm", layoutMatrix = layout)
        vis_dn <- visOptions(
          vis_dn,
          highlightNearest = list(
            enabled = TRUE, degree = 1, labelOnly = FALSE
          ),
          nodesIdSelection = TRUE
        )
        vis_dn <- visNodes(
          vis_dn, color = list(background = "rgba(0, 0, 0, 0)"),
          font = list(color = "white")
        )
      }

      output$vis_dn2 <- renderVisNetwork({vis_dn})
    })
  }
)
```

### FACS ssRNAseq - UP

```{r}
shinyApp(

  ui = fluidPage(
    fluidRow(
      column(
        width = 2,
        sliderInput(
          paste0("conn_up", 3),
          label = "Connectivity:",
          min = 0,
          max = 5,
          value = 1,
          step = 0.1
        )
      )
    ),
    visNetworkOutput(paste0("vis_up", 3), height = "660px")
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("conn_up", 3)]]
      # input[[paste0("init_up", 3)]]
    }, {

      de_gse <- rownames(res_gse[[3]][res_gse[[3]]$P.value.adj < 0.05, ])
      up_gse <- rownames(res_gse[[3]][res_gse[[3]]$Coef > 0, ])
      de_deg <- rownames(res_deg[[3]][res_deg[[3]]$P.value.adj < 0.05, ])
      up_deg <- rownames(res_deg[[3]][res_deg[[3]]$Coef > 0, ])

      net_data <- melt(geneIds(gene_sets[names(gene_sets) %in% de_gse]))
      names(net_data)[1] <- "ensembl"
      names(net_data)[2] <- "gene_set"

      gene_anno_sub <- gene_anno[gene_anno$ensembl %in% net_data$ensembl, ]
      net_data <- left_join(net_data, gene_anno_sub, by = "ensembl")
      net_data$gse_direction <- ifelse(
        net_data$gene_set %in% up_gse, yes = "up", no = "down"
      )
      net_data$gene_DE <- ifelse(
        net_data$ensembl %in% de_deg, yes = "yes", no = "no"
      )
      net_data$gene_direction <- ifelse(
        net_data$ensembl %in% up_deg, yes = "up", no = "down"
      )

      remove <- which(net_data$gse_direction != net_data$gene_direction)
      net_data <- net_data[-remove, ]
      net_data <- net_data[net_data$gene_DE == "yes", ]

      net_data_up <- net_data[which(net_data$gse_direction == "up"), ]
      hub_mat <- as.data.frame.matrix(table(net_data_up[ , c(3, 2)]))
      hub_mat <- hub_mat == 1
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )

      net_data_table <- table(net_data_up$symbol)
      gene_sets_table <- table(unlist(geneIds(gene_sets)))
      gene_anno_sub <- gene_anno[
        gene_anno$ensembl %in% names(gene_sets_table),
      ]
      gene_anno_sub <- gene_anno_sub[
        match(names(gene_sets_table), gene_anno_sub$ensembl),
      ]
      names(gene_sets_table) <- gene_anno_sub$symbol
      gene_sets_table <- gene_sets_table[
        names(gene_sets_table) %in% names(net_data_table)
      ]
      gene_sets_table <- gene_sets_table[
        match(names(net_data_table), names(gene_sets_table))
      ]

      net_data_table <- (net_data_table / gene_sets_table) * net_data_table

      remove <- net_data_table[
        net_data_table < input[[paste0("conn_up", 3)]]
      ]
      if (length(remove) > 0) {
        remove <- which(net_data_up$symbol %in% names(remove))
        net_data_up <- net_data_up[-remove, ]
        remove <- V(graph)$name %notin% net_data_up$symbol
        graph <- delete.vertices(graph, v = remove)
      }

      if (length(names(V(graph))) < 0) {
        return
      }

      graph <- subgraph(graph, which(degree(graph) > 0))

      clust <- cluster_edge_betweenness(graph)
      keep <- c()
      for (membership in seq_along(unique(clust$membership))) {
        sub <- clust$names[which(clust$membership == membership)]
        sub <- net_data_table[names(net_data_table) %in% sub]
        keep <- c(keep, names(sub[which.max(sub)]))
      }
      keep_idx <- which(V(graph)$name %in% keep)

      keep2 <- unique(unlist(neighborhood(graph, order = 1, nodes = keep_idx)))
      keep2 <- unique(c(keep_idx, keep2))
      graph <- induced_subgraph(graph, vids = keep2)

      edges <- get.edgelist(graph, names = FALSE)
      layout <- qgraph.layout.fruchtermanreingold(
        edges, vcount = vcount(graph), weights = E(graph)$weight,
        area = vcount(graph), repulse.rad = vcount(graph) / 4
      )
      vis_up <- toVisNetworkData(graph)

      net_data_up_nodup <- net_data_up[-which(duplicated(net_data_up$symbol)), ]
      net_data_up_nodup <- net_data_up_nodup[
        net_data_up_nodup$symbol %in% vis_up$nodes$id,
      ]
      net_data_up_nodup <- net_data_up_nodup[
        match(vis_up$nodes$id, net_data_up_nodup$symbol),
      ]
      vis_up$nodes$shape <- ifelse(
        net_data_up_nodup$symbol %in% keep, yes = "circle", no = "box"
      )

      net_data_table <- net_data_table[
        names(net_data_table) %in% vis_up$nodes$id
      ]
      vis_up$nodes$size <- (net_data_table + 5) ^ 1.8
      vis_up$edges$width <- (vis_up$edges$weight * 2) ^ 1.5

      net_res <- res_deg[[3]][res_deg[[3]]$Gene %in% vis_up$nodes$id, ]
      net_res <- net_res[match(vis_up$nodes$id, net_res$Gene), ]
      if (length(names(V(graph))) > 0) {
        col <- colorRampPalette(c("gray25", "red"))
        col <- col(100)[
          as.numeric(
            cut(
              rescale(log10(1 / net_res$P.value.adj), to = c(0, 1)),
              breaks = 100
            )
          )
        ]
        vis_up$nodes$color.border <- col
        vis_up$nodes$color.highlight <- col

        f <- pmax((max(nchar(vis_up$nodes$id)) - nchar(vis_up$nodes$id)) / 2, 0)
        vis_up$nodes$label <- sprintf(
          "%-*s%s%*s", f, "", vis_up$nodes$id, ceiling(f), ""
        )
        vis_up$nodes$font.size <- vis_up$nodes$size

        title <- vector(length = length(vis_up$nodes$id))
        for (j in seq_along(title)) {
          net_data_sub <- net_data_up[
            net_data_up$symbol == vis_up$nodes$id[j],
          ]
          title[j] <- paste(net_data_sub$gene_set, collapse = "<br>")
        }
        vis_up$nodes$title <- title

        vis_up <- visNetwork(vis_up$nodes, vis_up$edges, background = "black")
        vis_up <- visPhysics(vis_up, stabilization = FALSE)
        vis_up <- visEdges(
          vis_up, smooth = FALSE, color = list(opacity = 0.8, inherit = "both")
        )
        vis_up <- visIgraphLayout(vis_up, "layout.norm", layoutMatrix = layout)
        vis_up <- visOptions(
          vis_up,
          highlightNearest = list(
            enabled = TRUE, degree = 1, labelOnly = FALSE
          ),
          nodesIdSelection = TRUE
        )
        vis_up <- visNodes(
          vis_up, color = list(background = "rgba(0, 0, 0, 0)"),
          font = list(color = "white")
        )
      }

      output$vis_up3 <- renderVisNetwork({vis_up})
    })
  }
)
```

### FACS ssRNAseq - DOWN

```{r}
shinyApp(

  ui = fluidPage(
    fluidRow(
      column(
        width = 2,
        sliderInput(
          paste0("conn_dn", 3),
          label = "Connectivity:",
          min = 0,
          max = 5,
          value = 1,
          step = 0.1
        )
      )
    ),
    visNetworkOutput(paste0("vis_dn", 3), height = "660px")
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("conn_dn", 3)]]
      # input[[paste0("init_dn", 3)]]
    }, {

      de_gse <- rownames(res_gse[[3]][res_gse[[3]]$P.value.adj < 0.05, ])
      up_gse <- rownames(res_gse[[3]][res_gse[[3]]$Coef > 0, ])
      de_deg <- rownames(res_deg[[3]][res_deg[[3]]$P.value.adj < 0.05, ])
      up_deg <- rownames(res_deg[[3]][res_deg[[3]]$Coef > 0, ])

      net_data <- melt(geneIds(gene_sets[names(gene_sets) %in% de_gse]))
      names(net_data)[1] <- "ensembl"
      names(net_data)[2] <- "gene_set"

      gene_anno_sub <- gene_anno[gene_anno$ensembl %in% net_data$ensembl, ]
      net_data <- left_join(net_data, gene_anno_sub, by = "ensembl")
      net_data$gse_direction <- ifelse(
        net_data$gene_set %in% up_gse, yes = "up", no = "down"
      )
      net_data$gene_DE <- ifelse(
        net_data$ensembl %in% de_deg, yes = "yes", no = "no"
      )
      net_data$gene_direction <- ifelse(
        net_data$ensembl %in% up_deg, yes = "up", no = "down"
      )

      remove <- which(net_data$gse_direction != net_data$gene_direction)
      net_data <- net_data[-remove, ]
      net_data <- net_data[net_data$gene_DE == "yes", ]

      net_data_dn <- net_data[which(net_data$gse_direction == "down"), ]
      hub_mat <- as.data.frame.matrix(table(net_data_dn[ , c(3, 2)]))
      hub_mat <- hub_mat == 1
      hub_mat <- hub_mat %*% t(hub_mat)
      diag(hub_mat) <- 0
      graph <- graph_from_adjacency_matrix(
        hub_mat, mode = "upper", weighted = TRUE
      )

      net_data_table <- table(net_data_dn$symbol)
      gene_sets_table <- table(unlist(geneIds(gene_sets)))
      gene_anno_sub <- gene_anno[
        gene_anno$ensembl %in% names(gene_sets_table),
      ]
      gene_anno_sub <- gene_anno_sub[
        match(names(gene_sets_table), gene_anno_sub$ensembl),
      ]
      names(gene_sets_table) <- gene_anno_sub$symbol
      gene_sets_table <- gene_sets_table[
        names(gene_sets_table) %in% names(net_data_table)
      ]
      gene_sets_table <- gene_sets_table[
        match(names(net_data_table), names(gene_sets_table))
      ]
      net_data_table <- (net_data_table / gene_sets_table) * net_data_table

      remove <- net_data_table[
        net_data_table < input[[paste0("conn_dn", 3)]]
      ]
      if (length(remove) > 0) {
        remove <- which(net_data_dn$symbol %in% names(remove))
        net_data_dn <- net_data_dn[-remove, ]
        remove <- V(graph)$name %notin% net_data_dn$symbol
        graph <- delete.vertices(graph, v = remove)
      }

      if (length(names(V(graph))) < 0) {
        return
      }

      graph <- subgraph(graph, which(degree(graph) > 0))

      clust <- cluster_edge_betweenness(graph)
      keep <- c()
      for (membership in seq_along(unique(clust$membership))) {
        sub <- clust$names[which(clust$membership == membership)]
        sub <- net_data_table[names(net_data_table) %in% sub]
        keep <- c(keep, names(sub[which.max(sub)]))
      }
      keep_idx <- which(V(graph)$name %in% keep)

      keep2 <- unique(unlist(neighborhood(graph, order = 1, nodes = keep_idx)))
      keep2 <- unique(c(keep_idx, keep2))
      graph <- induced_subgraph(graph, vids = keep2)

      edges <- get.edgelist(graph, names = FALSE)
      layout <- qgraph.layout.fruchtermanreingold(
        edges, vcount = vcount(graph), weights = E(graph)$weight,
        area = vcount(graph), repulse.rad = vcount(graph) / 4
      )
      vis_dn <- toVisNetworkData(graph)

      net_data_dn_nodup <- net_data_dn[-which(duplicated(net_data_dn$symbol)), ]
      net_data_dn_nodup <- net_data_dn_nodup[
        net_data_dn_nodup$symbol %in% vis_dn$nodes$id,
      ]
      net_data_dn_nodup <- net_data_dn_nodup[
        match(vis_dn$nodes$id, net_data_dn_nodup$symbol),
      ]
      vis_dn$nodes$shape <- ifelse(
        net_data_dn_nodup$symbol %in% keep, yes = "circle", no = "box"
      )

      net_data_table <- net_data_table[
        names(net_data_table) %in% vis_dn$nodes$id
      ]
      vis_dn$nodes$size <- (net_data_table + 5) ^ 1.8
      vis_dn$edges$width <- (vis_dn$edges$weight * 2) ^ 1.5

      net_res <- res_deg[[3]][res_deg[[3]]$Gene %in% vis_dn$nodes$id, ]
      net_res <- net_res[match(vis_dn$nodes$id, net_res$Gene), ]
      if (length(names(V(graph))) > 0) {
        col <- colorRampPalette(c("gray25", "blue"))
        col <- col(100)[
          as.numeric(
            cut(
              rescale(log10(1 / net_res$P.value.adj), to = c(0, 1)),
              breaks = 100
            )
          )
        ]
        vis_dn$nodes$color.border <- col
        vis_dn$nodes$color.highlight <- col

        f <- pmax((max(nchar(vis_dn$nodes$id)) - nchar(vis_dn$nodes$id)) / 2, 0)
        vis_dn$nodes$label <- sprintf(
          "%-*s%s%*s", f, "", vis_dn$nodes$id, ceiling(f), ""
        )
        vis_dn$nodes$font.size <- vis_dn$nodes$size

        title <- vector(length = length(vis_dn$nodes$id))
        for (j in seq_along(title)) {
          net_data_sub <- net_data_dn[
            net_data_dn$symbol == vis_dn$nodes$id[j],
          ]
          title[j] <- paste(net_data_sub$gene_set, collapse = "<br>")
        }
        vis_dn$nodes$title <- title

        vis_dn <- visNetwork(vis_dn$nodes, vis_dn$edges, background = "black")
        vis_dn <- visPhysics(vis_dn, stabilization = FALSE)
        vis_dn <- visEdges(
          vis_dn, smooth = FALSE, color = list(opacity = 0.8, inherit = "both")
        )
        vis_dn <- visIgraphLayout(vis_dn, "layout.norm", layoutMatrix = layout)
        vis_dn <- visOptions(
          vis_dn,
          highlightNearest = list(
            enabled = TRUE, degree = 1, labelOnly = FALSE
          ),
          nodesIdSelection = TRUE
        )
        vis_dn <- visNodes(
          vis_dn, color = list(background = "rgba(0, 0, 0, 0)"),
          font = list(color = "white")
        )
      }

      output$vis_dn3 <- renderVisNetwork({vis_dn})
    })
  }
)
```

Gene/Protein Analysis
===

Row {.tabset}
---

### Search Genes/Proteins

```{r, include = FALSE}
dev.off()
dev.off()
```

```{r}
shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("deg", "_deg"),
      label = "Type to Search Genes/Proteins:",
      choices = unique(c(top_deg[[2]], top_deg[[3]])),
      selected = intersect(top_deg[[2]], top_deg[[3]])[1:25],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 2,
        radioButtons(
          paste0("scaling", "_deg"), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      )
    ),
    fluidPage(
      div(
        style = "display:flex;",
        plotOutput("heatmap2", height = "660px"),
        plotOutput("heatmap3", height = "660px")
      )
    )
  ),

  server = function (input, output, session) {
    observeEvent({
      input[[paste0("deg", "_deg")]]
      input[[paste0("scaling", "_deg")]]
    }, {

      features2 <- input[[paste0("deg", "_deg")]]
      sig2 <- sig_deg[[2]][
        rownames(sig_deg[[2]]) %in% input[[paste0("deg", "_deg")]], ,
        drop = FALSE
      ]
      features2 <- features2[features2 %in% rownames(sig2)]
      if (nrow(sig2) > 1) {
        sig2 <- sig2[match(features2, rownames(sig2)), , drop = FALSE]
      }

      gene_anno_sub2 <- gene_anno[gene_anno$symbol %in% features2, ]
      gene_anno_sub2 <- gene_anno_sub2[
        match(features2, gene_anno_sub2$symbol),
      ]

      mat2 <- deg[[2]][
        rownames(deg[[2]]) %in% gene_anno_sub2$ensembl, , drop = FALSE
      ]
      if (nrow(mat2) > 1) {
        mat2 <- mat2[
          match(gene_anno_sub2$ensembl, rownames(mat2)), , drop = FALSE
        ]
      }
      rownames(mat2) <- gene_anno_sub2$symbol

      gene_anno_sub2 <- gene_anno[gene_anno$ensembl %in% rownames(deg[[2]]), ]
      add_anno2 <- input[[paste0("deg", "_deg")]][
        input[[paste0("deg", "_deg")]] %notin% gene_anno_sub2$symbol
      ]
      add2 <- matrix(NA, nrow = length(add_anno2), ncol = ncol(deg[[2]]))
      rownames(add2) <- add_anno2
      mat2 <- rbind(mat2, add2)
      sig2 <- rbind(sig2, add2)
      sig2[is.na(sig2)] <- ""
      mat2 <- mat2[match(input[[paste0("deg", "_deg")]], rownames(mat2)), ]
      sig2 <- sig2[match(input[[paste0("deg", "_deg")]], rownames(sig2)), ]

      if (input[[paste0("scaling", "_deg")]] == "Scaled") {
        mat2 <- t(
          apply(
            mat2,
            MARGIN = 1,
            FUN = function (x) {
              (
                (
                  2 * (x - min(x, na.rm = TRUE)) /
                    (max(x, na.rm = TRUE) - min(x, na.rm = TRUE))
                ) - 1
              )
            }
          )
        )
        col2 <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title2 <- paste0("Scaled ", data_types[2])
      } else {
        col2 <- colorRamp2(range_deg[[2]], colors = c("blue", "white", "red"))
        title2 <- data_types[2]
      }

      output$heatmap2 <- renderPlot({
        draw(
          Heatmap(
            mat2,
            col = col2,
            na_col = "darkgrey",
            column_title = paste0(
              data_sets[2], "\n",
              "Significant Proteins (Adj. P < 0.05)"
            ),
            cluster_columns = FALSE,
            cluster_rows = FALSE,
            column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
            column_gap = unit(1, units = "mm"),
            rect_gp = gpar(col = "black", lwd = 0.25),
            row_names_gp = gpar(fontsize = 13),
            column_title_gp = gpar(fontsize = 14),
            row_title_rot = 0,
            top_annotation = top_anno[[2]],
            show_column_names = FALSE,
            cell_fun = function(j, i, x, y, width, height, fill) {
              grid.text(sig2[i, j], x = x, y = y, gp = gpar(fontsize = 11))
            },
            heatmap_legend_param = list(
              title = title2, direction = "horizontal",
              title_position = "topcenter", labels_gp = gpar(fontsize = 10),
              grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
            )
          ),
          heatmap_legend_side = "top"
        )
      }, res = 96)

      features3 <- input[[paste0("deg", "_deg")]]
      sig3 <- sig_deg[[3]][
        rownames(sig_deg[[3]]) %in% input[[paste0("deg", "_deg")]], ,
        drop = FALSE
      ]
      features3 <- features3[features3 %in% rownames(sig3)]
      if (nrow(sig3) > 1) {
        sig3 <- sig3[match(features3, rownames(sig3)), , drop = FALSE]
      }

      gene_anno_sub3 <- gene_anno[
        gene_anno$symbol %in% features3,
      ]
      gene_anno_sub3 <- gene_anno_sub3[
        match(features3, gene_anno_sub3$symbol),
      ]

      mat3 <- deg[[3]][
        rownames(deg[[3]]) %in% gene_anno_sub3$ensembl, , drop = FALSE
      ]
      if (nrow(mat3) > 1) {
        mat3 <- mat3[
          match(gene_anno_sub3$ensembl, rownames(mat3)), , drop = FALSE
        ]
      }
      rownames(mat3) <- gene_anno_sub3$symbol

      gene_anno_sub3 <- gene_anno[
        gene_anno$symbol %in% input[[paste0("deg", "_deg")]],
      ]
      gene_anno_sub3 <- gene_anno_sub3[
        match(input[[paste0("deg", "_deg")]], gene_anno_sub3$symbol),
      ]

      gene_anno_sub3 <- gene_anno[gene_anno$ensembl %in% rownames(deg[[3]]), ]
      add_anno3 <- input[[paste0("deg", "_deg")]][
        input[[paste0("deg", "_deg")]] %notin% gene_anno_sub3$symbol
      ]
      add3 <- matrix(NA, nrow = length(add_anno3), ncol = ncol(deg[[3]]))
      rownames(add3) <- add_anno3
      mat3 <- rbind(mat3, add3)
      sig3 <- rbind(sig3, add3)
      sig3[is.na(sig3)] <- ""
      mat3 <- mat3[match(input[[paste0("deg", "_deg")]], rownames(mat3)), ]
      sig3 <- sig3[match(input[[paste0("deg", "_deg")]], rownames(sig3)), ]

      if (input[[paste0("scaling", "_deg")]] == "Scaled") {
        mat3 <- t(
          apply(
            mat3,
            MARGIN = 1,
            FUN = function (x) {
              (
                (
                  2 * (x - min(x, na.rm = TRUE)) /
                    (max(x, na.rm = TRUE) - min(x, na.rm = TRUE))
                ) - 1
              )
            }
          )
        )
        col3 <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title3 <- paste0("Scaled ", data_types[3])
      } else {
        col3 <- colorRamp2(range_deg[[3]], colors = c("blue", "white", "red"))
        title3 <- data_types[3]
      }

      output$heatmap3 <- renderPlot({
        draw(
          Heatmap(
            mat3,
            col = col3,
            na_col = "darkgrey",
            column_title = paste0(
              data_sets[3], "\n",
              "Significant Genes (Adj. P < 0.05)"
            ),
            cluster_columns = FALSE,
            cluster_rows = FALSE,
            column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
            column_gap = unit(1, units = "mm"),
            rect_gp = gpar(col = "black", lwd = 0.25),
            row_names_gp = gpar(fontsize = 13),
            column_title_gp = gpar(fontsize = 14),
            row_title_rot = 0,
            top_annotation = top_anno[[3]],
            show_column_names = FALSE,
            cell_fun = function(j, i, x, y, width, height, fill) {
              grid.text(sig3[i, j], x = x, y = y, gp = gpar(fontsize = 11))
            },
            heatmap_legend_param = list(
              title = title3, direction = "horizontal",
              title_position = "topcenter", labels_gp = gpar(fontsize = 10),
              grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
            )
          ),
          heatmap_legend_side = "top"
        )
      }, res = 96)

    })
  }
)
```

### Gene Set Connectivity - LCM Mass Spec

```{r}
selectizeInput(
  paste0("deg", "_conn2"),
  label = "Type to Search Genes/Proteins:",
  choices = top_deg[[2]],
  selected = top_deg[[2]][1:5],
  multiple = TRUE,
  width = "100%"
)

renderSankeyNetwork({

  links_list2 <- vector(
    "list", length = length(input[[paste0("deg", "_conn2")]])
  )
  for (i in seq_along(input[[paste0("deg", "_conn2")]])) {
    gse_sub2 <- res_gse[[2]][res_gse[[2]]$P.value.adj < 0.05, ]
    deg_sub2 <- res_deg[[2]][
      res_deg[[2]]$Gene %in% input[[paste0("deg", "_conn2")]][i],
    ]$Coef
    if (deg_sub2 > 0) {
      gse_sub2 <- gse_sub2[gse_sub2$Coef > 0, ]
    } else {
      gse_sub2 <- gse_sub2[gse_sub2$Coef < 0, ]
    }
    gene_ids2 <- geneIds(gene_sets)[names(gene_sets) %in% gse_sub2$GeneSet]
    gene_ids2 <- gene_ids2[order(names(gene_ids2))]

    links_orig2 <- data.frame(
      source = rep(names(gene_ids2), times = lengths(gene_ids2)),
      target = gene_ids2 %>% map_dfr(as_tibble)
    )
    links2 <- links_orig2
    links_orig2[] <- map[unlist(links_orig2)]
    links2$value <- links_orig2$value
    colnames(links2) <- c("source", "target")
    links2 <- links2[
      links2$target %in% input[[paste0("deg", "_conn2")]][i],
    ]
    links2$value <- rep(1, times = nrow(links2))
    links_list2[[i]] <- links2
  }
  links_comb2 <- do.call(rbind, args = links_list2)

  nodes2 <- data.frame(
    name = c(
      as.character(links_comb2$source), as.character(links_comb2$target)
    ) %>% unique()
  )
  links_comb2$IDsource <- match(links_comb2$source, nodes2$name) - 1
  links_comb2$IDtarget <- match(links_comb2$target, nodes2$name) - 1

  if (nrow(nodes2) > 0) {
    select_length2 <- length(
      which(input[[paste0("deg", "_conn2")]] %in% nodes2$name)
    )
    nodes2$group <- c(
      rep("target", times = nrow(nodes2) - select_length2),
      nodes2$name[(nrow(nodes2) - (select_length2 - 1)):nrow(nodes2)]
    )
    colour2 <- c(hue_pal()(select_length2), "#D3D3D3")
    colour2 <- paste0(
      'd3.scaleOrdinal().range(["', paste(colour2, collapse = '", "'), '"])'
    )

    sankeyNetwork(
      Links = links_comb2, Nodes = nodes2, Source = "IDsource",
      Target = "IDtarget", Value = "value", NodeID = "name",
      sinksRight = FALSE, fontSize = 20, NodeGroup = "group",
      LinkGroup = "target", colourScale = colour2, margin = list(bottom = 100)
    )
  }

})
```

### Gene Set Connectivity - FACS ssRNAseq

```{r}
selectizeInput(
  paste0("deg", "_conn3"),
  label = "Type to Search Genes/Proteins:",
  choices = top_deg[[3]],
  selected = top_deg[[3]][1:5],
  multiple = TRUE,
  width = "100%"
)

renderSankeyNetwork({

  links_list3 <- vector(
    "list", length = length(input[[paste0("deg", "_conn3")]])
  )
  for (i in seq_along(input[[paste0("deg", "_conn3")]])) {
    gse_sub3 <- res_gse[[3]][res_gse[[3]]$P.value.adj < 0.05, ]
    deg_sub3 <- res_deg[[3]][
      res_deg[[3]]$Gene %in% input[[paste0("deg", "_conn3")]][i],
    ]$Coef
    if (deg_sub3 > 0) {
      gse_sub3 <- gse_sub3[gse_sub3$Coef > 0, ]
    } else {
      gse_sub3 <- gse_sub3[gse_sub3$Coef < 0, ]
    }
    gene_ids3 <- geneIds(gene_sets)[names(gene_sets) %in% gse_sub3$GeneSet]
    gene_ids3 <- gene_ids3[order(names(gene_ids3))]

    links_orig3 <- data.frame(
      source = rep(names(gene_ids3), times = lengths(gene_ids3)),
      target = gene_ids3 %>% map_dfr(as_tibble)
    )
    links3 <- links_orig3
    links_orig3[] <- map[unlist(links_orig3)]
    links3$value <- links_orig3$value
    colnames(links3) <- c("source", "target")
    links3 <- links3[
      links3$target %in% input[[paste0("deg", "_conn3")]][i],
    ]
    links3$value <- rep(1, times = nrow(links3))
    links_list3[[i]] <- links3
  }
  links_comb3 <- do.call(rbind, args = links_list3)

  nodes3 <- data.frame(
    name = c(
      as.character(links_comb3$source), as.character(links_comb3$target)
    ) %>% unique()
  )
  links_comb3$IDsource <- match(links_comb3$source, nodes3$name) - 1
  links_comb3$IDtarget <- match(links_comb3$target, nodes3$name) - 1

  if (nrow(nodes3) > 0) {
    select_length3 <- length(
      which(input[[paste0("deg", "_conn3")]] %in% nodes3$name)
    )
    nodes3$group <- c(
      rep("target", times = nrow(nodes3) - select_length3),
      nodes3$name[(nrow(nodes3) - (select_length3 - 1)):nrow(nodes3)]
    )
    colour3 <- c(hue_pal()(select_length3), "#D3D3D3")
    colour3 <- paste0(
      'd3.scaleOrdinal().range(["', paste(colour3, collapse = '", "'), '"])'
    )

    sankeyNetwork(
      Links = links_comb3, Nodes = nodes3, Source = "IDsource",
      Target = "IDtarget", Value = "value", NodeID = "name",
      sinksRight = FALSE, fontSize = 20, NodeGroup = "group",
      LinkGroup = "target", colourScale = colour3, margin = list(bottom = 100)
    )
  }

})
```

```{r, include = FALSE}
rm(
  deg, edges, gene_anno, gene_anno_sub, gene_sets, graph, gse, hub_mat, layout,
  meta, n_deg, n_gse, net_data, net_data_dn, net_data_dn_nodup, net_data_sub,
  net_data_up, net_data_up_nodup, net_res, p_adj, range_deg, range_gse, res_deg,
  res_gse, sig_deg, sig_gse, top_anno, top_deg, top_gse, vis_dn, vis_up, clust,
  net_data2, net_data3, mat, mat2, mat3, sig, sig2, sig3, add, add2, add3,
  add_anno, add_anno2, add_anno3, gene_anno_sub2, gene_anno_sub3, gse_sub2,
  gene_ids2, links2, links_orig2, links_comb2, links_list2, nodes2, gse_sub3,
  gene_ids3, links3, links_orig3, links_comb3, links_list3, nodes3, deg_sub2,
  deg_sub3
)
gc()
```
