---
title: "Tangle Bearing Neurons Viewer"
output:
  flexdashboard::flex_dashboard:
runtime: shiny
---

```{r global, include = FALSE}
#    This file is part of tangle-bearing-neurons.
#    Copyright (C) 2024  Emir Turkes, UK DRI at UCL
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#    Emir Turkes can be contacted at emir.turkes@eturkes.com

library(conflicted)
library(scales)
library(InteractiveComplexHeatmap)
library(circlize)
library(GSEABase)

`%notin%` <- Negate(`%in%`)

pdf(NULL)

data_sets <- c(
  "LCM Mass Spec - Human AD BA9 - Single-cell",
  "LCM Mass Spec - Human AD BA9 - Pooled-cells",
  "FACS-sorted ssRNAseq - Human AD BA9"
)
data_types <- c(
  "log2 LFQ Intensity", "log2 LFQ Intensity", "log2 CPM",
  "log2 Enrichment Score", "log2 Enrichment Score", "log2 Enrichment Score"
)

data_dir <- file.path("..", "data")
cache_dir <- file.path("..", "cache")
if (!dir.exists(cache_dir)) {
  dir.create(cache_dir, recursive = TRUE)
}

gene_sets <- readRDS(file.path(data_dir, "gene_sets.rds"))
gene_anno <- readRDS(file.path(data_dir, "gene_anno.rds"))

meta <- vector("list", length = length(data_sets))
deg <- vector("list", length = length(data_sets))
gse <- vector("list", length = length(data_sets))
res_deg <- vector("list", length = length(data_sets))
res_gse <- vector("list", length = length(data_sets))

data_dir <- list.dirs(data_dir)
data_dir <- data_dir[c(2, 3, 4)]
for (i in seq_along(data_sets)) {
  meta[[i]] <- readRDS(file.path(data_dir[i], "meta.rds"))
  deg[[i]] <- readRDS(file.path(data_dir[i], "deg.rds"))
  gse[[i]] <- readRDS(file.path(data_dir[i], "gse.rds"))
  res_deg[[i]] <- readRDS(file.path(data_dir[i], "res_deg.rds"))
  res_gse[[i]] <- readRDS(file.path(data_dir[i], "res_gse.rds"))
}

top_anno <- vector("list", length = length(data_sets))
for (i in seq_along(data_sets)) {
  n <- length(unique(meta[[i]]$donor))
  anno_colour <- hue_pal()(n)
  top_anno[[i]] <- HeatmapAnnotation(
    Donor = meta[[i]]$donor,
    group = anno_block(
      gp = gpar(fill = "lightgray", col = "lightgray"),
      labels = unique(meta[[i]]$tau),
      labels_gp = gpar(col = "black", fontsize = 14),
      height = unit(6, "mm")
    ),
    col = list(Donor = setNames(anno_colour, unique(meta[[i]]$donor))),
    simple_anno_size = unit(4, "mm"),
    show_legend = c("Donor" = FALSE),
    annotation_name_gp = gpar(fontsize = 12)
  )
}

p_adj <- p.adjust(
  unlist(c(lapply(res_deg, `[[`, "P.value"), lapply(res_gse, `[[`, "P.value"))),
  method = "BH"
)
p_adj <- split(
  p_adj,
  f = rep(
    seq_along(c(res_deg, res_gse)),
    times = sapply(c(res_deg, res_gse), FUN = nrow)
  )
)
for (i in seq_along(res_deg)) {
  res_deg[[i]]$P.value.adj <- p_adj[[i]]
}
for (i in seq_along(res_gse)) {
  res_gse[[i]]$P.value.adj <- p_adj[[i + length(data_sets)]]
}

priority_gse <- c(
  `Tau & Neurofibrillary Tangle` =
    "tau-protein | tau protein | neurofi | inclusion | aggrep",
  `Microtubule & Cytoskeletal` =
    "microtubule | dynein | cytoskele | laminin",
  `Death, Stress, & Ageing` =
    "death | necro | senescen | stress",
  `Protein Folding` =
    "folding | folded",
  `Synapse, Axon, & Dendrite` =
    "synap | axon | dendrit",
  `Exocytosis & Endocytosis` =
    "exocyt | endocyt",
  `Apoptosis & Autophagy`=
    "apopto | autophag | phagocyt",
  `Proteasomal & Lysosomal`=
    "proteasom | lysosom | endoplasmic | Golgi",
  `Glycosaminoglycan`=
    "heparan | chondroitin | sulfotransferase | glycosaminoglycan",
  `Transcription & Translation` =
    "transcription | translation",
  `Amyloid & Lipoprotein` =
    "amyloid | lipoprotein",
  `Lipid Metabolism`=
    "fatty | cholesterol | insulin",
  `Synaptic Pruning` =
    "pruning | phosphatidylserine | complement",
  `T Cell, B Cell, & Other Immune Cell`=
    "T cell | B cell | leukocyte | myeloid | lymphocyte",
  `Inflammatory Response` =
    "immun | inflammat | cortisol | interleukin | NF-kappaB | toll-like",
  `Macrophage & Astrocyte` =
    "macrophage | astrocyte",
  `Oligodendrocyte & Myelin` =
    "oligodendrocyte | Schwann | myelin"
)

if (file.exists(file.path(cache_dir, "range_gse.rds"))) {
  range_gse <- readRDS(file.path(cache_dir, "range_gse.rds"))
  sig_gse <- readRDS(file.path(cache_dir, "sig_gse.rds"))
  top_gse <- readRDS(file.path(cache_dir, "top_gse.rds"))
  n_gse <- readRDS(file.path(cache_dir, "n_gse.rds"))
  range_deg <- readRDS(file.path(cache_dir, "range_deg.rds"))
  sig_deg <- readRDS(file.path(cache_dir, "sig_deg.rds"))
  top_deg <- readRDS(file.path(cache_dir, "top_deg.rds"))
  n_deg <- readRDS(file.path(cache_dir, "n_deg.rds"))
} else {

  ref <- sapply(
    priority_gse,
    function (x) {
      patterns <- unlist(strsplit(x, " \\| "))
      matches <- sapply(
        patterns, function(y) {
          grep(paste0(".*", y, ".*"), names(gene_sets), value = TRUE)
        }
      )
      unique(unlist(matches))
    }
  )

  range_gse <- vector("list", length = length(data_sets))
  sig_gse <- vector("list", length = length(data_sets))
  top_gse <- vector("list", length = length(data_sets))
  n_gse <- vector("list", length = length(data_sets))

  for (i in seq_along(data_sets)) {

    range_gse[[i]] <- c(
      min(gse[[i]]), (min(gse[[i]]) + max(gse[[i]])) / 2, max(gse[[i]])
    )

    sig <- ifelse(res_gse[[i]]$P.value.adj < 0.05, yes = 1, no = 0)
    direction <- ifelse(res_gse[[i]]$Coef > 0, yes = 1, no = -1)
    sig <- sig * direction

    names(sig) <- rownames(res_gse[[i]])
    sig <- sig[c(which(sig == 1), which(sig == -1), which(sig == 0))]
    sig <- ifelse(
      sig == -1, yes = "↓", no = ifelse(sig == 1, yes = "↑", no = "")
    )

    sig <- data.frame(CTRL = rep("", times = length(sig)), NFT = sig)
    sig_gse[[i]] <- vector("list", length = 2)
    for (j in seq_along(sig)) {
      sig_gse[[i]][[j]] <- replicate(nrow(meta[[i]]) / 2, expr = sig[ , j])
    }
    sig_gse[[i]] <- do.call(cbind, args = sig_gse[[i]])

    res_gse[[i]] <- res_gse[[i]][match(rownames(sig), rownames(res_gse[[i]])), ]
    rownames(sig_gse[[i]]) <- res_gse[[i]]$GeneSet

    top_gse[[i]] <- sapply(
      priority_gse,
      function (x) {
        patterns <- unlist(strsplit(x, " \\| "))
        matches <- sapply(
          patterns, function(y) {
            grep(
              paste0(".*", y, ".*"),
              rownames(res_gse[[i]][res_gse[[i]]$P.value.adj < 0.05, ]),
              value = TRUE
            )
          }
        )
        unique(unlist(matches))
      }
    )

    n_gse[[i]] <- Map(
      function(x, y) {
        ratio <- round(length(x) / length(y), digits = 2) * 100
        ifelse(is.na(ratio), yes = 0, no = ratio)
      },
      top_gse[[i]], ref
    )

  }

  range_deg <- vector("list", length = length(data_sets))
  sig_deg <- vector("list", length = length(data_sets))
  top_deg <- vector("list", length = length(data_sets))
  n_deg <- vector("list", length = length(data_sets))

  for (i in seq_along(data_sets)) {

    range_deg[[i]] <- c(
      min(deg[[i]]), (min(deg[[i]]) + max(deg[[i]])) / 2, max(deg[[i]])
    )

    sig <- ifelse(res_deg[[i]]$P.value.adj < 0.05, yes = 1, no = 0)
    direction <- ifelse(res_deg[[i]]$Coef > 0, yes = 1, no = -1)
    sig <- sig * direction

    names(sig) <- rownames(res_deg[[i]])
    sig <- sig[c(which(sig == 1), which(sig == -1), which(sig == 0))]
    sig <- ifelse(
      sig == -1, yes = "↓", no = ifelse(sig == 1, yes = "↑", no = "")
    )

    sig <- data.frame(CTRL = rep("", times = length(sig)), NFT = sig)
    sig_deg[[i]] <- vector("list", length = 2)
    for (j in seq_along(sig)) {
      sig_deg[[i]][[j]] <- replicate(nrow(meta[[i]]) / 2, expr = sig[ , j])
    }
    sig_deg[[i]] <- do.call(cbind, args = sig_deg[[i]])

    res_deg[[i]] <- res_deg[[i]][match(rownames(sig), rownames(res_deg[[i]])), ]
    rownames(sig_deg[[i]]) <- res_deg[[i]]$Gene

  }
  saveRDS(range_gse, file = file.path(cache_dir, "range_gse.rds"))
  saveRDS(sig_gse, file = file.path(cache_dir, "sig_gse.rds"))
  saveRDS(top_gse, file = file.path(cache_dir, "top_gse.rds"))
  saveRDS(n_gse, file = file.path(cache_dir, "n_gse.rds"))
  saveRDS(range_deg, file = file.path(cache_dir, "range_deg.rds"))
  saveRDS(sig_deg, file = file.path(cache_dir, "sig_deg.rds"))
  saveRDS(top_deg, file = file.path(cache_dir, "top_deg.rds"))
  saveRDS(n_deg, file = file.path(cache_dir, "n_deg.rds"))
}
```

Laser-capture Microdissection Mass Spec - Human AD BA9 - Single-cell
===

```{r}
idx <- 1
sub_idx <- 0
```

Row {.tabset}
---

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%",
    " Of Tested Pathways Significant)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_11_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_11_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = "Click row to See proteins. Drag bottom-right corner to resize.",
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 11)]]
      input[[paste0("scaling", 11)]]
      input[[paste0("init", 11)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 11)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 11)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 11)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[1], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 11)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 11), click_action = click_11_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 11), click_action = click_11_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_12_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_12_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 12)]]
      input[[paste0("scaling", 12)]]
      input[[paste0("init", 12)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 12)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 12)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 12)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[2], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 12)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 12), click_action = click_12_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 12), click_action = click_12_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_13_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_13_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 13)]]
      input[[paste0("scaling", 13)]]
      input[[paste0("init", 13)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 13)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 13)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 13)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[3], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 13)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 13), click_action = click_13_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 13), click_action = click_13_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_14_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_14_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 14)]]
      input[[paste0("scaling", 14)]]
      input[[paste0("init", 14)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 14)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 14)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 14)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[4], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 14)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 14), click_action = click_14_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 14), click_action = click_14_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_15_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_15_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 15)]]
      input[[paste0("scaling", 15)]]
      input[[paste0("init", 15)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 15)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 15)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 15)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[5], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 15)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 15), click_action = click_15_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 15), click_action = click_15_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_16_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_16_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 16)]]
      input[[paste0("scaling", 16)]]
      input[[paste0("init", 16)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 16)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 16)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 16)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[6], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 16)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 16), click_action = click_16_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 16), click_action = click_16_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_17_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_17_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 17)]]
      input[[paste0("scaling", 17)]]
      input[[paste0("init", 17)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 17)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 17)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 17)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[7], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 17)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 17), click_action = click_17_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 17), click_action = click_17_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_18_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_18_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 18)]]
      input[[paste0("scaling", 18)]]
      input[[paste0("init", 18)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 18)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 18)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 18)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[8], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 18)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 18), click_action = click_18_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 18), click_action = click_18_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_19_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_19_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 19)]]
      input[[paste0("scaling", 19)]]
      input[[paste0("init", 19)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 19)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 19)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 19)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[9], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 19)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 19), click_action = click_19_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 19), click_action = click_19_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_110_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_110_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 110)]]
      input[[paste0("scaling", 110)]]
      input[[paste0("init", 110)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 110)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 110)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 110)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[10], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 110)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 110), click_action = click_110_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 110), click_action = click_110_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_111_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[11]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_111_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[11]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 111)]]
      input[[paste0("scaling", 111)]]
      input[[paste0("init", 111)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 111)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 111)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 111)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[11], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 111)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 111), click_action = click_111_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 111), click_action = click_111_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_112_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[12]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_112_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[12]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 112)]]
      input[[paste0("scaling", 112)]]
      input[[paste0("init", 112)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 112)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 112)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 112)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[12], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 112)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 112), click_action = click_112_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 112), click_action = click_112_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_113_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[13]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_113_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[13]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 113)]]
      input[[paste0("scaling", 113)]]
      input[[paste0("init", 113)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 113)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 113)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 113)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[13], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 113)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 113), click_action = click_113_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 113), click_action = click_113_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_114_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[14]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_114_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[14]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 114)]]
      input[[paste0("scaling", 114)]]
      input[[paste0("init", 114)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 114)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 114)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 114)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[14], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 114)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 114), click_action = click_114_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 114), click_action = click_114_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_115_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[15]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_115_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[15]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 115)]]
      input[[paste0("scaling", 115)]]
      input[[paste0("init", 115)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 115)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 115)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 115)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[15], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 115)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 115), click_action = click_115_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 115), click_action = click_115_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_116_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[16]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_116_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[16]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 116)]]
      input[[paste0("scaling", 116)]]
      input[[paste0("init", 116)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 116)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 116)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 116)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[16], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 116)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 116), click_action = click_116_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 116), click_action = click_116_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_117_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[17]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[1]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[1], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_117_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[1]][[17]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[1]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[1]]),
  ]
  sig <- sig_deg[[1]][
    rownames(sig_deg[[1]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[1]][
    rownames(deg[[1]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[1]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[1]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[1]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 117)]]
      input[[paste0("scaling", 117)]]
      input[[paste0("init", 117)]]
    }, {

      sig <- sig_gse[[1]][
        rownames(sig_gse[[1]]) %in% input[[paste0("gse", 117)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 117)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[1]][rownames(gse[[1]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 117)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[1 + 3])
      } else {
        col <- colorRamp2(range_gse[[1]], colors = c("blue", "white", "red"))
        title <- data_types[1 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[1], "\n",
            "Significant ", names(priority_gse)[17], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[1]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[1]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 117)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 117), click_action = click_117_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 117), click_action = click_117_unscaled
        )
      }
    })
  },
)
```

Laser-capture Microdissection Mass Spec - Human AD BA9 - Pooled-cells
===

```{r}
idx <- 2
sub_idx <- 0
```

Row {.tabset}
---

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%",
    " Of Tested Pathways Significant)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_21_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_21_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = "Click row to See proteins. Drag bottom-right corner to resize.",
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 21)]]
      input[[paste0("scaling", 21)]]
      input[[paste0("init", 21)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 21)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 21)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 21)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[2], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 21)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 21), click_action = click_21_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 21), click_action = click_21_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_22_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_22_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 22)]]
      input[[paste0("scaling", 22)]]
      input[[paste0("init", 22)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 22)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 22)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 22)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[2], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 22)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 22), click_action = click_22_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 22), click_action = click_22_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_23_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_23_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 23)]]
      input[[paste0("scaling", 23)]]
      input[[paste0("init", 23)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 23)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 23)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 23)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[3], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 23)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 23), click_action = click_23_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 23), click_action = click_23_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_24_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_24_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 24)]]
      input[[paste0("scaling", 24)]]
      input[[paste0("init", 24)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 24)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 24)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 24)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[4], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 24)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 24), click_action = click_24_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 24), click_action = click_24_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_25_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_25_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 25)]]
      input[[paste0("scaling", 25)]]
      input[[paste0("init", 25)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 25)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 25)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 25)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[5], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 25)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 25), click_action = click_25_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 25), click_action = click_25_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_26_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_26_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 26)]]
      input[[paste0("scaling", 26)]]
      input[[paste0("init", 26)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 26)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 26)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 26)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[6], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 26)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 26), click_action = click_26_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 26), click_action = click_26_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_27_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_27_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 27)]]
      input[[paste0("scaling", 27)]]
      input[[paste0("init", 27)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 27)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 27)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 27)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[7], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 27)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 27), click_action = click_27_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 27), click_action = click_27_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_28_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_28_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 28)]]
      input[[paste0("scaling", 28)]]
      input[[paste0("init", 28)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 28)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 28)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 28)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[8], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 28)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 28), click_action = click_28_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 28), click_action = click_28_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_29_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_29_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 29)]]
      input[[paste0("scaling", 29)]]
      input[[paste0("init", 29)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 29)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 29)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 29)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[9], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 29)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 29), click_action = click_29_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 29), click_action = click_29_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_210_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_210_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 210)]]
      input[[paste0("scaling", 210)]]
      input[[paste0("init", 210)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 210)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 210)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 210)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[10], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 210)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 210), click_action = click_210_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 210), click_action = click_210_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_211_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[11]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_211_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[11]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 211)]]
      input[[paste0("scaling", 211)]]
      input[[paste0("init", 211)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 211)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 211)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 211)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[11], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 211)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 211), click_action = click_211_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 211), click_action = click_211_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_212_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[12]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_212_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[12]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 212)]]
      input[[paste0("scaling", 212)]]
      input[[paste0("init", 212)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 212)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 212)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 212)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[12], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 212)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 212), click_action = click_212_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 212), click_action = click_212_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_213_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[13]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_213_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[13]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 213)]]
      input[[paste0("scaling", 213)]]
      input[[paste0("init", 213)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 213)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 213)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 213)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[13], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 213)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 213), click_action = click_213_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 213), click_action = click_213_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_214_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[14]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_214_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[14]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 214)]]
      input[[paste0("scaling", 214)]]
      input[[paste0("init", 214)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 214)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 214)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 214)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[14], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 214)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 214), click_action = click_214_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 214), click_action = click_214_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_215_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[15]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_215_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[15]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 215)]]
      input[[paste0("scaling", 215)]]
      input[[paste0("init", 215)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 215)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 215)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 215)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[15], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 215)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 215), click_action = click_215_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 215), click_action = click_215_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_216_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[16]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_216_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[16]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 216)]]
      input[[paste0("scaling", 216)]]
      input[[paste0("init", 216)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 216)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 216)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 216)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[16], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 216)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 216), click_action = click_216_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 216), click_action = click_216_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_217_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[17]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[2]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[2], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_217_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[2]][[17]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[2]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[2]]),
  ]
  sig <- sig_deg[[2]][
    rownames(sig_deg[[2]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[2]][
    rownames(deg[[2]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[2]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[2]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[2]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 217)]]
      input[[paste0("scaling", 217)]]
      input[[paste0("init", 217)]]
    }, {

      sig <- sig_gse[[2]][
        rownames(sig_gse[[2]]) %in% input[[paste0("gse", 217)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 217)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[2]][rownames(gse[[2]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 217)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[2 + 3])
      } else {
        col <- colorRamp2(range_gse[[2]], colors = c("blue", "white", "red"))
        title <- data_types[2 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[2], "\n",
            "Significant ", names(priority_gse)[17], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[2]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[2]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 217)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 217), click_action = click_217_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 217), click_action = click_217_unscaled
        )
      }
    })
  },
)
```

FACS-sorted Single-soma RNAseq - Human AD BA9
===

```{r}
idx <- 3
sub_idx <- 0
```

Row {.tabset}
---

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%",
    " Of Tested Pathways Significant)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_31_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_31_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[1]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = "Click row to See proteins. Drag bottom-right corner to resize.",
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 31)]]
      input[[paste0("scaling", 31)]]
      input[[paste0("init", 31)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 31)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 31)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 31)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[3], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 31)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 31), click_action = click_31_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 31), click_action = click_31_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_32_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_32_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[2]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 32)]]
      input[[paste0("scaling", 32)]]
      input[[paste0("init", 32)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 32)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 32)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 32)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[3], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 32)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 32), click_action = click_32_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 32), click_action = click_32_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_33_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_33_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[3]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 33)]]
      input[[paste0("scaling", 33)]]
      input[[paste0("init", 33)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 33)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 33)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 33)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[3], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 33)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 33), click_action = click_33_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 33), click_action = click_33_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_34_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_34_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[4]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 34)]]
      input[[paste0("scaling", 34)]]
      input[[paste0("init", 34)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 34)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 34)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 34)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[4], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 34)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 34), click_action = click_34_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 34), click_action = click_34_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_35_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_35_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[5]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 35)]]
      input[[paste0("scaling", 35)]]
      input[[paste0("init", 35)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 35)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 35)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 35)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[5], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 35)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 35), click_action = click_35_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 35), click_action = click_35_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_36_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_36_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[6]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 36)]]
      input[[paste0("scaling", 36)]]
      input[[paste0("init", 36)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 36)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 36)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 36)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[6], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 36)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 36), click_action = click_36_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 36), click_action = click_36_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_37_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_37_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[7]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 37)]]
      input[[paste0("scaling", 37)]]
      input[[paste0("init", 37)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 37)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 37)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 37)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[7], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 37)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 37), click_action = click_37_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 37), click_action = click_37_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_38_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_38_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[8]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 38)]]
      input[[paste0("scaling", 38)]]
      input[[paste0("init", 38)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 38)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 38)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 38)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[8], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 38)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 38), click_action = click_38_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 38), click_action = click_38_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_39_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_39_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[9]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 39)]]
      input[[paste0("scaling", 39)]]
      input[[paste0("init", 39)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 39)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 39)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 39)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[9], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 39)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 39), click_action = click_39_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 39), click_action = click_39_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_310_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_310_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[10]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 310)]]
      input[[paste0("scaling", 310)]]
      input[[paste0("init", 310)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 310)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 310)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 310)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[10], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 310)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 310), click_action = click_310_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 310), click_action = click_310_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_311_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[11]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_311_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[11]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 311)]]
      input[[paste0("scaling", 311)]]
      input[[paste0("init", 311)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 311)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 311)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 311)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[11], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 311)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 311), click_action = click_311_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 311), click_action = click_311_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_312_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[12]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_312_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[12]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 312)]]
      input[[paste0("scaling", 312)]]
      input[[paste0("init", 312)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 312)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 312)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 312)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[12], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 312)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 312), click_action = click_312_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 312), click_action = click_312_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_313_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[13]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_313_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[13]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 313)]]
      input[[paste0("scaling", 313)]]
      input[[paste0("init", 313)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 313)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 313)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 313)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[13], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 313)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 313), click_action = click_313_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 313), click_action = click_313_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_314_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[14]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_314_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[14]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 314)]]
      input[[paste0("scaling", 314)]]
      input[[paste0("init", 314)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 314)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 314)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 314)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[14], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 314)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 314), click_action = click_314_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 314), click_action = click_314_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_315_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[15]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_315_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[15]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 315)]]
      input[[paste0("scaling", 315)]]
      input[[paste0("init", 315)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 315)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 315)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 315)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[15], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 315)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 315), click_action = click_315_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 315), click_action = click_315_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_316_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[16]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_316_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[16]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 316)]]
      input[[paste0("scaling", 316)]]
      input[[paste0("init", 316)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 316)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 316)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 316)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[16], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 316)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 316), click_action = click_316_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 316), click_action = click_316_unscaled
        )
      }
    })
  },
)
```

```{r results = "asis"}
cat(
  paste0(
    "### ", names(priority_gse)[sub_idx + 1],
    " (", n_gse[[idx]][[sub_idx + 1]], "%)"
  )
)
```

```{r}
sub_idx <- sub_idx + 1
comb_idx <- as.numeric(paste0(idx, sub_idx))

click_317_unscaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[17]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(range_deg[[3]], colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = data_types[3], direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

click_317_scaled <- function(df, output) {
  if (is.null(df)) {
    df$row_label <- top_gse[[3]][[17]][1]
  }

  features <- unlist(geneIds(gene_sets[names(gene_sets) %in% df$row_label]))
  gene_anno_sub <- gene_anno[gene_anno$ensembl %in% features, ]
  gene_anno_sub <- gene_anno_sub[order(gene_anno_sub$symbol), ]
  add_anno <- features[features %notin% rownames(deg[[3]])]
  add_anno <- gene_anno_sub[gene_anno_sub$ensembl %in% add_anno, ]
  gene_anno_sub <- gene_anno_sub[
    gene_anno_sub$symbol %in% rownames(sig_deg[[3]]),
  ]
  sig <- sig_deg[[3]][
    rownames(sig_deg[[3]]) %in% gene_anno_sub$symbol, , drop = FALSE
  ]
  if (nrow(sig) > 1) {
    sig <- sig[match(gene_anno_sub$symbol, rownames(sig)), , drop = FALSE]
  }

  mat <- deg[[3]][
    rownames(deg[[3]]) %in% gene_anno_sub$ensembl, , drop = FALSE
  ]
  if (nrow(mat) > 1) {
    mat <- mat[
      match(gene_anno_sub$ensembl, rownames(mat)), , drop = FALSE
    ]
  }
  if (nrow(mat) > 0) {
    rownames(mat) <- gene_anno_sub$symbol
  }

  add <- matrix(NA, nrow = nrow(add_anno), ncol = ncol(deg[[3]]))
  rownames(add) <- add_anno$symbol
  mat <- rbind(mat, add)
  sig <- rbind(sig, add)
  sig[is.na(sig)] <- ""

  mat <- t(
    apply(
      mat, MARGIN = 1,
      FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
    )
  )
  output$heatmap <- renderPlot({
    draw(
      Heatmap(
        mat,
        col = colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red")),
        na_col = "darkgrey",
        column_title = paste0(
          gene_sets[[df$row_label]]@shortDescription, "\n", df$row_label
        ),
        cluster_columns = FALSE,
        cluster_rows = FALSE,
        column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
        column_gap = unit(1, units = "mm"),
        rect_gp = gpar(col = "black", lwd = 0.25),
        row_names_gp = gpar(fontsize = 10),
        column_title_gp = gpar(fontsize = 14),
        row_title_rot = 0,
        top_annotation = top_anno[[3]],
        show_column_names = FALSE,
        cell_fun = function(j, i, x, y, width, height, fill) {
          grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 8))
        },
        heatmap_legend_param = list(
          title = paste0("Scaled ", data_types[3]), direction = "horizontal",
          title_position = "topcenter", labels_gp = gpar(fontsize = 10),
          grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
        )
      ),
      heatmap_legend_side = "top"
    )
  }, res = 80)
}

shinyApp(

  ui = fluidPage(
    selectizeInput(
      paste0("gse", comb_idx),
      label = "Type to Search Pathways:",
      choices = top_gse[[idx]][[sub_idx]],
      selected = top_gse[[idx]][[sub_idx]][1:20],
      multiple = TRUE,
      width = "100%"
    ),
    fluidRow(
      column(
        width = 3,
        radioButtons(
          paste0("scaling", comb_idx), label = NULL,
          c("Scaled", "Unscaled"), inline = TRUE
        )
      ),
      column(
        width = 1,
        actionButton(paste0("init", comb_idx), label = "Initialise"),
        style = 'padding-left: 120px;'
      ),
    ),
    InteractiveComplexHeatmapOutput(
      heatmap_id = paste0("ht", comb_idx),
      title1 = c(
        "Click row to see Proteins. Drag bottom-right corner to resize."
      ),
      title3 = c(
        "Arrows indicate significance at Adj. P < 0.05.
        Grey cells indicate lowly or undetected proteins."
      ),
      width1 = 1150, height1 = 425, response = "click",
      output_ui = plotOutput("heatmap", width = 675, height = 480)
    )
  ),
  
  server = function (input, output, session) {
    observeEvent({
      input[[paste0("gse", 317)]]
      input[[paste0("scaling", 317)]]
      input[[paste0("init", 317)]]
    }, {

      sig <- sig_gse[[3]][
        rownames(sig_gse[[3]]) %in% input[[paste0("gse", 317)]], , drop = FALSE
      ]
      if (nrow(sig) > 1) {
        sig <- sig[
          match(input[[paste0("gse", 317)]], rownames(sig)), , drop = FALSE
        ]
      }

      mat <- gse[[3]][rownames(gse[[3]]) %in% rownames(sig), , drop = FALSE]
      if (nrow(mat) > 1) {
        mat <- mat[match(rownames(sig), rownames(mat)), , drop = FALSE]
      }

      if (input[[paste0("scaling", 317)]] == "Scaled") {
        mat <- t(
          apply(
            mat, MARGIN = 1,
            FUN = function (x) ((2 * (x - min(x)) / (max(x) - min(x))) - 1)
          )
        )
        col <- colorRamp2(c(-1, 0, 1), colors = c("blue", "white", "red"))
        title <- paste0("Scaled ", data_types[3 + 3])
      } else {
        col <- colorRamp2(range_gse[[3]], colors = c("blue", "white", "red"))
        title <- data_types[3 + 3]
      }

      ht <- draw(
        Heatmap(
          mat,
          col = col,
          column_title = paste0(
            data_sets[3], "\n",
            "Significant ", names(priority_gse)[17], " Pathways ",
            "(Adj. P < 0.05)"
          ),
          cluster_columns = FALSE,
          cluster_rows = FALSE,
          column_split = rep(seq(2), each = nrow(meta[[3]]) / 2),
          column_gap = unit(1, units = "mm"),
          rect_gp = gpar(col = "black", lwd = 0.25),
          row_names_gp = gpar(fontsize = 13),
          column_title_gp = gpar(fontsize = 14),
          row_title_rot = 0,
          top_annotation = top_anno[[3]],
          show_column_names = FALSE,
          row_names_max_width = max_text_width(
            rownames(mat), gp = gpar(fontsize = 14)
          ),
          cell_fun = function(j, i, x, y, width, height, fill) {
            grid.text(sig[i, j], x = x, y = y, gp = gpar(fontsize = 11))
          },
          heatmap_legend_param = list(
            title = title, direction = "horizontal",
            title_position = "topcenter", labels_gp = gpar(fontsize = 10),
            grid_height = unit(3, "mm"), title_gp = gpar(fontsize = 12)
          )
        ),
        heatmap_legend_side = "top"
      )

      if (input[[paste0("scaling", 317)]] == "Scaled") {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 317), click_action = click_317_scaled
        )
      } else {
        makeInteractiveComplexHeatmap(
          input, output, session, ht, res = 80,
          heatmap_id = paste0("ht", 217), click_action = click_317_unscaled
        )
      }
    })
  },
)
```

```{r, include = FALSE}
rm(p_adj, anno_colour, cache_dir, data_dir, priority_gse)
gc()
```
